<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.1.189">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>alphabase - Fragment Functionalities</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="../styles.css">
<meta property="og:title" content="alphabase - Fragment Functionalities">
<meta property="og:description" content="Just in case that we have two modifications, one is on the peptide N-term, and the other is on the N-term AA site chain. Similar for C-term sites.">
<meta property="og:site-name" content="alphabase">
<meta name="twitter:title" content="alphabase - Fragment Functionalities">
<meta name="twitter:description" content="Just in case that we have two modifications, one is on the peptide N-term, and the other is on the N-term AA site chain. Similar for C-term sites.">
<meta name="twitter:card" content="summary">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">alphabase</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/MannLabs/alphabase"><i class="bi bi-github" role="img">
</i> 
 </a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">Fragment Functionalities</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">AlphaBase</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../utils.html" class="sidebar-item-text sidebar-link">Utils</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../yaml_utils.html" class="sidebar-item-text sidebar-link">YAML Utils</a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">constants</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../constants/aa.html" class="sidebar-item-text sidebar-link">Amino acid information</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../constants/element.html" class="sidebar-item-text sidebar-link">Atom element information</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../constants/isotope.html" class="sidebar-item-text sidebar-link">Isotope distribution</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../constants/modification.html" class="sidebar-item-text sidebar-link">Modification information</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">io</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../io/hdf.html" class="sidebar-item-text sidebar-link">HDF functionalities</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">peptide</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../peptide/fragment.html" class="sidebar-item-text sidebar-link active">Fragment Functionalities</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../peptide/mass_calc.html" class="sidebar-item-text sidebar-link">Mass Calculation</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../peptide/mobility.html" class="sidebar-item-text sidebar-link">CCS/Mobility Functionalities</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../peptide/precursor.html" class="sidebar-item-text sidebar-link">Precursor Functionalities</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">protein</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../protein/fasta.html" class="sidebar-item-text sidebar-link">Protein and Peptide Processing</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../protein/test_fasta.html" class="sidebar-item-text sidebar-link">Testing fasta</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">psm_reader</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../psm_reader/alphapept_reader.html" class="sidebar-item-text sidebar-link">AlphaPept PSM Reader</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../psm_reader/dia_psm_reader.html" class="sidebar-item-text sidebar-link">DIA PSM reader</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../psm_reader/maxquant_reader.html" class="sidebar-item-text sidebar-link">MaxQuant PSM reader</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../psm_reader/msfragger_reader.html" class="sidebar-item-text sidebar-link">MSFragger Reader</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../psm_reader/pfind_reader.html" class="sidebar-item-text sidebar-link">pFind PSM Reader</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../psm_reader/psm_reader.html" class="sidebar-item-text sidebar-link">Base Class for PSM Readers</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true">spectral_library</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../spectral_library/decoy_library.html" class="sidebar-item-text sidebar-link">Decoy Libraries</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../spectral_library/library_base.html" class="sidebar-item-text sidebar-link">Base Class for Spectral Libraries</a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#first-it-is-worth-mentioning-that-in-alphabase" id="toc-first-it-is-worth-mentioning-that-in-alphabase" class="nav-link active" data-scroll-target="#first-it-is-worth-mentioning-that-in-alphabase">First, it is worth mentioning that, in AlphaBase:</a></li>
  <li><a href="#parse_charged_frag_type" id="toc-parse_charged_frag_type" class="nav-link" data-scroll-target="#parse_charged_frag_type">parse_charged_frag_type</a></li>
  <li><a href="#get_charged_frag_types" id="toc-get_charged_frag_types" class="nav-link" data-scroll-target="#get_charged_frag_types">get_charged_frag_types</a></li>
  <li><a href="#fragment-dataframe-processing" id="toc-fragment-dataframe-processing" class="nav-link" data-scroll-target="#fragment-dataframe-processing">Fragment dataframe processing</a>
  <ul class="collapse">
  <li><a href="#init_fragment_by_precursor_dataframe" id="toc-init_fragment_by_precursor_dataframe" class="nav-link" data-scroll-target="#init_fragment_by_precursor_dataframe">init_fragment_by_precursor_dataframe</a></li>
  <li><a href="#init_fragment_dataframe_from_other" id="toc-init_fragment_dataframe_from_other" class="nav-link" data-scroll-target="#init_fragment_dataframe_from_other">init_fragment_dataframe_from_other</a></li>
  <li><a href="#init_zero_fragment_dataframe" id="toc-init_zero_fragment_dataframe" class="nav-link" data-scroll-target="#init_zero_fragment_dataframe">init_zero_fragment_dataframe</a></li>
  <li><a href="#get_sliced_fragment_dataframe" id="toc-get_sliced_fragment_dataframe" class="nav-link" data-scroll-target="#get_sliced_fragment_dataframe">get_sliced_fragment_dataframe</a></li>
  <li><a href="#update_sliced_fragment_dataframe" id="toc-update_sliced_fragment_dataframe" class="nav-link" data-scroll-target="#update_sliced_fragment_dataframe">update_sliced_fragment_dataframe</a></li>
  <li><a href="#concat_precursor_fragment_dataframes" id="toc-concat_precursor_fragment_dataframes" class="nav-link" data-scroll-target="#concat_precursor_fragment_dataframes">concat_precursor_fragment_dataframes</a></li>
  <li><a href="#calc_fragment_mz_values_for_same_naa" id="toc-calc_fragment_mz_values_for_same_naa" class="nav-link" data-scroll-target="#calc_fragment_mz_values_for_same_naa">calc_fragment_mz_values_for_same_nAA</a></li>
  <li><a href="#mask_fragments_for_charge_greater_than_precursor_charge" id="toc-mask_fragments_for_charge_greater_than_precursor_charge" class="nav-link" data-scroll-target="#mask_fragments_for_charge_greater_than_precursor_charge">mask_fragments_for_charge_greater_than_precursor_charge</a></li>
  </ul></li>
  <li><a href="#create-fragment-mz-dataframe" id="toc-create-fragment-mz-dataframe" class="nav-link" data-scroll-target="#create-fragment-mz-dataframe">Create fragment mz dataframe</a>
  <ul class="collapse">
  <li><a href="#create_fragment_mz_dataframe" id="toc-create_fragment_mz_dataframe" class="nav-link" data-scroll-target="#create_fragment_mz_dataframe">create_fragment_mz_dataframe</a></li>
  <li><a href="#create_fragment_mz_dataframe_by_sort_precursor" id="toc-create_fragment_mz_dataframe_by_sort_precursor" class="nav-link" data-scroll-target="#create_fragment_mz_dataframe_by_sort_precursor">create_fragment_mz_dataframe_by_sort_precursor</a></li>
  <li><a href="#examples-and-unittests" id="toc-examples-and-unittests" class="nav-link" data-scroll-target="#examples-and-unittests">Examples and unittests:</a></li>
  <li><a href="#test-mod-deltas" id="toc-test-mod-deltas" class="nav-link" data-scroll-target="#test-mod-deltas">Test mod deltas</a></li>
  </ul></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/MannLabs/alphabase/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block">Fragment Functionalities</h1>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header>

<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<section id="first-it-is-worth-mentioning-that-in-alphabase" class="level3">
<h3 class="anchored" data-anchor-id="first-it-is-worth-mentioning-that-in-alphabase">First, it is worth mentioning that, in AlphaBase:</h3>
<ol type="1">
<li>peptide N-term modification site is 0</li>
<li>C-term modification site is -1</li>
<li>other modifications sites are integers from 1 to nAA</li>
</ol>
<p>Just in case that we have two modifications, one is on the peptide N-term, and the other is on the N-term AA site chain. Similar for C-term sites.</p>
<hr>
<p><a href="https://github.com/MannLabs/alphabase/blob/main/alphabase/peptide/fragment.py#L64" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="parse_charged_frag_type" class="level3">
<h3 class="anchored" data-anchor-id="parse_charged_frag_type">parse_charged_frag_type</h3>
<blockquote class="blockquote">
<pre><code> parse_charged_frag_type (charged_frag_type:str)</code></pre>
</blockquote>
<p>Oppsite to <code>get_charged_frag_types</code>.</p>
<table class="table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>charged_frag_type</td>
<td>str</td>
<td>e.g.&nbsp;‘y_z1’, ‘b_modloss_z1’</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>typing.Tuple[str, int]</strong></td>
<td><strong>str. Fragment type, e.g.&nbsp;‘b’,‘y’<br><br>int. Charge state</strong></td>
</tr>
</tbody>
</table>
<pre><code>/Users/zengwenfeng/opt/anaconda3/lib/python3.8/site-packages/fastcore/docscrape.py:225: UserWarning: Unknown section Examples
  else: warn(msg)</code></pre>
<hr>
<p><a href="https://github.com/MannLabs/alphabase/blob/main/alphabase/peptide/fragment.py#L32" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="get_charged_frag_types" class="level3">
<h3 class="anchored" data-anchor-id="get_charged_frag_types">get_charged_frag_types</h3>
<blockquote class="blockquote">
<pre><code> get_charged_frag_types (frag_types:List[str], max_frag_charge:int=2)</code></pre>
</blockquote>
<p>Combine fragment types and charge states.</p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>frag_types</td>
<td>typing.List[str]</td>
<td></td>
<td>e.g.&nbsp;[‘b’,‘y’,‘b_modloss’,‘y_modloss’]</td>
</tr>
<tr class="even">
<td>max_frag_charge</td>
<td>int</td>
<td>2</td>
<td>max fragment charge. (default: 2)</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>typing.List[str]</strong></td>
<td></td>
<td><strong>charged fragment types</strong></td>
</tr>
</tbody>
</table>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> np.<span class="bu">all</span>(np.array(get_charged_frag_types([<span class="st">'b'</span>,<span class="st">'b_modloss'</span>],<span class="dv">2</span>))<span class="op">==</span>np.array([<span class="st">'b_z1'</span>, <span class="st">'b_z2'</span>, <span class="st">'b_modloss_z1'</span>, <span class="st">'b_modloss_z2'</span>]))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> parse_charged_frag_type(<span class="st">'b_z2'</span>)<span class="op">==</span>(<span class="st">'b'</span>,<span class="dv">2</span>)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> parse_charged_frag_type(<span class="st">'b_modloss_z2'</span>)<span class="op">==</span>(<span class="st">'b_modloss'</span>, <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="fragment-dataframe-processing" class="level1">
<h1>Fragment dataframe processing</h1>
<p>In AlphaX Ecosystem, library fragments are stored in a dataframe, where the columns are charged_frag_types (<code>['b_z1','b_z2','y_z1','y_z2','b_modloss_z1','y_H2O_z1'...]</code>) and the rows are corresponding positions (starting with peptide N-term) of the fragments. Library precursor/peptide dataframe must contain <code>frag_start_idx</code> and <code>frag_end_idx</code> columns to tell us where are the fragments of each precursor/peptide.</p>
<p>We provide different ways to initialize fragment dataframes, see below:</p>
<hr>
<p><a href="https://github.com/MannLabs/alphabase/blob/main/alphabase/peptide/fragment.py#L135" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="init_fragment_by_precursor_dataframe" class="level3">
<h3 class="anchored" data-anchor-id="init_fragment_by_precursor_dataframe">init_fragment_by_precursor_dataframe</h3>
<blockquote class="blockquote">
<pre><code> init_fragment_by_precursor_dataframe (precursor_df,
                                       charged_frag_types:List[str], refer
                                       ence_fragment_df:pandas.core.frame.
                                       DataFrame=None,
                                       dtype:numpy.dtype=&lt;class
                                       'numpy.float64'&gt;,
                                       inplace_in_reference:bool=False)</code></pre>
</blockquote>
<p>Init zero fragment dataframe for the <code>precursor_df</code>. If the <code>reference_fragment_df</code> is provided, the result dataframe’s length will be the same as reference_fragment_df. Otherwise it generates the dataframe from scratch.</p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>precursor_df</td>
<td>pd.DataFrame</td>
<td></td>
<td>precursors to generate fragment masses,<br>if <code>precursor_df</code> contains the ‘frag_start_idx’ column, <br>it is better to provide <code>reference_fragment_df</code> as <br><code>precursor_df.frag_start_idx</code> and <code>precursor.frag_end_idx</code> <br>point to the indices in <code>reference_fragment_df</code></td>
</tr>
<tr class="even">
<td>charged_frag_types</td>
<td>typing.List[str]</td>
<td></td>
<td><code>['b_z1','b_z2','y_z1','y_z2','b_modloss_z1','y_H2O_z1'...]</code></td>
</tr>
<tr class="odd">
<td>reference_fragment_df</td>
<td>DataFrame</td>
<td>None</td>
<td>init zero fragment_mz_df based<br>on this reference. If None, fragment_mz_df will be <br>initialized by :func:<code>alphabase.peptide.fragment.init_zero_fragment_dataframe</code>.<br>Defaults to None.</td>
</tr>
<tr class="even">
<td>dtype</td>
<td>dtype</td>
<td>float64</td>
<td></td>
</tr>
<tr class="odd">
<td>inplace_in_reference</td>
<td>bool</td>
<td>False</td>
<td>if calculate the fragment mz <br>inplace in the reference_fragment_df (default: False)</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>pd.DataFrame</strong></td>
<td></td>
<td><strong>zero <code>fragment_df</code> with given <code>charged_frag_types</code> columns</strong></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/MannLabs/alphabase/blob/main/alphabase/peptide/fragment.py#L123" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="init_fragment_dataframe_from_other" class="level3">
<h3 class="anchored" data-anchor-id="init_fragment_dataframe_from_other">init_fragment_dataframe_from_other</h3>
<blockquote class="blockquote">
<pre><code> init_fragment_dataframe_from_other
                                     (reference_fragment_df:pandas.core.fr
                                     ame.DataFrame, dtype=&lt;class
                                     'numpy.float64'&gt;)</code></pre>
</blockquote>
<p>Init zero fragment dataframe from the <code>reference_fragment_df</code> (same rows and same columns)</p>
<hr>
<p><a href="https://github.com/MannLabs/alphabase/blob/main/alphabase/peptide/fragment.py#L88" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="init_zero_fragment_dataframe" class="level3">
<h3 class="anchored" data-anchor-id="init_zero_fragment_dataframe">init_zero_fragment_dataframe</h3>
<blockquote class="blockquote">
<pre><code> init_zero_fragment_dataframe (peplen_array:numpy.ndarray,
                               charged_frag_types:List[str], dtype=&lt;class
                               'numpy.float64'&gt;)</code></pre>
</blockquote>
<p>Initialize a zero dataframe based on peptide length (nAA) array (peplen_array) and charge_frag_types (column number). The row number of returned dataframe is np.sum(peplen_array-1).</p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>peplen_array</td>
<td>ndarray</td>
<td></td>
<td>peptide lengths for the fragment dataframe</td>
</tr>
<tr class="even">
<td>charged_frag_types</td>
<td>typing.List[str]</td>
<td></td>
<td><code>['b_z1','b_z2','y_z1','y_z2','b_modloss_z1','y_H2O_z1'...]</code></td>
</tr>
<tr class="odd">
<td>dtype</td>
<td>type</td>
<td>float64</td>
<td></td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>typing.Tuple[pandas.core.frame.DataFrame, numpy.ndarray, numpy.ndarray]</strong></td>
<td></td>
<td><strong>pd.DataFrame, <code>fragment_df</code> with zero values<br><br>np.ndarray (int64), the start indices point to the <code>fragment_df</code> for each peptide<br><br>np.ndarray (int64), the end indices point to the <code>fragment_df</code> for each peptide</strong></td>
</tr>
</tbody>
</table>
<p>For a subset of the precursor dataframe, we need to set or get fragment values for the slicing (by <code>frag_start_idx</code> and <code>frag_end_idx</code>in <code>precursor_df</code>) of the fragment dataframe. We use <a href="https://MannLabs.github.io/alphabase/peptide/fragment.html#update_sliced_fragment_dataframe"><code>update_sliced_fragment_dataframe</code></a> to set the values, and <a href="https://MannLabs.github.io/alphabase/peptide/fragment.html#get_sliced_fragment_dataframe"><code>get_sliced_fragment_dataframe</code></a> to get values.</p>
<hr>
<p><a href="https://github.com/MannLabs/alphabase/blob/main/alphabase/peptide/fragment.py#L268" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="get_sliced_fragment_dataframe" class="level3">
<h3 class="anchored" data-anchor-id="get_sliced_fragment_dataframe">get_sliced_fragment_dataframe</h3>
<blockquote class="blockquote">
<pre><code> get_sliced_fragment_dataframe (fragment_df:pandas.core.frame.DataFrame,
                                frag_start_end_list:Union[List,numpy.ndarr
                                ay], charged_frag_types:List=None)</code></pre>
</blockquote>
<p>Get the sliced fragment_df from <code>frag_start_end_list=[(start,end),(start,end),...]</code>.</p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>fragment_df</td>
<td>DataFrame</td>
<td></td>
<td>fragment dataframe to get values</td>
</tr>
<tr class="even">
<td>frag_start_end_list</td>
<td>typing.Union[typing.List, numpy.ndarray]</td>
<td></td>
<td>List[Tuple[int,int]], e.g.&nbsp;<code>[(start,end),(start,end),...]</code> or np.ndarray</td>
</tr>
<tr class="odd">
<td>charged_frag_types</td>
<td>typing.List</td>
<td>None</td>
<td>e.g.&nbsp;<code>['b_z1','b_z2','y_z1','y_z2']</code>.<br>if None, all columns will be considered</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>DataFrame</strong></td>
<td></td>
<td><strong>sliced fragment_df. If <code>charged_frag_types</code> is None, <br>return fragment_df with all columns</strong></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/MannLabs/alphabase/blob/main/alphabase/peptide/fragment.py#L226" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="update_sliced_fragment_dataframe" class="level3">
<h3 class="anchored" data-anchor-id="update_sliced_fragment_dataframe">update_sliced_fragment_dataframe</h3>
<blockquote class="blockquote">
<pre><code> update_sliced_fragment_dataframe
                                   (fragment_df:pandas.core.frame.DataFram
                                   e, values:numpy.ndarray, frag_start_end
                                   _list:List[Tuple[int,int]],
                                   charged_frag_types:List[str]=None)</code></pre>
</blockquote>
<p>Set the values of the slices <code>frag_start_end_list=[(start,end),(start,end),...]</code> of fragment_df.</p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>fragment_df</td>
<td>DataFrame</td>
<td></td>
<td>fragment dataframe to set the values</td>
</tr>
<tr class="even">
<td>values</td>
<td>ndarray</td>
<td></td>
<td>values to set</td>
</tr>
<tr class="odd">
<td>frag_start_end_list</td>
<td>typing.List[typing.Tuple[int, int]]</td>
<td></td>
<td>e.g.&nbsp;<code>[(start,end),(start,end),...]</code></td>
</tr>
<tr class="even">
<td>charged_frag_types</td>
<td>typing.List[str]</td>
<td>None</td>
<td>e.g.&nbsp;<code>['b_z1','b_z2','y_z1','y_z2']</code>.<br>If None, the columns of values should be the same as fragment_df’s columns.<br>It is much faster if charged_frag_types is None as we use numpy slicing, <br>otherwise we use pd.loc (much slower).<br>Defaults to None.</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>DataFrame</strong></td>
<td></td>
<td><strong>fragment_df after the values are set into slices</strong></td>
</tr>
</tbody>
</table>
<p>For some search engines, it reports different result files for different raw files. After load them separately, we concatenate <code>precursor_df_list</code> and <code>fragment_df_list</code> into single dataframes respectively. The main processing here is to cumulate <code>frag_start_idx</code> and <code>frag_end_idx</code> for different <code>precursor_df</code>s.</p>
<hr>
<p><a href="https://github.com/MannLabs/alphabase/blob/main/alphabase/peptide/fragment.py#L304" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="concat_precursor_fragment_dataframes" class="level3">
<h3 class="anchored" data-anchor-id="concat_precursor_fragment_dataframes">concat_precursor_fragment_dataframes</h3>
<blockquote class="blockquote">
<pre><code> concat_precursor_fragment_dataframes
                                       (precursor_df_list:List[pandas.core
                                       .frame.DataFrame], fragment_df_list
                                       :List[pandas.core.frame.DataFrame],
                                       *other_fragment_df_lists)</code></pre>
</blockquote>
<p>Since fragment_df is indexed by precursor_df, when we concatenate multiple fragment_df, the indexed positions will change for in precursor_dfs,<br>
this function keeps the correct indexed positions of precursor_df when concatenating multiple fragment_df dataframes.</p>
<table class="table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>precursor_df_list</td>
<td>typing.List[pandas.core.frame.DataFrame]</td>
<td>precursor dataframe list to concatenate</td>
</tr>
<tr class="even">
<td>fragment_df_list</td>
<td>typing.List[pandas.core.frame.DataFrame]</td>
<td>fragment dataframe list to concatenate</td>
</tr>
<tr class="odd">
<td>other_fragment_df_lists</td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>typing.Tuple[pandas.core.frame.DataFrame, …]</strong></td>
<td>**concatenated precursor_df, fragment_df, *other_fragment_df …**</td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/MannLabs/alphabase/blob/main/alphabase/peptide/fragment.py#L346" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="calc_fragment_mz_values_for_same_naa" class="level3">
<h3 class="anchored" data-anchor-id="calc_fragment_mz_values_for_same_naa">calc_fragment_mz_values_for_same_nAA</h3>
<blockquote class="blockquote">
<pre><code> calc_fragment_mz_values_for_same_nAA
                                       (df_group:pandas.core.frame.DataFra
                                       me, nAA:int,
                                       charged_frag_types:list)</code></pre>
</blockquote>
<hr>
<p><a href="https://github.com/MannLabs/alphabase/blob/main/alphabase/peptide/fragment.py#L443" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="mask_fragments_for_charge_greater_than_precursor_charge" class="level3">
<h3 class="anchored" data-anchor-id="mask_fragments_for_charge_greater_than_precursor_charge">mask_fragments_for_charge_greater_than_precursor_charge</h3>
<blockquote class="blockquote">
<pre><code> mask_fragments_for_charge_greater_than_precursor_charge
                                                          (fragment_df:pan
                                                          das.core.frame.D
                                                          ataFrame, precur
                                                          sor_charge_array
                                                          :numpy.ndarray, 
                                                          nAA_array:numpy.
                                                          ndarray, candida
                                                          te_fragment_char
                                                          ges:list=[2, 3,
                                                          4])</code></pre>
</blockquote>
<p>Mask the fragment dataframe when the fragment charge is larger than the precursor charge</p>
</section>
</section>
<section id="create-fragment-mz-dataframe" class="level1">
<h1>Create fragment mz dataframe</h1>
<p>This is one of the most important functions in alphabase. For a given <code>precursor_df</code>, it calculates the fragment ion dataframe, and also set the <code>frag_start_idx</code> and <code>frag_end_idx</code> column values to connect the <code>precursor_df</code> and <code>fragment_mz_df</code>.</p>
<p>When creating a new fragment mz/intensity dataframes for a precursor, alphabase will check if <code>frag_start_idx</code> exists. As the <code>frag_start_idx</code> points to an existing fragment dataframe (refers to <code>reference_frag_df</code>), so we have to provide the <code>reference_frag_df</code> to make sure that <code>reference_frag_df</code> and newly created fragment_df are consisitent.</p>
<p>For the more convenient and faster calculation, we should do as follows: - Sort <code>precursor_df</code> by ‘nAA’ (<code>precursor_df.sort_values('nAA', inplace=True)</code>) to make sure groupby(‘nAA’) will not change the order of the <code>precursor_df</code>. - Reset index (<code>precursor_df.reset_index(drop=True, inplace=True)</code>) to make sure iloc and loc will index the same dataframe subset. - Delete <code>frag_start_idx</code> and <code>frag_end_idx</code> columns if they exist, otherwise the creation speed wil be slower. - Call <code>create_fragment_mz_dataframe_by_sort_precursor(precursor_df, charged_frag_types)</code> or <code>create_fragment_mz_dataframe(precursor_df, charged_frag_types)</code>. <a href="https://MannLabs.github.io/alphabase/peptide/fragment.html#create_fragment_mz_dataframe"><code>create_fragment_mz_dataframe</code></a> will also call <a href="https://MannLabs.github.io/alphabase/peptide/fragment.html#create_fragment_mz_dataframe_by_sort_precursor"><code>create_fragment_mz_dataframe_by_sort_precursor</code></a> if there is no <code>frag_start_idx</code> column. - If we need to predict/calculate <code>fragment_intensity_df</code>, we can redo step 3 (delete frag idxes columns) and then call ‘intensity prediction’ or ‘intensity calculation’.</p>
<hr>
<p><a href="https://github.com/MannLabs/alphabase/blob/main/alphabase/peptide/fragment.py#L520" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="create_fragment_mz_dataframe" class="level3">
<h3 class="anchored" data-anchor-id="create_fragment_mz_dataframe">create_fragment_mz_dataframe</h3>
<blockquote class="blockquote">
<pre><code> create_fragment_mz_dataframe (precursor_df:pandas.core.frame.DataFrame,
                               charged_frag_types:List, reference_fragment
                               _df:pandas.core.frame.DataFrame=None,
                               inplace_in_reference:bool=False,
                               batch_size:int=500000)</code></pre>
</blockquote>
<p>Generate fragment mass dataframe for the precursor_df. If the <code>reference_fragment_df</code> is provided and precursor_df contains <code>frag_start_idx</code>, it will generate the mz dataframe based on the reference. Otherwise it generates the mz dataframe from scratch.</p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>precursor_df</td>
<td>DataFrame</td>
<td></td>
<td>precursors to generate fragment masses,<br>if <code>precursor_df</code> contains the ‘frag_start_idx’ column, <br><code>reference_fragment_df</code> must be provided</td>
</tr>
<tr class="even">
<td>charged_frag_types</td>
<td>typing.List</td>
<td></td>
<td><code>['b_z1','b_z2','y_z1','y_z2','b_modloss_1','y_H2O_z1'...]</code></td>
</tr>
<tr class="odd">
<td>reference_fragment_df</td>
<td>DataFrame</td>
<td>None</td>
<td>kwargs only. Generate fragment_mz_df based on this reference, <br>as <code>precursor_df.frag_start_idx</code> and <br><code>precursor.frag_end_idx</code> point to the indices in <br><code>reference_fragment_df</code>.<br>Defaults to None</td>
</tr>
<tr class="even">
<td>inplace_in_reference</td>
<td>bool</td>
<td>False</td>
<td>kwargs only. Change values in place in the <code>reference_fragment_df</code>.<br>Defaults to False</td>
</tr>
<tr class="odd">
<td>batch_size</td>
<td>int</td>
<td>500000</td>
<td></td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>DataFrame</strong></td>
<td></td>
<td><strong><code>fragment_mz_df</code> with given <code>charged_frag_types</code></strong></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/MannLabs/alphabase/blob/main/alphabase/peptide/fragment.py#L464" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="create_fragment_mz_dataframe_by_sort_precursor" class="level3">
<h3 class="anchored" data-anchor-id="create_fragment_mz_dataframe_by_sort_precursor">create_fragment_mz_dataframe_by_sort_precursor</h3>
<blockquote class="blockquote">
<pre><code> create_fragment_mz_dataframe_by_sort_precursor
                                                 (precursor_df:pandas.core
                                                 .frame.DataFrame,
                                                 charged_frag_types:List,
                                                 batch_size:int=500000)</code></pre>
</blockquote>
<p>Sort nAA in precursor_df for faster fragment mz dataframe creation.</p>
<p>Because the fragment mz values are continous in memory, so it is faster when setting values in pandas.</p>
<p>Note that this function will change the order and index of precursor_df</p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>precursor_df</td>
<td>DataFrame</td>
<td></td>
<td>precursor dataframe</td>
</tr>
<tr class="even">
<td>charged_frag_types</td>
<td>typing.List</td>
<td></td>
<td>fragment types list</td>
</tr>
<tr class="odd">
<td>batch_size</td>
<td>int</td>
<td>500000</td>
<td>Calculate fragment mz values in batch. <br>Defaults to 500000.</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>DataFrame</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
</section>
<section id="examples-and-unittests" class="level3">
<h3 class="anchored" data-anchor-id="examples-and-unittests">Examples and unittests:</h3>
<p>Test <a href="https://MannLabs.github.io/alphabase/peptide/fragment.html#create_fragment_mz_dataframe_by_sort_precursor"><code>create_fragment_mz_dataframe_by_sort_precursor</code></a></p>
<p><a href="https://MannLabs.github.io/alphabase/peptide/fragment.html#create_fragment_mz_dataframe_by_sort_precursor"><code>create_fragment_mz_dataframe_by_sort_precursor</code></a> will sort <code>nAA</code> columns in <code>precursor_df</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>repeat <span class="op">=</span> <span class="dv">2</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>peptides <span class="op">=</span> [<span class="st">'AGHCEWQMKAADER'</span>]<span class="op">*</span>repeat</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>mods <span class="op">=</span> [<span class="st">'Acetyl@Protein N-term;Carbamidomethyl@C;Oxidation@M'</span>]<span class="op">*</span>repeat</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>sites <span class="op">=</span> [<span class="st">'0;4;8'</span>]<span class="op">*</span>repeat</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>peptides <span class="op">+=</span> [<span class="st">'AGHCEWQMK'</span>]<span class="op">*</span>repeat</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>mods <span class="op">+=</span> [<span class="st">''</span>]<span class="op">*</span>repeat</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>sites <span class="op">+=</span> [<span class="st">''</span>]<span class="op">*</span>repeat</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>precursor_df <span class="op">=</span> pd.DataFrame({</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">'sequence'</span>: peptides,</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>    <span class="st">'mods'</span>: mods,</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>    <span class="st">'mod_sites'</span>: sites</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>precursor_df[<span class="st">'nAA'</span>] <span class="op">=</span> precursor_df[<span class="st">'sequence'</span>].<span class="bu">str</span>.<span class="bu">len</span>()</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>precursor_df[<span class="st">'charge'</span>] <span class="op">=</span> [<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">2</span>]</span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>precursor_df <span class="op">=</span> update_precursor_mz(precursor_df)</span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a>fragment_mz_df <span class="op">=</span> create_fragment_mz_dataframe_by_sort_precursor(</span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a>    precursor_df,</span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a>    get_charged_frag_types([<span class="st">'b'</span>,<span class="st">'y'</span>,<span class="st">'b_modloss'</span>,<span class="st">'y_modloss'</span>],<span class="dv">2</span>)</span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> precursor_df.nAA.is_monotonic</span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> np.allclose(precursor_df.precursor_mz.values, [<span class="fl">545.233862</span>, <span class="fl">545.233862</span>, <span class="fl">1746.732265</span>, <span class="fl">1746.732265</span>])</span>
<span id="cb16-24"><a href="#cb16-24" aria-hidden="true" tabindex="-1"></a>ith_pep <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb16-25"><a href="#cb16-25" aria-hidden="true" tabindex="-1"></a>frag_start, frag_end <span class="op">=</span> precursor_df[[<span class="st">'frag_start_idx'</span>,<span class="st">'frag_end_idx'</span>]].values[ith_pep]</span>
<span id="cb16-26"><a href="#cb16-26" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> np.allclose(fragment_mz_df.iloc[frag_start:frag_end][<span class="st">'b_z1'</span>].values, </span>
<span id="cb16-27"><a href="#cb16-27" aria-hidden="true" tabindex="-1"></a>        [ <span class="fl">72.04439025</span>,  <span class="fl">129.06585397</span>,  <span class="fl">266.12476583</span>,  <span class="fl">369.13395079</span>,</span>
<span id="cb16-28"><a href="#cb16-28" aria-hidden="true" tabindex="-1"></a>        <span class="fl">498.17654388</span>,  <span class="fl">684.25585683</span>,  <span class="fl">812.31443434</span>,  <span class="fl">943.35491942</span>]</span>
<span id="cb16-29"><a href="#cb16-29" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb16-30"><a href="#cb16-30" aria-hidden="true" tabindex="-1"></a>ith_pep <span class="op">=</span> <span class="dv">2</span></span>
<span id="cb16-31"><a href="#cb16-31" aria-hidden="true" tabindex="-1"></a>frag_start, frag_end <span class="op">=</span> precursor_df[[<span class="st">'frag_start_idx'</span>,<span class="st">'frag_end_idx'</span>]].values[ith_pep]</span>
<span id="cb16-32"><a href="#cb16-32" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> (fragment_mz_df.iloc[frag_start:frag_end][<span class="st">'b_z2'</span>].values<span class="op">==</span><span class="dv">0</span>).<span class="bu">all</span>()</span>
<span id="cb16-33"><a href="#cb16-33" aria-hidden="true" tabindex="-1"></a>precursor_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>sequence</th>
      <th>mods</th>
      <th>mod_sites</th>
      <th>nAA</th>
      <th>charge</th>
      <th>precursor_mz</th>
      <th>frag_start_idx</th>
      <th>frag_end_idx</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>AGHCEWQMK</td>
      <td></td>
      <td></td>
      <td>9</td>
      <td>2</td>
      <td>545.233862</td>
      <td>0</td>
      <td>8</td>
    </tr>
    <tr>
      <th>1</th>
      <td>AGHCEWQMK</td>
      <td></td>
      <td></td>
      <td>9</td>
      <td>2</td>
      <td>545.233862</td>
      <td>8</td>
      <td>16</td>
    </tr>
    <tr>
      <th>2</th>
      <td>AGHCEWQMKAADER</td>
      <td>Acetyl@Protein N-term;Carbamidomethyl@C;Oxidat...</td>
      <td>0;4;8</td>
      <td>14</td>
      <td>1</td>
      <td>1746.732265</td>
      <td>16</td>
      <td>29</td>
    </tr>
    <tr>
      <th>3</th>
      <td>AGHCEWQMKAADER</td>
      <td>Acetyl@Protein N-term;Carbamidomethyl@C;Oxidat...</td>
      <td>0;4;8</td>
      <td>14</td>
      <td>1</td>
      <td>1746.732265</td>
      <td>29</td>
      <td>42</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
<p>Test <a href="https://MannLabs.github.io/alphabase/peptide/fragment.html#get_sliced_fragment_dataframe"><code>get_sliced_fragment_dataframe</code></a> and <a href="https://MannLabs.github.io/alphabase/peptide/fragment.html#update_sliced_fragment_dataframe"><code>update_sliced_fragment_dataframe</code></a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>sliced_frag_df <span class="op">=</span> get_sliced_fragment_dataframe(</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>    fragment_mz_df, </span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>    precursor_df.loc[:<span class="dv">0</span>,[<span class="st">'frag_start_idx'</span>,<span class="st">'frag_end_idx'</span>]].values,</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>    fragment_mz_df.columns.values</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> np.allclose(fragment_mz_df.iloc[</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>        precursor_df[<span class="st">'frag_start_idx'</span>].values[<span class="dv">0</span>]:precursor_df[<span class="st">'frag_end_idx'</span>].values[<span class="dv">0</span>],:</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>    ].values, sliced_frag_df.values</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>sliced_frag_df <span class="op">=</span> get_sliced_fragment_dataframe(</span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>    fragment_mz_df, </span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>    precursor_df.loc[:<span class="dv">0</span>,[<span class="st">'frag_start_idx'</span>,<span class="st">'frag_end_idx'</span>]].values,</span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> np.allclose(fragment_mz_df.iloc[</span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a>        precursor_df[<span class="st">'frag_start_idx'</span>].values[<span class="dv">0</span>]:precursor_df[<span class="st">'frag_end_idx'</span>].values[<span class="dv">0</span>],:</span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a>    ].values, sliced_frag_df.values</span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-19"><a href="#cb17-19" aria-hidden="true" tabindex="-1"></a>ith_pep <span class="op">=</span> <span class="dv">1</span></span>
<span id="cb17-20"><a href="#cb17-20" aria-hidden="true" tabindex="-1"></a>update_sliced_fragment_dataframe(</span>
<span id="cb17-21"><a href="#cb17-21" aria-hidden="true" tabindex="-1"></a>    fragment_mz_df, </span>
<span id="cb17-22"><a href="#cb17-22" aria-hidden="true" tabindex="-1"></a>    <span class="op">-</span>np.ones((precursor_df.nAA.values[ith_pep]<span class="op">-</span><span class="dv">1</span>,<span class="bu">len</span>(fragment_mz_df.columns))),</span>
<span id="cb17-23"><a href="#cb17-23" aria-hidden="true" tabindex="-1"></a>    [(precursor_df[<span class="st">'frag_start_idx'</span>].values[ith_pep],precursor_df[<span class="st">'frag_end_idx'</span>].values[ith_pep])]</span>
<span id="cb17-24"><a href="#cb17-24" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb17-25"><a href="#cb17-25" aria-hidden="true" tabindex="-1"></a>sliced_frag_df <span class="op">=</span> get_sliced_fragment_dataframe(</span>
<span id="cb17-26"><a href="#cb17-26" aria-hidden="true" tabindex="-1"></a>    fragment_mz_df, </span>
<span id="cb17-27"><a href="#cb17-27" aria-hidden="true" tabindex="-1"></a>    precursor_df.loc[ith_pep:ith_pep,[<span class="st">'frag_start_idx'</span>,<span class="st">'frag_end_idx'</span>]].values,</span>
<span id="cb17-28"><a href="#cb17-28" aria-hidden="true" tabindex="-1"></a>    fragment_mz_df.columns.values</span>
<span id="cb17-29"><a href="#cb17-29" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb17-30"><a href="#cb17-30" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> np.allclose(</span>
<span id="cb17-31"><a href="#cb17-31" aria-hidden="true" tabindex="-1"></a>    <span class="op">-</span>np.ones((precursor_df.nAA.values[ith_pep]<span class="op">-</span><span class="dv">1</span>,<span class="bu">len</span>(fragment_mz_df.columns))), </span>
<span id="cb17-32"><a href="#cb17-32" aria-hidden="true" tabindex="-1"></a>    sliced_frag_df.values</span>
<span id="cb17-33"><a href="#cb17-33" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb17-34"><a href="#cb17-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-35"><a href="#cb17-35" aria-hidden="true" tabindex="-1"></a>ith_pep <span class="op">=</span> <span class="dv">2</span></span>
<span id="cb17-36"><a href="#cb17-36" aria-hidden="true" tabindex="-1"></a>update_sliced_fragment_dataframe(</span>
<span id="cb17-37"><a href="#cb17-37" aria-hidden="true" tabindex="-1"></a>    fragment_mz_df, </span>
<span id="cb17-38"><a href="#cb17-38" aria-hidden="true" tabindex="-1"></a>    <span class="op">-</span><span class="dv">2</span><span class="op">*</span>np.ones((precursor_df.nAA.values[ith_pep]<span class="op">-</span><span class="dv">1</span>,<span class="bu">len</span>(fragment_mz_df.columns))),</span>
<span id="cb17-39"><a href="#cb17-39" aria-hidden="true" tabindex="-1"></a>    [(precursor_df[<span class="st">'frag_start_idx'</span>].values[ith_pep],precursor_df[<span class="st">'frag_end_idx'</span>].values[ith_pep])],</span>
<span id="cb17-40"><a href="#cb17-40" aria-hidden="true" tabindex="-1"></a>    charged_frag_types<span class="op">=</span>fragment_mz_df.columns.values</span>
<span id="cb17-41"><a href="#cb17-41" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb17-42"><a href="#cb17-42" aria-hidden="true" tabindex="-1"></a>sliced_frag_df <span class="op">=</span> get_sliced_fragment_dataframe(</span>
<span id="cb17-43"><a href="#cb17-43" aria-hidden="true" tabindex="-1"></a>    fragment_mz_df, </span>
<span id="cb17-44"><a href="#cb17-44" aria-hidden="true" tabindex="-1"></a>    precursor_df.loc[ith_pep:ith_pep,[<span class="st">'frag_start_idx'</span>,<span class="st">'frag_end_idx'</span>]].values,</span>
<span id="cb17-45"><a href="#cb17-45" aria-hidden="true" tabindex="-1"></a>    fragment_mz_df.columns.values</span>
<span id="cb17-46"><a href="#cb17-46" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb17-47"><a href="#cb17-47" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> np.allclose(</span>
<span id="cb17-48"><a href="#cb17-48" aria-hidden="true" tabindex="-1"></a>    <span class="op">-</span><span class="dv">2</span><span class="op">*</span>np.ones((precursor_df.nAA.values[ith_pep]<span class="op">-</span><span class="dv">1</span>,<span class="bu">len</span>(fragment_mz_df.columns))), </span>
<span id="cb17-49"><a href="#cb17-49" aria-hidden="true" tabindex="-1"></a>    sliced_frag_df.values</span>
<span id="cb17-50"><a href="#cb17-50" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Test <a href="https://MannLabs.github.io/alphabase/peptide/fragment.html#create_fragment_mz_dataframe"><code>create_fragment_mz_dataframe</code></a></p>
<p>If nAA column is not sorted, <a href="https://MannLabs.github.io/alphabase/peptide/fragment.html#create_fragment_mz_dataframe"><code>create_fragment_mz_dataframe</code></a> also works. But it would be much slower for large peptide sets.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>repeat <span class="op">=</span> <span class="dv">2</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>peptides <span class="op">=</span> [<span class="st">'AGHCEWQMKAADER'</span>]<span class="op">*</span>repeat</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>mods <span class="op">=</span> [<span class="st">'Acetyl@Protein N-term;Carbamidomethyl@C;Oxidation@M'</span>]<span class="op">*</span>repeat</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>sites <span class="op">=</span> [<span class="st">'0;4;8'</span>]<span class="op">*</span>repeat</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>peptides <span class="op">+=</span> [<span class="st">'AGHCEWQMK'</span>]<span class="op">*</span>repeat</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>mods <span class="op">+=</span> [<span class="st">''</span>]<span class="op">*</span>repeat</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>sites <span class="op">+=</span> [<span class="st">''</span>]<span class="op">*</span>repeat</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>precursor_df <span class="op">=</span> pd.DataFrame({</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">'sequence'</span>: peptides,</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>    <span class="st">'mods'</span>: mods,</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>    <span class="st">'mod_sites'</span>: sites</span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>precursor_df[<span class="st">'nAA'</span>] <span class="op">=</span> precursor_df[<span class="st">'sequence'</span>].<span class="bu">str</span>.<span class="bu">len</span>()</span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>precursor_df[<span class="st">'charge'</span>] <span class="op">=</span> <span class="dv">2</span></span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> <span class="kw">not</span> precursor_df.nAA.is_monotonic_increasing</span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a>fragment_mz_df <span class="op">=</span> create_fragment_mz_dataframe_by_sort_precursor(</span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a>    precursor_df,</span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a>    get_charged_frag_types([<span class="st">'b'</span>,<span class="st">'y'</span>,<span class="st">'b_modloss'</span>,<span class="st">'y_modloss'</span>],<span class="dv">2</span>)</span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a>precursor_df <span class="op">=</span> precursor_df.sort_values(<span class="st">'nAA'</span>, ascending<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb18-22"><a href="#cb18-22" aria-hidden="true" tabindex="-1"></a>fragment_mz_df1 <span class="op">=</span> create_fragment_mz_dataframe(</span>
<span id="cb18-23"><a href="#cb18-23" aria-hidden="true" tabindex="-1"></a>    precursor_df,</span>
<span id="cb18-24"><a href="#cb18-24" aria-hidden="true" tabindex="-1"></a>    get_charged_frag_types([<span class="st">'b'</span>,<span class="st">'y'</span>,<span class="st">'b_modloss'</span>,<span class="st">'y_modloss'</span>],<span class="dv">2</span>),</span>
<span id="cb18-25"><a href="#cb18-25" aria-hidden="true" tabindex="-1"></a>    reference_fragment_df<span class="op">=</span>fragment_mz_df</span>
<span id="cb18-26"><a href="#cb18-26" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb18-27"><a href="#cb18-27" aria-hidden="true" tabindex="-1"></a>ith_pep <span class="op">=</span> <span class="dv">2</span></span>
<span id="cb18-28"><a href="#cb18-28" aria-hidden="true" tabindex="-1"></a>frag_start, frag_end <span class="op">=</span> precursor_df[[<span class="st">'frag_start_idx'</span>,<span class="st">'frag_end_idx'</span>]].values[ith_pep]</span>
<span id="cb18-29"><a href="#cb18-29" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> np.allclose(fragment_mz_df.values, fragment_mz_df1.values)</span>
<span id="cb18-30"><a href="#cb18-30" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> np.allclose(fragment_mz_df.iloc[frag_start:frag_end][<span class="st">'b_z1'</span>].values, </span>
<span id="cb18-31"><a href="#cb18-31" aria-hidden="true" tabindex="-1"></a>        [ <span class="fl">72.04439025</span>,  <span class="fl">129.06585397</span>,  <span class="fl">266.12476583</span>,  <span class="fl">369.13395079</span>,</span>
<span id="cb18-32"><a href="#cb18-32" aria-hidden="true" tabindex="-1"></a>        <span class="fl">498.17654388</span>,  <span class="fl">684.25585683</span>,  <span class="fl">812.31443434</span>,  <span class="fl">943.35491942</span>]</span>
<span id="cb18-33"><a href="#cb18-33" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb18-34"><a href="#cb18-34" aria-hidden="true" tabindex="-1"></a>ith_pep <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb18-35"><a href="#cb18-35" aria-hidden="true" tabindex="-1"></a>frag_start, frag_end <span class="op">=</span> precursor_df[[<span class="st">'frag_start_idx'</span>,<span class="st">'frag_end_idx'</span>]].values[ith_pep]</span>
<span id="cb18-36"><a href="#cb18-36" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> np.allclose(fragment_mz_df.iloc[frag_start:frag_end][<span class="st">'b_z2'</span>].values, </span>
<span id="cb18-37"><a href="#cb18-37" aria-hidden="true" tabindex="-1"></a>        [ <span class="fl">57.5311157</span> ,  <span class="fl">86.04184756</span>, <span class="fl">154.57130349</span>, <span class="fl">234.58662783</span>,</span>
<span id="cb18-38"><a href="#cb18-38" aria-hidden="true" tabindex="-1"></a>            <span class="fl">299.10792438</span>, <span class="fl">392.14758085</span>, <span class="fl">456.1768696</span> , <span class="fl">529.69456946</span>,</span>
<span id="cb18-39"><a href="#cb18-39" aria-hidden="true" tabindex="-1"></a>            <span class="fl">593.74205097</span>, <span class="fl">629.26060786</span>, <span class="fl">664.77916475</span>, <span class="fl">722.29263626</span>,</span>
<span id="cb18-40"><a href="#cb18-40" aria-hidden="true" tabindex="-1"></a>            <span class="fl">786.81393281</span>]</span>
<span id="cb18-41"><a href="#cb18-41" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb18-42"><a href="#cb18-42" aria-hidden="true" tabindex="-1"></a>precursor_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>sequence</th>
      <th>mods</th>
      <th>mod_sites</th>
      <th>nAA</th>
      <th>charge</th>
      <th>frag_start_idx</th>
      <th>frag_end_idx</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2</th>
      <td>AGHCEWQMKAADER</td>
      <td>Acetyl@Protein N-term;Carbamidomethyl@C;Oxidat...</td>
      <td>0;4;8</td>
      <td>14</td>
      <td>2</td>
      <td>16</td>
      <td>29</td>
    </tr>
    <tr>
      <th>3</th>
      <td>AGHCEWQMKAADER</td>
      <td>Acetyl@Protein N-term;Carbamidomethyl@C;Oxidat...</td>
      <td>0;4;8</td>
      <td>14</td>
      <td>2</td>
      <td>29</td>
      <td>42</td>
    </tr>
    <tr>
      <th>0</th>
      <td>AGHCEWQMK</td>
      <td></td>
      <td></td>
      <td>9</td>
      <td>2</td>
      <td>0</td>
      <td>8</td>
    </tr>
    <tr>
      <th>1</th>
      <td>AGHCEWQMK</td>
      <td></td>
      <td></td>
      <td>9</td>
      <td>2</td>
      <td>8</td>
      <td>16</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>_reference_frag_df <span class="op">=</span> fragment_mz_df</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>fragment_mz_df <span class="op">=</span> create_fragment_mz_dataframe(</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>    precursor_df,</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>    [<span class="st">'b_z1'</span>,<span class="st">'y_z1'</span>],</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>    reference_fragment_df<span class="op">=</span>_reference_frag_df</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> np.allclose(fragment_mz_df.values, _reference_frag_df[fragment_mz_df.columns])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="test-mod-deltas" class="level2">
<h2 class="anchored" data-anchor-id="test-mod-deltas">Test mod deltas</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>repeat <span class="op">=</span> <span class="dv">1</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>peptides <span class="op">=</span> [<span class="st">'AGHCEWQMK'</span>]<span class="op">*</span>repeat</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>mods <span class="op">=</span> [<span class="st">'Acetyl@Protein N-term;Carbamidomethyl@C;Oxidation@M'</span>]<span class="op">*</span>repeat</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>sites <span class="op">=</span> [<span class="st">'0;4;8'</span>]<span class="op">*</span>repeat</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>peptides <span class="op">+=</span> [<span class="st">'AGHCEWQMK'</span>]<span class="op">*</span>repeat</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>mods <span class="op">+=</span> [<span class="st">'Acetyl@Protein N-term;Carbamidomethyl@C;Oxidation@M'</span>]<span class="op">*</span>repeat</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>sites <span class="op">+=</span> [<span class="st">'0;4;8'</span>]<span class="op">*</span>repeat</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>precursor_df <span class="op">=</span> pd.DataFrame({</span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">'sequence'</span>: peptides,</span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>    <span class="st">'mods'</span>: mods,</span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>    <span class="st">'mod_sites'</span>: sites</span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>precursor_df[<span class="st">'nAA'</span>] <span class="op">=</span> precursor_df[<span class="st">'sequence'</span>].<span class="bu">str</span>.<span class="bu">len</span>()</span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>precursor_df[<span class="st">'charge'</span>] <span class="op">=</span> <span class="dv">2</span></span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a>mod_deltas <span class="op">=</span> [<span class="st">''</span>]<span class="op">*</span><span class="bu">len</span>(precursor_df)</span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a>mod_delta_sites <span class="op">=</span> [<span class="st">''</span>]<span class="op">*</span><span class="bu">len</span>(precursor_df)</span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a>mod_deltas[<span class="dv">0</span>],mod_delta_sites[<span class="dv">0</span>] <span class="op">=</span> <span class="st">'100;200'</span>,<span class="st">'0;-1'</span></span>
<span id="cb20-19"><a href="#cb20-19" aria-hidden="true" tabindex="-1"></a>precursor_df[<span class="st">'mod_deltas'</span>] <span class="op">=</span> mod_deltas</span>
<span id="cb20-20"><a href="#cb20-20" aria-hidden="true" tabindex="-1"></a>precursor_df[<span class="st">'mod_delta_sites'</span>] <span class="op">=</span> mod_delta_sites</span>
<span id="cb20-21"><a href="#cb20-21" aria-hidden="true" tabindex="-1"></a>update_precursor_mz(precursor_df)</span>
<span id="cb20-22"><a href="#cb20-22" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> np.allclose(precursor_df.precursor_mz.values, [<span class="fl">752.747333</span>, <span class="fl">602.747333</span>])</span>
<span id="cb20-23"><a href="#cb20-23" aria-hidden="true" tabindex="-1"></a>fragment_mz_df <span class="op">=</span> create_fragment_mz_dataframe(precursor_df, charged_frag_types<span class="op">=</span>fragment_mz_df.columns.values)</span>
<span id="cb20-24"><a href="#cb20-24" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> np.allclose(fragment_mz_df[<span class="st">'y_z1'</span>].values[precursor_df.frag_start_idx[<span class="dv">0</span>]:precursor_df.frag_end_idx[<span class="dv">0</span>]], </span>
<span id="cb20-25"><a href="#cb20-25" aria-hidden="true" tabindex="-1"></a>    [<span class="fl">1291.43971168</span>, <span class="fl">1234.41824796</span>, <span class="fl">1097.3593361</span> ,  <span class="fl">937.32868742</span>,</span>
<span id="cb20-26"><a href="#cb20-26" aria-hidden="true" tabindex="-1"></a>        <span class="fl">808.28609433</span>,  <span class="fl">622.20678138</span>,  <span class="fl">494.14820387</span>,  <span class="fl">347.11280417</span>]</span>
<span id="cb20-27"><a href="#cb20-27" aria-hidden="true" tabindex="-1"></a>),  <span class="ss">f'200 Da must be added to all y-ions'</span></span>
<span id="cb20-28"><a href="#cb20-28" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> np.allclose(fragment_mz_df[<span class="st">'b_z1'</span>].values[precursor_df.frag_start_idx[<span class="dv">0</span>]:precursor_df.frag_end_idx[<span class="dv">0</span>]], </span>
<span id="cb20-29"><a href="#cb20-29" aria-hidden="true" tabindex="-1"></a>    [<span class="fl">214.05495494</span>,  <span class="fl">271.07641866</span>,  <span class="fl">408.13533052</span>,  <span class="fl">568.1659792</span> ,</span>
<span id="cb20-30"><a href="#cb20-30" aria-hidden="true" tabindex="-1"></a>        <span class="fl">697.20857228</span>,  <span class="fl">883.28788524</span>, <span class="fl">1011.34646274</span>, <span class="fl">1158.38186245</span>]</span>
<span id="cb20-31"><a href="#cb20-31" aria-hidden="true" tabindex="-1"></a>),  <span class="ss">f'100 Da must be added to all b-ions'</span></span>
<span id="cb20-32"><a href="#cb20-32" aria-hidden="true" tabindex="-1"></a>precursor_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>sequence</th>
      <th>mods</th>
      <th>mod_sites</th>
      <th>nAA</th>
      <th>charge</th>
      <th>mod_deltas</th>
      <th>mod_delta_sites</th>
      <th>precursor_mz</th>
      <th>frag_start_idx</th>
      <th>frag_end_idx</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>AGHCEWQMK</td>
      <td>Acetyl@Protein N-term;Carbamidomethyl@C;Oxidat...</td>
      <td>0;4;8</td>
      <td>9</td>
      <td>2</td>
      <td>100;200</td>
      <td>0;-1</td>
      <td>752.747333</td>
      <td>0</td>
      <td>8</td>
    </tr>
    <tr>
      <th>1</th>
      <td>AGHCEWQMK</td>
      <td>Acetyl@Protein N-term;Carbamidomethyl@C;Oxidat...</td>
      <td>0;4;8</td>
      <td>9</td>
      <td>2</td>
      <td></td>
      <td></td>
      <td>602.747333</td>
      <td>8</td>
      <td>16</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>