<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.1.251">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>alphabase - Modification information</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="../styles.css">
<meta property="og:title" content="alphabase - Modification information">
<meta property="og:description" content="The default modification TSV is stored in alphabase/constants/const_files/modification.tsv. Please check it to add more modifications.">
<meta property="og:site-name" content="alphabase">
<meta name="twitter:title" content="alphabase - Modification information">
<meta name="twitter:description" content="The default modification TSV is stored in alphabase/constants/const_files/modification.tsv. Please check it to add more modifications.">
<meta name="twitter:card" content="summary">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">alphabase</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/MannLabs/alphabase"><i class="bi bi-github" role="img">
</i> 
 </a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">Modification information</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">AlphaBase</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../utils.html" class="sidebar-item-text sidebar-link">Utils</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../yaml_utils.html" class="sidebar-item-text sidebar-link">YAML Utils</a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">constants</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../constants/constants.aa.html" class="sidebar-item-text sidebar-link">Amino acid information</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../constants/constants.element.html" class="sidebar-item-text sidebar-link">Atom element information</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../constants/constants.isotope.html" class="sidebar-item-text sidebar-link">Isotope distribution</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../constants/constants.modification.html" class="sidebar-item-text sidebar-link active">Modification information</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">io</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../io/hdf.html" class="sidebar-item-text sidebar-link">HDF functionalities</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">peptide</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../peptide/peptide.fragment.html" class="sidebar-item-text sidebar-link">Fragment Functionalities</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../peptide/peptide.mass_calc.html" class="sidebar-item-text sidebar-link">Mass Calculation</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../peptide/peptide.mobility.html" class="sidebar-item-text sidebar-link">CCS/Mobility Functionalities</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../peptide/peptide.precursor.html" class="sidebar-item-text sidebar-link">Precursor Functionalities</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">protein</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../protein/protein.fasta.html" class="sidebar-item-text sidebar-link">Protein and Peptide Processing</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../protein/test_fasta.html" class="sidebar-item-text sidebar-link">Testing fasta</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">psm_reader</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../psm_reader/psm_reader.alphapept_reader.html" class="sidebar-item-text sidebar-link">AlphaPept PSM Reader</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../psm_reader/psm_reader.dia_psm_reader.html" class="sidebar-item-text sidebar-link">DIA PSM reader</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../psm_reader/psm_reader.maxquant_reader.html" class="sidebar-item-text sidebar-link">MaxQuant PSM reader</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../psm_reader/psm_reader.msfragger_reader.html" class="sidebar-item-text sidebar-link">MSFragger Reader</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../psm_reader/psm_reader.pfind_reader.html" class="sidebar-item-text sidebar-link">pFind PSM Reader</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../psm_reader/psm_reader.psm_reader.html" class="sidebar-item-text sidebar-link">Base Class for PSM Readers</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true">scoring</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../scoring/scoring.fdr.html" class="sidebar-item-text sidebar-link">FDR functionalities</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../scoring/scoring.feature_extraction_base.html" class="sidebar-item-text sidebar-link">Base Class of Feature Extractors</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../scoring/scoring.ml_scoring_base.html" class="sidebar-item-text sidebar-link">Base Class of ML Scoring Methods</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="true">spectral_library</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../spectral_library/spectral_library.decoy_library.html" class="sidebar-item-text sidebar-link">Decoy Libraries</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../spectral_library/spectral_library.library_base.html" class="sidebar-item-text sidebar-link">Base Class for Spectral Libraries</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../spectral_library/spectral_library.translate.html" class="sidebar-item-text sidebar-link">Translate Spectral Libraries</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-8" aria-expanded="true">statistics</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-8" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-8" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../statistics/regression.html" class="sidebar-item-text sidebar-link">Regression</a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#add_modifications_for_lower_case_aa" id="toc-add_modifications_for_lower_case_aa" class="nav-link active" data-scroll-target="#add_modifications_for_lower_case_aa">add_modifications_for_lower_case_AA</a></li>
  <li><a href="#update_all_by_mod_df" id="toc-update_all_by_mod_df" class="nav-link" data-scroll-target="#update_all_by_mod_df">update_all_by_MOD_DF</a></li>
  <li><a href="#load_mod_df" id="toc-load_mod_df" class="nav-link" data-scroll-target="#load_mod_df">load_mod_df</a></li>
  <li><a href="#keep_modloss_by_importance" id="toc-keep_modloss_by_importance" class="nav-link" data-scroll-target="#keep_modloss_by_importance">keep_modloss_by_importance</a></li>
  <li><a href="#with-nist-elements-all-calculated-masses-are-the-same-as-unimod_mass-now." id="toc-with-nist-elements-all-calculated-masses-are-the-same-as-unimod_mass-now." class="nav-link" data-scroll-target="#with-nist-elements-all-calculated-masses-are-the-same-as-unimod_mass-now.">With NIST elements, all calculated masses are the same as unimod_mass now.</a></li>
  <li><a href="#mod-site-representation" id="toc-mod-site-representation" class="nav-link" data-scroll-target="#mod-site-representation">Mod site representation</a>
  <ul>
  <li><a href="#calc_modification_mass_sum" id="toc-calc_modification_mass_sum" class="nav-link" data-scroll-target="#calc_modification_mass_sum">calc_modification_mass_sum</a></li>
  <li><a href="#calc_mod_masses_for_same_len_seqs" id="toc-calc_mod_masses_for_same_len_seqs" class="nav-link" data-scroll-target="#calc_mod_masses_for_same_len_seqs">calc_mod_masses_for_same_len_seqs</a></li>
  <li><a href="#calc_modification_mass" id="toc-calc_modification_mass" class="nav-link" data-scroll-target="#calc_modification_mass">calc_modification_mass</a></li>
  <li><a href="#calc_modloss_mass" id="toc-calc_modloss_mass" class="nav-link" data-scroll-target="#calc_modloss_mass">calc_modloss_mass</a></li>
  <li><a href="#calc_modloss_mass_with_importance" id="toc-calc_modloss_mass_with_importance" class="nav-link" data-scroll-target="#calc_modloss_mass_with_importance">calc_modloss_mass_with_importance</a></li>
  <li><a href="#note-that-get_modloss_mass-is-a-little-bit-time-comsuming" id="toc-note-that-get_modloss_mass-is-a-little-bit-time-comsuming" class="nav-link" data-scroll-target="#note-that-get_modloss_mass-is-a-little-bit-time-comsuming">Note that get_modloss_mass is a little bit time comsuming</a></li>
  <li><a href="#we-can-update-modification-list-for-differet-requirements-for-example" id="toc-we-can-update-modification-list-for-differet-requirements-for-example" class="nav-link" data-scroll-target="#we-can-update-modification-list-for-differet-requirements-for-example">We can update modification list for differet requirements, for example:</a>
  <ul class="collapse">
  <li><a href="#add_new_modifications" id="toc-add_new_modifications" class="nav-link" data-scroll-target="#add_new_modifications">add_new_modifications</a></li>
  </ul></li>
  </ul></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/MannLabs/alphabase/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block">Modification information</h1>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header>

<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<p>The default modification TSV is stored in <code>alphabase/constants/const_files/modification.tsv</code>. Please check it to add more modifications.</p>
<p>First, we load <code>modification.tsv</code> into <code>MOD_DF</code>.</p>
<p>Then, we extract information of <code>MOD_CHEM</code> (dict), <code>MOD_MASS</code> (dict), <code>MOD_LOSS_MASS</code> (dict), <code>MOD_INFO_DICT</code> (dict) … from <code>MOD_DF</code>. This step is done in <a href="https://MannLabs.github.io/alphabase/constants/modification.html#update_all_by_mod_df"><code>update_all_by_MOD_DF</code></a>.</p>
<p>All these steps are done by <a href="https://MannLabs.github.io/alphabase/constants/modification.html#load_mod_df"><code>load_mod_df</code></a>.</p>
<hr>
<p><a href="https://github.com/MannLabs/alphabase/blob/main/alphabase/constants/modification.py#L59" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="add_modifications_for_lower_case_aa" class="level3">
<h3 class="anchored" data-anchor-id="add_modifications_for_lower_case_aa">add_modifications_for_lower_case_AA</h3>
<blockquote class="blockquote">
<pre><code> add_modifications_for_lower_case_AA ()</code></pre>
</blockquote>
<p>Add modifications for lower-case AAs for advanced usages</p>
<hr>
<p><a href="https://github.com/MannLabs/alphabase/blob/main/alphabase/constants/modification.py#L37" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="update_all_by_mod_df" class="level3">
<h3 class="anchored" data-anchor-id="update_all_by_mod_df">update_all_by_MOD_DF</h3>
<blockquote class="blockquote">
<pre><code> update_all_by_MOD_DF ()</code></pre>
</blockquote>
<p>As DataFrame is more conveneint in data operation, we can also process MOD_DF and then update all global modification variables from MOD_DF</p>
<hr>
<p><a href="https://github.com/MannLabs/alphabase/blob/main/alphabase/constants/modification.py#L88" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="load_mod_df" class="level3">
<h3 class="anchored" data-anchor-id="load_mod_df">load_mod_df</h3>
<blockquote class="blockquote">
<pre><code> load_mod_df (tsv:str='/Users/georgwallmann/Documents/git/alphabase/alphab
              ase/constants/const_files/modification.tsv',
              modloss_importance_level=1)</code></pre>
</blockquote>
<hr>
<p><a href="https://github.com/MannLabs/alphabase/blob/main/alphabase/constants/modification.py#L82" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="keep_modloss_by_importance" class="level3">
<h3 class="anchored" data-anchor-id="keep_modloss_by_importance">keep_modloss_by_importance</h3>
<blockquote class="blockquote">
<pre><code> keep_modloss_by_importance (modloss_importance_level:float=1.0)</code></pre>
</blockquote>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>MOD_DF</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>mod_name</th>
      <th>avge_mass</th>
      <th>classification</th>
      <th>composition</th>
      <th>modloss_composition</th>
      <th>mono_mass</th>
      <th>unimod_id</th>
      <th>unimod_mass</th>
      <th>unimod_modloss</th>
      <th>modloss_importance</th>
      <th>mass</th>
      <th>modloss_original</th>
      <th>modloss</th>
    </tr>
    <tr>
      <th>mod_name</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>GlyGly@K</th>
      <td>GlyGly@K</td>
      <td>114.042927</td>
      <td>Post-translational</td>
      <td>H(6)C(4)N(2)O(2)</td>
      <td>H(6)C(4)N(2)O(2)</td>
      <td>114.042927</td>
      <td>121</td>
      <td>114.042927</td>
      <td>114.042927</td>
      <td>1000000.0</td>
      <td>114.042927</td>
      <td>114.042927</td>
      <td>114.042927</td>
    </tr>
    <tr>
      <th>15N-oxobutanoic@C^Any N-term</th>
      <td>15N-oxobutanoic@C^Any N-term</td>
      <td>-18.023900</td>
      <td>Artefact</td>
      <td>H(-3)15N(-1)</td>
      <td></td>
      <td>-18.023584</td>
      <td>1419</td>
      <td>-18.023584</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>-18.023584</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>15N-oxobutanoic@S^Protein N-term</th>
      <td>15N-oxobutanoic@S^Protein N-term</td>
      <td>-18.023900</td>
      <td>Post-translational</td>
      <td>H(-3)15N(-1)</td>
      <td></td>
      <td>-18.023584</td>
      <td>1419</td>
      <td>-18.023584</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>-18.023584</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>15N-oxobutanoic@T^Protein N-term</th>
      <td>15N-oxobutanoic@T^Protein N-term</td>
      <td>-18.023900</td>
      <td>Post-translational</td>
      <td>H(-3)15N(-1)</td>
      <td></td>
      <td>-18.023584</td>
      <td>1419</td>
      <td>-18.023584</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>-18.023584</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>2-dimethylsuccinyl@C</th>
      <td>2-dimethylsuccinyl@C</td>
      <td>144.125300</td>
      <td>Chemical derivative</td>
      <td>H(8)C(6)O(4)</td>
      <td></td>
      <td>144.042259</td>
      <td>1262</td>
      <td>144.042259</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>144.042259</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>spermidine@Q</th>
      <td>spermidine@Q</td>
      <td>128.215300</td>
      <td>Chemical derivative</td>
      <td>H(16)C(7)N(2)</td>
      <td></td>
      <td>128.131349</td>
      <td>1421</td>
      <td>128.131349</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>128.131349</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>spermine@Q</th>
      <td>spermine@Q</td>
      <td>185.309700</td>
      <td>Chemical derivative</td>
      <td>H(23)C(10)N(3)</td>
      <td></td>
      <td>185.189198</td>
      <td>1420</td>
      <td>185.189198</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>185.189198</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>sulfo+amino@Y</th>
      <td>sulfo+amino@Y</td>
      <td>95.077800</td>
      <td>Chemical derivative</td>
      <td>H(1)N(1)O(3)S(1)</td>
      <td></td>
      <td>94.967714</td>
      <td>997</td>
      <td>94.967714</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>94.967714</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>thioacylPA@K</th>
      <td>thioacylPA@K</td>
      <td>159.206200</td>
      <td>Chemical derivative</td>
      <td>H(9)C(6)N(1)O(2)S(1)</td>
      <td></td>
      <td>159.035399</td>
      <td>967</td>
      <td>159.035399</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>159.035400</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>trifluoro@L</th>
      <td>trifluoro@L</td>
      <td>53.971400</td>
      <td>Non-standard residue</td>
      <td>H(-3)F(3)</td>
      <td></td>
      <td>53.971735</td>
      <td>750</td>
      <td>53.971735</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>53.971734</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
  </tbody>
</table>
<p>2685 rows × 13 columns</p>
</div>
</div>
</div>
</section>
<section id="with-nist-elements-all-calculated-masses-are-the-same-as-unimod_mass-now." class="level3">
<h3 class="anchored" data-anchor-id="with-nist-elements-all-calculated-masses-are-the-same-as-unimod_mass-now.">With NIST elements, all calculated masses are the same as unimod_mass now.</h3>
</section>
<section id="mod-site-representation" class="level1">
<h1>Mod site representation</h1>
<ul>
<li><code>site=0</code> refers to an N-term modification</li>
<li><code>site=-1</code> refers to a C-term modification</li>
<li><code>1&lt;=site&lt;=peplen</code> refers to a normal modification</li>
</ul>
<p>For example: _0A1B2C3D4E5F6G7H8I9J10K11_-1</p>
<p><a href="https://MannLabs.github.io/alphabase/constants/modification.html#calc_modification_mass"><code>calc_modification_mass</code></a> and <a href="https://MannLabs.github.io/alphabase/constants/modification.html#calc_modification_mass_sum"><code>calc_modification_mass_sum</code></a> are base functions to calculate masses. But for large sets of peptides, we recommend to use <a href="https://MannLabs.github.io/alphabase/constants/modification.html#calc_mod_masses_for_same_len_seqs"><code>calc_mod_masses_for_same_len_seqs</code></a> instead of <a href="https://MannLabs.github.io/alphabase/constants/modification.html#calc_modification_mass"><code>calc_modification_mass</code></a>.</p>
<hr>
<p><a href="https://github.com/MannLabs/alphabase/blob/main/alphabase/constants/modification.py#L190" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="calc_modification_mass_sum" class="level3">
<h3 class="anchored" data-anchor-id="calc_modification_mass_sum">calc_modification_mass_sum</h3>
<blockquote class="blockquote">
<pre><code> calc_modification_mass_sum (mod_names:List[str])</code></pre>
</blockquote>
<p>Calculate summed mass of the given modification without knowing the sites and peptide length. It is useful to calculate peptide mass.</p>
<table class="table">
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>mod_names</td>
<td>typing.List[str]</td>
<td>Modification name list</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>float</strong></td>
<td><strong>Total mass</strong></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/MannLabs/alphabase/blob/main/alphabase/constants/modification.py#L147" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="calc_mod_masses_for_same_len_seqs" class="level3">
<h3 class="anchored" data-anchor-id="calc_mod_masses_for_same_len_seqs">calc_mod_masses_for_same_len_seqs</h3>
<blockquote class="blockquote">
<pre><code> calc_mod_masses_for_same_len_seqs (nAA:int,
                                    mod_names_list:List[List[str]],
                                    mod_sites_list:List[List[int]])</code></pre>
</blockquote>
<p>Calculate modification masses for the given peptides with same peptide length (<code>nAA</code>).</p>
<table class="table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>nAA</td>
<td>int</td>
<td>Peptide length</td>
</tr>
<tr class="even">
<td>mod_names_list</td>
<td>typing.List[typing.List[str]]</td>
<td>List (pep_count) of modification list (n_mod on each peptide)</td>
</tr>
<tr class="odd">
<td>mod_sites_list</td>
<td>typing.List[typing.List[int]]</td>
<td>List of modification site list corresponding to <code>mod_names_list</code>.<br>* <code>site=0</code> refers to an N-term modification<br>* <code>site=-1</code> refers to a C-term modification<br>* <code>1&lt;=site&lt;=peplen</code> refers to a normal modification</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>ndarray</strong></td>
<td><strong>2-D array with shape=<code>(nAA, pep_count or len(mod_names_list)))</code>. <br>Masses of modifications through all the peptides, <br><code>0</code> if sites without modifications.</strong></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/MannLabs/alphabase/blob/main/alphabase/constants/modification.py#L107" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="calc_modification_mass" class="level3">
<h3 class="anchored" data-anchor-id="calc_modification_mass">calc_modification_mass</h3>
<blockquote class="blockquote">
<pre><code> calc_modification_mass (nAA:int, mod_names:List[str],
                         mod_sites:List[int])</code></pre>
</blockquote>
<p>Calculate modification masses for the given peptide length (<code>nAA</code>), and modified site list.</p>
<table class="table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>nAA</td>
<td>int</td>
<td>Peptide length</td>
</tr>
<tr class="even">
<td>mod_names</td>
<td>typing.List[str]</td>
<td>List[str]. Modification name list</td>
</tr>
<tr class="odd">
<td>mod_sites</td>
<td>typing.List[int]</td>
<td>List[int]. Modification site list corresponding to <code>mod_names</code>.<br>* <code>site=0</code> refers to an N-term modification<br>* <code>site=-1</code> refers to a C-term modification<br>* <code>1&lt;=site&lt;=peplen</code> refers to a normal modification</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>ndarray</strong></td>
<td><strong>1-D array with length=<code>nAA</code>. <br>Masses of modifications through the peptide, <br><code>0</code> if sites has no modifications</strong></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/MannLabs/alphabase/blob/main/alphabase/constants/modification.py#L325" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="calc_modloss_mass" class="level3">
<h3 class="anchored" data-anchor-id="calc_modloss_mass">calc_modloss_mass</h3>
<blockquote class="blockquote">
<pre><code> calc_modloss_mass (nAA:int, mod_names:List, mod_sites:List,
                    for_nterm_frag:bool)</code></pre>
</blockquote>
<p>Calculate modification loss masses (e.g.&nbsp;-98 Da for Phospho@S/T, -64 Da for Oxidation@M). The mod loss mass is calculated by the modification closer to the fragment sites. For example, the modloss of the b3 ion for <code>AS(Phospho@S)M(Oxidation@M)...</code> will be -64 Da.</p>
<table class="table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>nAA</td>
<td>int</td>
<td>Peptide length</td>
</tr>
<tr class="even">
<td>mod_names</td>
<td>typing.List</td>
<td>Modification name list</td>
</tr>
<tr class="odd">
<td>mod_sites</td>
<td>typing.List</td>
<td>Modification site list corresponding</td>
</tr>
<tr class="even">
<td>for_nterm_frag</td>
<td>bool</td>
<td>If <code>True</code>, the loss will be on the <br>N-term fragments (mainly <code>b</code> ions); <br>If <code>False</code>, the loss will be on the <br>C-term fragments (mainly <code>y</code> ions)</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>ndarray</strong></td>
<td><strong>mod_loss masses</strong></td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/MannLabs/alphabase/blob/main/alphabase/constants/modification.py#L249" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="calc_modloss_mass_with_importance" class="level3">
<h3 class="anchored" data-anchor-id="calc_modloss_mass_with_importance">calc_modloss_mass_with_importance</h3>
<blockquote class="blockquote">
<pre><code> calc_modloss_mass_with_importance (nAA:int, mod_names:List,
                                    mod_sites:List, for_nterm_frag:bool)</code></pre>
</blockquote>
<p>Calculate modification loss masses (e.g.&nbsp;-98 Da for Phospho@S/T, -64 Da for Oxidation@M). Modifications with higher <code>MOD_LOSS_IMPORTANCE</code> have higher priorities. For example, <code>AS(Phospho@S)M(Oxidation@M)...</code>, importance of Phospho@S &gt; importance of Oxidation@M, so the modloss of b3 ion will be -98 Da, not -64 Da.</p>
<table class="table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>nAA</td>
<td>int</td>
<td>Peptide length</td>
</tr>
<tr class="even">
<td>mod_names</td>
<td>typing.List</td>
<td>Modification name list</td>
</tr>
<tr class="odd">
<td>mod_sites</td>
<td>typing.List</td>
<td>Modification site list</td>
</tr>
<tr class="even">
<td>for_nterm_frag</td>
<td>bool</td>
<td>If <code>True</code>, the loss will be on the <br>N-term fragments (mainly <code>b</code> ions); <br>If <code>False</code>, the loss will be on the <br>C-term fragments (mainly <code>y</code> ions)</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>ndarray</strong></td>
<td><strong>mod_loss masses</strong></td>
</tr>
</tbody>
</table>
</section>
<section id="note-that-get_modloss_mass-is-a-little-bit-time-comsuming" class="level3">
<h3 class="anchored" data-anchor-id="note-that-get_modloss_mass-is-a-little-bit-time-comsuming">Note that get_modloss_mass is a little bit time comsuming</h3>
<p><code>%timeit get_modloss_mass(10, mod_names, mod_sites, False)</code></p>
<p><code>Results (12 seconds in total): 12.6 µs ± 96.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>MOD_DF.drop_duplicates(<span class="st">'classification'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>mod_name</th>
      <th>avge_mass</th>
      <th>classification</th>
      <th>composition</th>
      <th>modloss_composition</th>
      <th>mono_mass</th>
      <th>unimod_id</th>
      <th>unimod_mass</th>
      <th>unimod_modloss</th>
      <th>modloss_importance</th>
      <th>mass</th>
      <th>modloss_original</th>
      <th>modloss</th>
    </tr>
    <tr>
      <th>mod_name</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>GlyGly@K</th>
      <td>GlyGly@K</td>
      <td>114.042927</td>
      <td>Post-translational</td>
      <td>H(6)C(4)N(2)O(2)</td>
      <td>H(6)C(4)N(2)O(2)</td>
      <td>114.042927</td>
      <td>121</td>
      <td>114.042927</td>
      <td>114.042927</td>
      <td>1000000.0</td>
      <td>114.042927</td>
      <td>114.042927</td>
      <td>114.042927</td>
    </tr>
    <tr>
      <th>15N-oxobutanoic@C^Any N-term</th>
      <td>15N-oxobutanoic@C^Any N-term</td>
      <td>-18.023900</td>
      <td>Artefact</td>
      <td>H(-3)15N(-1)</td>
      <td></td>
      <td>-18.023584</td>
      <td>1419</td>
      <td>-18.023584</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>-18.023584</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>2-dimethylsuccinyl@C</th>
      <td>2-dimethylsuccinyl@C</td>
      <td>144.125300</td>
      <td>Chemical derivative</td>
      <td>H(8)C(6)O(4)</td>
      <td></td>
      <td>144.042259</td>
      <td>1262</td>
      <td>144.042259</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>144.042259</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>3-deoxyglucosone@R</th>
      <td>3-deoxyglucosone@R</td>
      <td>144.125300</td>
      <td>Multiple</td>
      <td>H(8)C(6)O(4)</td>
      <td></td>
      <td>144.042259</td>
      <td>949</td>
      <td>144.042259</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>144.042259</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>ADP-Ribosyl@C</th>
      <td>ADP-Ribosyl@C</td>
      <td>541.300500</td>
      <td>Other glycosylation</td>
      <td>H(21)C(15)N(5)O(13)P(2)</td>
      <td></td>
      <td>541.061110</td>
      <td>213</td>
      <td>541.061110</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>541.061110</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>ADP-Ribosyl@N</th>
      <td>ADP-Ribosyl@N</td>
      <td>541.300500</td>
      <td>N-linked glycosylation</td>
      <td>H(21)C(15)N(5)O(13)P(2)</td>
      <td>H(21)C(15)N(5)O(13)P(2)</td>
      <td>541.061110</td>
      <td>213</td>
      <td>541.061110</td>
      <td>541.061110</td>
      <td>0.0</td>
      <td>541.061110</td>
      <td>541.061110</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>ADP-Ribosyl@S</th>
      <td>ADP-Ribosyl@S</td>
      <td>541.300500</td>
      <td>O-linked glycosylation</td>
      <td>H(21)C(15)N(5)O(13)P(2)</td>
      <td>H(21)C(15)N(5)O(13)P(2)</td>
      <td>541.061110</td>
      <td>213</td>
      <td>541.061110</td>
      <td>541.061110</td>
      <td>0.0</td>
      <td>541.061110</td>
      <td>541.061110</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>AEC-MAEC:2H(4)@S</th>
      <td>AEC-MAEC:2H(4)@S</td>
      <td>63.158000</td>
      <td>Isotopic label</td>
      <td>H(1)2H(4)C(2)N(1)O(-1)S(1)</td>
      <td></td>
      <td>63.044462</td>
      <td>792</td>
      <td>63.044462</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>63.044463</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>Ahx2+Hsl@Any C-term</th>
      <td>Ahx2+Hsl@Any C-term</td>
      <td>309.403900</td>
      <td>Non-standard residue</td>
      <td>H(27)C(16)N(3)O(3)</td>
      <td></td>
      <td>309.205242</td>
      <td>1015</td>
      <td>309.205242</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>309.205242</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>Ala-&gt;Arg@A</th>
      <td>Ala-&gt;Arg@A</td>
      <td>85.107800</td>
      <td>AA substitution</td>
      <td>H(7)C(3)N(3)O(0)S(0)</td>
      <td></td>
      <td>85.063997</td>
      <td>1052</td>
      <td>85.063997</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>85.063997</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>Arg-loss@R^Any C-term</th>
      <td>Arg-loss@R^Any C-term</td>
      <td>-156.185700</td>
      <td>Other</td>
      <td>H(-12)C(-6)N(-4)O(-1)</td>
      <td></td>
      <td>-156.101111</td>
      <td>1287</td>
      <td>-156.101111</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>-156.101111</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>FormylMet@Protein N-term</th>
      <td>FormylMet@Protein N-term</td>
      <td>159.206200</td>
      <td>Pre-translational</td>
      <td>H(9)C(6)N(1)O(2)S(1)</td>
      <td></td>
      <td>159.035399</td>
      <td>107</td>
      <td>159.035399</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>159.035400</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>HexN@K</th>
      <td>HexN@K</td>
      <td>161.155800</td>
      <td>Synth. pep. protect. gp.</td>
      <td>H(11)C(6)N(1)O(4)</td>
      <td></td>
      <td>161.068808</td>
      <td>454</td>
      <td>161.068808</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>161.068808</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>Met-loss+Acetyl@M^Protein N-term</th>
      <td>Met-loss+Acetyl@M^Protein N-term</td>
      <td>-89.159400</td>
      <td>Co-translational</td>
      <td>H(-7)C(-3)N(-1)S(-1)</td>
      <td></td>
      <td>-89.029920</td>
      <td>766</td>
      <td>-89.029920</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>-89.029920</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
</section>
<section id="we-can-update-modification-list-for-differet-requirements-for-example" class="level2">
<h2 class="anchored" data-anchor-id="we-can-update-modification-list-for-differet-requirements-for-example">We can update modification list for differet requirements, for example:</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>add_modifications_for_lower_case_AA()</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>MOD_DF <span class="op">=</span> MOD_DF[</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>    (MOD_DF[<span class="st">'classification'</span>].isin([<span class="st">'Post-translational'</span>,<span class="st">'O-linked glycosylation'</span>,<span class="st">'AA substitution'</span>,<span class="st">'Multiple'</span>,<span class="st">'Non-standard residue'</span>,<span class="st">'Pre-translational'</span>]))</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>    <span class="op">&amp;</span> MOD_DF[<span class="st">'lower_case_AA'</span>]</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>] <span class="co"># we only need PTMs</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>update_all_by_MOD_DF()</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="co"># MOD_INFO_DICT is also updated</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>pd.DataFrame().from_dict(MOD_INFO_DICT, orient<span class="op">=</span><span class="st">'index'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>mod_name</th>
      <th>avge_mass</th>
      <th>classification</th>
      <th>composition</th>
      <th>modloss_composition</th>
      <th>mono_mass</th>
      <th>unimod_id</th>
      <th>unimod_mass</th>
      <th>unimod_modloss</th>
      <th>modloss_importance</th>
      <th>mass</th>
      <th>modloss_original</th>
      <th>modloss</th>
      <th>lower_case_AA</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>GlyGly@k</th>
      <td>GlyGly@k</td>
      <td>114.042927</td>
      <td>Post-translational</td>
      <td>H(6)C(4)N(2)O(2)</td>
      <td>H(6)C(4)N(2)O(2)</td>
      <td>114.042927</td>
      <td>121</td>
      <td>114.042927</td>
      <td>114.042927</td>
      <td>1000000.0</td>
      <td>114.042927</td>
      <td>114.042927</td>
      <td>114.042927</td>
      <td>True</td>
    </tr>
    <tr>
      <th>15N-oxobutanoic@s^Protein N-term</th>
      <td>15N-oxobutanoic@s^Protein N-term</td>
      <td>-18.023900</td>
      <td>Post-translational</td>
      <td>H(-3)15N(-1)</td>
      <td></td>
      <td>-18.023584</td>
      <td>1419</td>
      <td>-18.023584</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>-18.023584</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>True</td>
    </tr>
    <tr>
      <th>15N-oxobutanoic@t^Protein N-term</th>
      <td>15N-oxobutanoic@t^Protein N-term</td>
      <td>-18.023900</td>
      <td>Post-translational</td>
      <td>H(-3)15N(-1)</td>
      <td></td>
      <td>-18.023584</td>
      <td>1419</td>
      <td>-18.023584</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>-18.023584</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>True</td>
    </tr>
    <tr>
      <th>3-deoxyglucosone@r</th>
      <td>3-deoxyglucosone@r</td>
      <td>144.125300</td>
      <td>Multiple</td>
      <td>H(8)C(6)O(4)</td>
      <td></td>
      <td>144.042259</td>
      <td>949</td>
      <td>144.042259</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>144.042259</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>True</td>
    </tr>
    <tr>
      <th>3-phosphoglyceryl@k</th>
      <td>3-phosphoglyceryl@k</td>
      <td>168.042000</td>
      <td>Post-translational</td>
      <td>H(5)C(3)O(6)P(1)</td>
      <td></td>
      <td>167.982375</td>
      <td>1387</td>
      <td>167.982375</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>167.982375</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>True</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>pyrophospho@t</th>
      <td>pyrophospho@t</td>
      <td>159.959800</td>
      <td>Post-translational</td>
      <td>H(2)O(6)P(2)</td>
      <td>H(3)O(7)P(2)</td>
      <td>159.932662</td>
      <td>898</td>
      <td>159.932662</td>
      <td>176.935402</td>
      <td>0.0</td>
      <td>159.932662</td>
      <td>176.935401</td>
      <td>0.000000</td>
      <td>True</td>
    </tr>
    <tr>
      <th>s-GlcNAc@s</th>
      <td>s-GlcNAc@s</td>
      <td>283.255700</td>
      <td>O-linked glycosylation</td>
      <td>H(13)C(8)N(1)O(8)S(1)</td>
      <td>H(13)C(8)N(1)O(8)S(1)</td>
      <td>283.036187</td>
      <td>1412</td>
      <td>283.036187</td>
      <td>283.036187</td>
      <td>0.0</td>
      <td>283.036188</td>
      <td>283.036188</td>
      <td>0.000000</td>
      <td>True</td>
    </tr>
    <tr>
      <th>s-GlcNAc@t</th>
      <td>s-GlcNAc@t</td>
      <td>283.255700</td>
      <td>O-linked glycosylation</td>
      <td>H(13)C(8)N(1)O(8)S(1)</td>
      <td>H(13)C(8)N(1)O(8)S(1)</td>
      <td>283.036187</td>
      <td>1412</td>
      <td>283.036187</td>
      <td>283.036187</td>
      <td>0.0</td>
      <td>283.036188</td>
      <td>283.036188</td>
      <td>0.000000</td>
      <td>True</td>
    </tr>
    <tr>
      <th>serotonylation@q</th>
      <td>serotonylation@q</td>
      <td>159.184600</td>
      <td>Post-translational</td>
      <td>H(9)C(10)N(1)O(1)</td>
      <td></td>
      <td>159.068414</td>
      <td>1992</td>
      <td>159.068414</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>159.068414</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>True</td>
    </tr>
    <tr>
      <th>trifluoro@l</th>
      <td>trifluoro@l</td>
      <td>53.971400</td>
      <td>Non-standard residue</td>
      <td>H(-3)F(3)</td>
      <td></td>
      <td>53.971735</td>
      <td>750</td>
      <td>53.971735</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>53.971734</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>True</td>
    </tr>
  </tbody>
</table>
<p>1249 rows × 14 columns</p>
</div>
</div>
</div>
<hr>
<p><a href="https://github.com/MannLabs/alphabase/blob/main/alphabase/constants/modification.py#L370" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="add_new_modifications" class="level3">
<h3 class="anchored" data-anchor-id="add_new_modifications">add_new_modifications</h3>
<blockquote class="blockquote">
<pre><code> add_new_modifications (new_mods:list)</code></pre>
</blockquote>
<p>Add new modifications into MOD_DF</p>
<table class="table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>new_mods</td>
<td>list</td>
<td>list of tuples. Tuple example:<br>(<br> modname@site:str (e.g.&nbsp;Mod@S), <br> chemical compositions:str (e.g.&nbsp;“H(4)O(2)”),<br> [optional] modloss compositions:str (e.g.&nbsp;“H(2)O(1)”),<br>)</td>
</tr>
</tbody>
</table>


</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>