{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## The constants module (alphabase.constants)\n",
    "\n",
    "The `alphabase.constants` module contains chemical element (`alphabase.constants.element.py`), amino acid (`alphabase.constants.aa.py`) and modification (`alphabase.constants.modification.py`) information, which are basic information for all alphabase functionalities. Check `alphabase/constants/nist_element.yaml`, `alphabase/constants/amino_acid.yaml`, `alphabase/constants/used_mod.yaml` for details.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## The peptide module (alphabase.peptide)\n",
    "\n",
    "The `peptide` module contains peptide-related functionalities.\n",
    "\n",
    "### alphabase.peptide.precursor\n",
    "\n",
    "There are two functions in `alphabase.peptide.precursor`: \n",
    "\n",
    "- `refine_precursor_df(df)`. Sort df by 'nAA' and `reset_index` inplace. For faster precursor and fragment mass calculation.\n",
    "- `update_precursor_mz(df)`/`calc_precursor_mz(df)`. The dataframe must contain `sequence, mods, mod_sites, charge` columns.\n",
    "- `hash_precursor_df(df, seed=0)`. It generates int64-based unique identity for the modified peptide (`sequence`, `mods`, and `mod_sites`) in `mod_seq_hash` column. It will also generates int64-based unique identity for the precursor (`sequence`, `mods`, `mod_sites`, and `charge`) in the `mod_seq_charge_hash` column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>mods</th>\n",
       "      <th>mod_sites</th>\n",
       "      <th>charge</th>\n",
       "      <th>nAA</th>\n",
       "      <th>precursor_mz</th>\n",
       "      <th>mod_seq_hash</th>\n",
       "      <th>mod_seq_charge_hash</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACDEFHMK</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>980.396450</td>\n",
       "      <td>-3613625140520963925</td>\n",
       "      <td>-3613625140520963924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACDEFHMK</td>\n",
       "      <td>Acetyl@Protein N-term</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>1022.407015</td>\n",
       "      <td>-4469340833947450</td>\n",
       "      <td>-4469340833947449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACDEFHMK</td>\n",
       "      <td>Carbamidomethyl@C;Oxidation@M</td>\n",
       "      <td>2;7</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>527.210052</td>\n",
       "      <td>4151334950393863927</td>\n",
       "      <td>4151334950393863929</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sequence                           mods mod_sites  charge  nAA  \\\n",
       "0  ACDEFHMK                                                1    8   \n",
       "1  ACDEFHMK          Acetyl@Protein N-term         0       1    8   \n",
       "2  ACDEFHMK  Carbamidomethyl@C;Oxidation@M       2;7       2    8   \n",
       "\n",
       "   precursor_mz         mod_seq_hash  mod_seq_charge_hash  \n",
       "0    980.396450 -3613625140520963925 -3613625140520963924  \n",
       "1   1022.407015    -4469340833947450    -4469340833947449  \n",
       "2    527.210052  4151334950393863927  4151334950393863929  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import alphabase.peptide.precursor as precursor_utils\n",
    "import pandas as pd\n",
    "\n",
    "df = pd.DataFrame()\n",
    "df['sequence'] = ['ACDEFHMK']*3\n",
    "df['mods'] = ['',\n",
    "    'Acetyl@Protein N-term', \n",
    "    'Carbamidomethyl@C;Oxidation@M'\n",
    "]\n",
    "df['mod_sites'] = ['','0','2;7']\n",
    "df['charge'] = [1,1,2]\n",
    "\n",
    "precursor_utils.update_precursor_mz(df)\n",
    "precursor_utils.hash_precursor_df(df, seed=0)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There should be no duplicated hash codes for different modified peptides and precursors in the df:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert ~df.mod_seq_charge_hash.duplicated().any()\n",
    "assert ~df.mod_seq_hash.duplicated().any()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If hash collision or duplicates are detected, we can change the seed to generate other hash codes. We should make sure that all mod_seq_hash and mod_seq_charge_hash are unique for modified peptides () and precursors."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>mods</th>\n",
       "      <th>mod_sites</th>\n",
       "      <th>charge</th>\n",
       "      <th>nAA</th>\n",
       "      <th>precursor_mz</th>\n",
       "      <th>mod_seq_hash</th>\n",
       "      <th>mod_seq_charge_hash</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACDEFHMK</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>980.396450</td>\n",
       "      <td>4883417763805589795</td>\n",
       "      <td>4883417763805589796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACDEFHMK</td>\n",
       "      <td>Acetyl@Protein N-term</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>1022.407015</td>\n",
       "      <td>287597180480789294</td>\n",
       "      <td>287597180480789295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACDEFHMK</td>\n",
       "      <td>Carbamidomethyl@C;Oxidation@M</td>\n",
       "      <td>2;7</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>527.210052</td>\n",
       "      <td>4243306788669639832</td>\n",
       "      <td>4243306788669639834</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sequence                           mods mod_sites  charge  nAA  \\\n",
       "0  ACDEFHMK                                                1    8   \n",
       "1  ACDEFHMK          Acetyl@Protein N-term         0       1    8   \n",
       "2  ACDEFHMK  Carbamidomethyl@C;Oxidation@M       2;7       2    8   \n",
       "\n",
       "   precursor_mz         mod_seq_hash  mod_seq_charge_hash  \n",
       "0    980.396450  4883417763805589795  4883417763805589796  \n",
       "1   1022.407015   287597180480789294   287597180480789295  \n",
       "2    527.210052  4243306788669639832  4243306788669639834  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "precursor_utils.hash_precursor_df(df, seed=1337)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### alphabase.peptide.mobility\n",
    "\n",
    "The `mobility` module provides functionalities converting CCS values into mobility values and vice versa."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>mods</th>\n",
       "      <th>mod_sites</th>\n",
       "      <th>charge</th>\n",
       "      <th>ccs_pred</th>\n",
       "      <th>mobility_whatever</th>\n",
       "      <th>nAA</th>\n",
       "      <th>precursor_mz</th>\n",
       "      <th>mobility_pred</th>\n",
       "      <th>ccs_whatever</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACDEFHMK</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>300</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8</td>\n",
       "      <td>980.396450</td>\n",
       "      <td>1.477183</td>\n",
       "      <td>203.089245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACDEFHMK</td>\n",
       "      <td>Acetyl@Protein N-term</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>300</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8</td>\n",
       "      <td>1022.407015</td>\n",
       "      <td>1.478027</td>\n",
       "      <td>202.973356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACDEFHMK</td>\n",
       "      <td>Carbamidomethyl@C;Oxidation@M</td>\n",
       "      <td>2;7</td>\n",
       "      <td>2</td>\n",
       "      <td>300</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8</td>\n",
       "      <td>527.210052</td>\n",
       "      <td>0.739312</td>\n",
       "      <td>405.782410</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sequence                           mods mod_sites  charge  ccs_pred  \\\n",
       "0  ACDEFHMK                                                1       300   \n",
       "1  ACDEFHMK          Acetyl@Protein N-term         0       1       300   \n",
       "2  ACDEFHMK  Carbamidomethyl@C;Oxidation@M       2;7       2       300   \n",
       "\n",
       "   mobility_whatever  nAA  precursor_mz  mobility_pred  ccs_whatever  \n",
       "0                1.0    8    980.396450       1.477183    203.089245  \n",
       "1                1.0    8   1022.407015       1.478027    202.973356  \n",
       "2                1.0    8    527.210052       0.739312    405.782410  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import alphabase.peptide.mobility as mobility\n",
    "df = pd.DataFrame()\n",
    "df['sequence'] = ['ACDEFHMK']*3\n",
    "df['mods'] = ['',\n",
    "    'Acetyl@Protein N-term', \n",
    "    'Carbamidomethyl@C;Oxidation@M'\n",
    "]\n",
    "df['mod_sites'] = ['','0','2;7']\n",
    "df['charge'] = [1,1,2]\n",
    "df['ccs_pred'] = 300 # if we predict ccs values\n",
    "df['mobility_whatever'] = 1.0\n",
    "\n",
    "df['mobility_pred'] = mobility.ccs_to_mobility_for_df(\n",
    "    df, 'ccs_pred'\n",
    ")\n",
    "df['ccs_whatever'] = mobility.mobility_to_ccs_for_df(\n",
    "    df, 'mobility_whatever'\n",
    ")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### alphabase.peptide.fragment\n",
    "\n",
    "This is the most comprehensive module in alphabase. The most important function is `create_fragment_mz_dataframe`, which creates fragment dataframe for the given precursor_df and insert `frag_start_idx` and `frag_end_idx` in the precursor_df."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>b_z1</th>\n",
       "      <th>b_z2</th>\n",
       "      <th>y_z1</th>\n",
       "      <th>y_z2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>72.044390</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>909.359336</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>175.053575</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>806.350151</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>290.080518</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>691.323208</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>419.123111</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>562.280615</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>566.191525</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>415.212201</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>703.250437</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>278.153289</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>834.290922</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>147.112804</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>114.054955</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>909.359336</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>217.064140</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>806.350151</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>332.091083</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>691.323208</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>461.133676</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>562.280615</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>608.202090</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>415.212201</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>745.261002</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>278.153289</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>876.301487</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>147.112804</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>72.044390</td>\n",
       "      <td>36.525833</td>\n",
       "      <td>982.375714</td>\n",
       "      <td>491.691495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>232.075039</td>\n",
       "      <td>116.541158</td>\n",
       "      <td>822.345066</td>\n",
       "      <td>411.676171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>347.101982</td>\n",
       "      <td>174.054629</td>\n",
       "      <td>707.318123</td>\n",
       "      <td>354.162700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>476.144575</td>\n",
       "      <td>238.575926</td>\n",
       "      <td>578.275530</td>\n",
       "      <td>289.641403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>623.212989</td>\n",
       "      <td>312.110133</td>\n",
       "      <td>431.207116</td>\n",
       "      <td>216.107196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>760.271901</td>\n",
       "      <td>380.639589</td>\n",
       "      <td>294.148204</td>\n",
       "      <td>147.577740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>907.307301</td>\n",
       "      <td>454.157288</td>\n",
       "      <td>147.112804</td>\n",
       "      <td>74.060040</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          b_z1        b_z2        y_z1        y_z2\n",
       "0    72.044390    0.000000  909.359336    0.000000\n",
       "1   175.053575    0.000000  806.350151    0.000000\n",
       "2   290.080518    0.000000  691.323208    0.000000\n",
       "3   419.123111    0.000000  562.280615    0.000000\n",
       "4   566.191525    0.000000  415.212201    0.000000\n",
       "5   703.250437    0.000000  278.153289    0.000000\n",
       "6   834.290922    0.000000  147.112804    0.000000\n",
       "7   114.054955    0.000000  909.359336    0.000000\n",
       "8   217.064140    0.000000  806.350151    0.000000\n",
       "9   332.091083    0.000000  691.323208    0.000000\n",
       "10  461.133676    0.000000  562.280615    0.000000\n",
       "11  608.202090    0.000000  415.212201    0.000000\n",
       "12  745.261002    0.000000  278.153289    0.000000\n",
       "13  876.301487    0.000000  147.112804    0.000000\n",
       "14   72.044390   36.525833  982.375714  491.691495\n",
       "15  232.075039  116.541158  822.345066  411.676171\n",
       "16  347.101982  174.054629  707.318123  354.162700\n",
       "17  476.144575  238.575926  578.275530  289.641403\n",
       "18  623.212989  312.110133  431.207116  216.107196\n",
       "19  760.271901  380.639589  294.148204  147.577740\n",
       "20  907.307301  454.157288  147.112804   74.060040"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import alphabase.peptide.fragment as fragment\n",
    "\n",
    "df = pd.DataFrame()\n",
    "df['sequence'] = ['ACDEFHMK']*3\n",
    "df['mods'] = ['',\n",
    "    'Acetyl@Protein N-term', \n",
    "    'Carbamidomethyl@C;Oxidation@M'\n",
    "]\n",
    "df['mod_sites'] = ['','0','2;7']\n",
    "df['charge'] = [1,1,2]\n",
    "frag_types = fragment.get_charged_frag_types(['b','y'], 2)\n",
    "\n",
    "frag_df = fragment.create_fragment_mz_dataframe(\n",
    "    df, frag_types\n",
    ")\n",
    "assert 'frag_start_idx' in df.columns and 'frag_end_idx' in df.columns\n",
    "frag_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that all fragment ions are stored from peptide's N-terminal to C-terminal, so the b-ions are in the ascending order (from b1 to bn) and y-ions are in the decending order (from yn to y1)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Performance for fragment mz calculation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For large number of precursors in the precursor dataframe, we should use `create_fragment_mz_dataframe_by_sort_precursor`. It is similar to `create_fragment_mz_dataframe` but will change the order of the input precursor_df."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## alphabase.io.hdf\n",
    "\n",
    "It defines functionalities to operate HDF files. We use it to save the spectrum libraries."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## alphabase.spectral_library\n",
    "\n",
    "Spectrum library functionalities.\n",
    "\n",
    "### alphabase.spectral_library.base\n",
    "\n",
    "The base structure (`SpecLibBase` class) of the spectrum library in alphabase. The main functions of this class are to save (`SpecLibBase.save_hdf`) and load (`SpecLibBase.load_hdf`) library into/from hdf file, it will also generate `mod_seq_hash` and `mod_seq_charge_hash` for the precurosr_df.\n",
    "\n",
    "`SpecLibBase` does not define how to get its fragment_mz_df and fragment_intensity_df, this could be done in its sub-classes. We can also generate fragment_mz_df and fragment_intensity_df outside and assign them to `SpecLibBase._fragment_mz_df` and `SpecLibBase._fragment_intensity_df`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import alphabase.spectral_library.base as lib_base\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "import tempfile\n",
    "import os\n",
    "TEMPDIR = tempfile.gettempdir()\n",
    "hdf_path = os.path.join(TEMPDIR, \"alphabase_speclib.hdf\")\n",
    "\n",
    "hdflib = lib_base.SpecLibBase(['b_z1','y_z1'])\n",
    "\n",
    "df = pd.DataFrame()\n",
    "df['sequence'] = ['ACDEFHMK']*3\n",
    "df['mods'] = ['',\n",
    "    'Acetyl@Protein N-term', \n",
    "    'Carbamidomethyl@C;Oxidation@M'\n",
    "]\n",
    "df['mod_sites'] = ['','0','2;7']\n",
    "df['charge'] = [1,1,2]\n",
    "hdflib.precursor_df = df\n",
    "\n",
    "# generate fragment_mz_df outside\n",
    "frag_df = fragment.create_fragment_mz_dataframe(df, ['b_z1','y_z1'])\n",
    "hdflib._fragment_mz_df = frag_df\n",
    "\n",
    "hdflib._fragment_intensity_df = pd.DataFrame()\n",
    "hdflib.save_hdf(hdf_path)\n",
    "\n",
    "#test load_hdf\n",
    "_hdflib = lib_base.SpecLibBase(['b_z1','y_z1'])\n",
    "_hdflib.load_hdf(hdf_path, load_mod_seq=True)\n",
    "assert all(_hdflib.precursor_df.sequence.values==df.sequence.values)\n",
    "assert all(_hdflib.precursor_df.charge.values==df.charge.values)\n",
    "assert np.allclose(_hdflib.fragment_mz_df.values, frag_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we have `mod_seq_charge_hash` to uniqely identify the precursors, we split precursor_df into two sub dataframes in the hdf file. \n",
    "\n",
    "One is still `precursor_df`, but we remove all str columns from it, thus to speed up the data loading and save the memories:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>charge</th>\n",
       "      <th>frag_end_idx</th>\n",
       "      <th>frag_start_idx</th>\n",
       "      <th>mod_seq_charge_hash</th>\n",
       "      <th>mod_seq_hash</th>\n",
       "      <th>nAA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>-3613625140520963924</td>\n",
       "      <td>-3613625140520963925</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>7</td>\n",
       "      <td>-4469340833947449</td>\n",
       "      <td>-4469340833947450</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>21</td>\n",
       "      <td>14</td>\n",
       "      <td>4151334950393863929</td>\n",
       "      <td>4151334950393863927</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   charge  frag_end_idx  frag_start_idx  mod_seq_charge_hash  \\\n",
       "0       1             7               0 -3613625140520963924   \n",
       "1       1            14               7    -4469340833947449   \n",
       "2       2            21              14  4151334950393863929   \n",
       "\n",
       "          mod_seq_hash  nAA  \n",
       "0 -3613625140520963925    8  \n",
       "1    -4469340833947450    8  \n",
       "2  4151334950393863927    8  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_hdflib.load_hdf(hdf_path)\n",
    "_hdflib.precursor_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The other is `mod_seq_df`, all str columns are moved into this df, as well as the hash columns:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mod_seq_charge_hash</th>\n",
       "      <th>mod_seq_hash</th>\n",
       "      <th>mod_sites</th>\n",
       "      <th>mods</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-3613625140520963924</td>\n",
       "      <td>-3613625140520963925</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ACDEFHMK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-4469340833947449</td>\n",
       "      <td>-4469340833947450</td>\n",
       "      <td>0</td>\n",
       "      <td>Acetyl@Protein N-term</td>\n",
       "      <td>ACDEFHMK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4151334950393863929</td>\n",
       "      <td>4151334950393863927</td>\n",
       "      <td>2;7</td>\n",
       "      <td>Carbamidomethyl@C;Oxidation@M</td>\n",
       "      <td>ACDEFHMK</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   mod_seq_charge_hash         mod_seq_hash mod_sites  \\\n",
       "0 -3613625140520963924 -3613625140520963925             \n",
       "1    -4469340833947449    -4469340833947450         0   \n",
       "2  4151334950393863929  4151334950393863927       2;7   \n",
       "\n",
       "                            mods  sequence  \n",
       "0                                 ACDEFHMK  \n",
       "1          Acetyl@Protein N-term  ACDEFHMK  \n",
       "2  Carbamidomethyl@C;Oxidation@M  ACDEFHMK  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mod_seq_df = _hdflib.load_df_from_hdf(hdf_path, 'mod_seq_df')\n",
    "mod_seq_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can then easily merge precursor_df and mod_seq_df by `mod_seq_charge_hash` to create the original precursor dataframe. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>charge</th>\n",
       "      <th>frag_end_idx</th>\n",
       "      <th>frag_start_idx</th>\n",
       "      <th>mod_seq_charge_hash</th>\n",
       "      <th>mod_seq_hash</th>\n",
       "      <th>nAA</th>\n",
       "      <th>mod_sites</th>\n",
       "      <th>mods</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>-3613625140520963924</td>\n",
       "      <td>-3613625140520963925</td>\n",
       "      <td>8</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ACDEFHMK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>7</td>\n",
       "      <td>-4469340833947449</td>\n",
       "      <td>-4469340833947450</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>Acetyl@Protein N-term</td>\n",
       "      <td>ACDEFHMK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>21</td>\n",
       "      <td>14</td>\n",
       "      <td>4151334950393863929</td>\n",
       "      <td>4151334950393863927</td>\n",
       "      <td>8</td>\n",
       "      <td>2;7</td>\n",
       "      <td>Carbamidomethyl@C;Oxidation@M</td>\n",
       "      <td>ACDEFHMK</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   charge  frag_end_idx  frag_start_idx  mod_seq_charge_hash  \\\n",
       "0       1             7               0 -3613625140520963924   \n",
       "1       1            14               7    -4469340833947449   \n",
       "2       2            21              14  4151334950393863929   \n",
       "\n",
       "          mod_seq_hash  nAA mod_sites                           mods  sequence  \n",
       "0 -3613625140520963925    8                                           ACDEFHMK  \n",
       "1    -4469340833947450    8         0          Acetyl@Protein N-term  ACDEFHMK  \n",
       "2  4151334950393863927    8       2;7  Carbamidomethyl@C;Oxidation@M  ACDEFHMK  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "del mod_seq_df['mod_seq_hash']\n",
    "_hdflib.precursor_df = _hdflib.precursor_df.merge(mod_seq_df, on='mod_seq_charge_hash')\n",
    "_hdflib.precursor_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "... and it is extactly the same with the original one. That says, if we have the fragment_mz/intensity_df and unique identifiers for precursors, we don't need the original `['sequence','mods','mod_sites']` information any more during searching. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert all(_hdflib.precursor_df.sequence.values==df.sequence.values)\n",
    "assert all(_hdflib.precursor_df.charge.values==df.charge.values)\n",
    "assert np.allclose(_hdflib.fragment_mz_df.values, frag_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## alphabase.io.psm_reader\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from alphabase.io.psm_reader import psm_reader_provider"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`alphabase.io.psm_reader.psm_reader_provider` has registered some basic reader classes for the following search engine results:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'alphapept': alphabase.io.psm_reader.alphapept_reader.AlphaPeptReader,\n",
       " 'maxquant': alphabase.io.psm_reader.maxquant_reader.MaxQuantReader,\n",
       " 'spectronaut': alphabase.io.psm_reader.dia_search_reader.SpectronautReader,\n",
       " 'openswath': alphabase.io.psm_reader.dia_search_reader.SwathReader,\n",
       " 'swath': alphabase.io.psm_reader.dia_search_reader.SwathReader,\n",
       " 'diann': alphabase.io.psm_reader.dia_search_reader.DiannReader,\n",
       " 'pfind': alphabase.io.psm_reader.pfind_reader.pFindReader,\n",
       " 'msfragger_pepxml': alphabase.io.psm_reader.msfragger_reader.MSFraggerPepXML}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "psm_reader_provider.reader_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can then use `psm_reader_provider.get_reader` to get a reader and then `import_file`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>charge</th>\n",
       "      <th>rt</th>\n",
       "      <th>scan_num</th>\n",
       "      <th>raw_name</th>\n",
       "      <th>precursor_mz</th>\n",
       "      <th>score</th>\n",
       "      <th>proteins</th>\n",
       "      <th>decoy</th>\n",
       "      <th>spec_idx</th>\n",
       "      <th>mods</th>\n",
       "      <th>mod_sites</th>\n",
       "      <th>nAA</th>\n",
       "      <th>rt_norm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAAAAAGDSDSWDADAFSVEDPVRK</td>\n",
       "      <td>3</td>\n",
       "      <td>90.978</td>\n",
       "      <td>107307</td>\n",
       "      <td>20190402_QX1_SeVW_MA_HeLa_500ng_LC11</td>\n",
       "      <td>879.06841</td>\n",
       "      <td>148.950</td>\n",
       "      <td>sp|O75822|EIF3J_HUMAN</td>\n",
       "      <td>0</td>\n",
       "      <td>107306</td>\n",
       "      <td>Acetyl@Protein N-term</td>\n",
       "      <td>0</td>\n",
       "      <td>26</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAAAAAAAPAAAATAPTTAATTAATAAQ</td>\n",
       "      <td>3</td>\n",
       "      <td>70.261</td>\n",
       "      <td>81358</td>\n",
       "      <td>20190402_QX1_SeVW_MA_HeLa_500ng_LC11</td>\n",
       "      <td>790.07495</td>\n",
       "      <td>41.423</td>\n",
       "      <td>sp|P37108|SRP14_HUMAN</td>\n",
       "      <td>0</td>\n",
       "      <td>81357</td>\n",
       "      <td>Acetyl@Protein N-term;Oxidation@M</td>\n",
       "      <td>0;9</td>\n",
       "      <td>29</td>\n",
       "      <td>0.772286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAAAAAAAAPAAAATAPTTAATTAATAAQ</td>\n",
       "      <td>2</td>\n",
       "      <td>70.287</td>\n",
       "      <td>81391</td>\n",
       "      <td>20190402_QX1_SeVW_MA_HeLa_500ng_LC11</td>\n",
       "      <td>1184.60880</td>\n",
       "      <td>118.210</td>\n",
       "      <td>sp|P37108|SRP14_HUMAN</td>\n",
       "      <td>0</td>\n",
       "      <td>81390</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>30</td>\n",
       "      <td>0.772571</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         sequence  charge      rt  scan_num  \\\n",
       "0      AAAAAAAGDSDSWDADAFSVEDPVRK       3  90.978    107307   \n",
       "1   AAAAAAAAAPAAAATAPTTAATTAATAAQ       3  70.261     81358   \n",
       "2  AAAAAAAAAAPAAAATAPTTAATTAATAAQ       2  70.287     81391   \n",
       "\n",
       "                               raw_name  precursor_mz    score  \\\n",
       "0  20190402_QX1_SeVW_MA_HeLa_500ng_LC11     879.06841  148.950   \n",
       "1  20190402_QX1_SeVW_MA_HeLa_500ng_LC11     790.07495   41.423   \n",
       "2  20190402_QX1_SeVW_MA_HeLa_500ng_LC11    1184.60880  118.210   \n",
       "\n",
       "                proteins  decoy  spec_idx                               mods  \\\n",
       "0  sp|O75822|EIF3J_HUMAN      0    107306              Acetyl@Protein N-term   \n",
       "1  sp|P37108|SRP14_HUMAN      0     81357  Acetyl@Protein N-term;Oxidation@M   \n",
       "2  sp|P37108|SRP14_HUMAN      0     81390                                      \n",
       "\n",
       "  mod_sites  nAA   rt_norm  \n",
       "0         0   26  1.000000  \n",
       "1       0;9   29  0.772286  \n",
       "2             30  0.772571  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import io\n",
    "import numpy as np\n",
    "msms_txt = io.StringIO('''Raw file\tScan number\tScan index\tSequence\tLength\tMissed cleavages\tModifications\tModified sequence\tOxidation (M) Probabilities\tOxidation (M) Score diffs\tAcetyl (Protein N-term)\tOxidation (M)\tProteins\tCharge\tFragmentation\tMass analyzer\tType\tScan event number\tIsotope index\tm/z\tMass\tMass error [ppm]\tMass error [Da]\tSimple mass error [ppm]\tRetention time\tPEP\tScore\tDelta score\tScore diff\tLocalization prob\tCombinatorics\tPIF\tFraction of total spectrum\tBase peak fraction\tPrecursor full scan number\tPrecursor Intensity\tPrecursor apex fraction\tPrecursor apex offset\tPrecursor apex offset time\tMatches\tIntensities\tMass deviations [Da]\tMass deviations [ppm]\tMasses\tNumber of matches\tIntensity coverage\tPeak coverage\tNeutral loss level\tETD identification type\tReverse\tAll scores\tAll sequences\tAll modified sequences\tReporter PIF\tReporter fraction\tid\tProtein group IDs\tPeptide ID\tMod. peptide ID\tEvidence ID\tOxidation (M) site IDs\n",
    "20190402_QX1_SeVW_MA_HeLa_500ng_LC11\t81358\t73979\tAAAAAAAAAPAAAATAPTTAATTAATAAQ\t29\t0\tUnmodified\t_(Acetyl (Protein N-term))AAAAAAAAM(Oxidation (M))PAAAATAPTTAATTAATAAQ_\t\t\t0\t0\tsp|P37108|SRP14_HUMAN\t3\tHCD\tFTMS\tMULTI-MSMS\t13\t1\t790.07495\t2367.203\t0.35311\t0.00027898\t-0.061634807\t70.261\t0.012774\t41.423\t36.666\tNaN\tNaN\t1\t0\t0\t0\t81345\t10653955\t0.0338597821787898\t-11\t0.139877319335938\ty1;y2;y3;y4;y11;y1-NH3;y2-NH3;a2;b2;b3;b4;b5;b6;b7;b8;b9;b11;b12;b6(2+);b8(2+);b13(2+);b18(2+)\t2000000;2000000;300000;400000;200000;1000000;400000;300000;600000;1000000;2000000;3000000;3000000;3000000;3000000;2000000;600000;500000;1000000;2000000;300000;200000\t5.2861228709844E-06;-6.86980268369553E-05;-0.00238178789771837;0.000624715964988809;-0.0145624692099773;-0.000143471782706683;-0.000609501446461991;-0.000524972720768346;0.00010190530804266;5.8620815195809E-05;0.000229901232955854;-0.000108750048696038;-0.000229593152369034;0.00183148682538103;0.00276641182404092;0.000193118923334623;0.00200988580445483;0.000102216846016745;5.86208151389656E-05;0.000229901232955854;-0.00104559184393338;0.00525030008475369\t0.0359413365445091;-0.314964433555295;-8.23711898839045;1.60102421155213;-14.8975999917227;-1.10320467763838;-3.03102462870716;-4.56152475051625;0.712219104095465;0.273777366204575;0.806231096969562;-0.305312183824154;-0.537399178230218;3.67572664689217;4.85930954169285;0.301587577451224;2.48616190909398;0.116225745519871;0.273777365939099;0.806231096969562;-2.19774169175011;7.53961026980589\t147.076413378177;218.113601150127;289.153028027798;390.197699998035;977.50437775671;130.050013034583;201.087592852046;115.087114392821;143.081402136892;214.118559209185;285.155501716567;356.192954155649;427.230188786552;498.265241494374;569.301420357176;640.341107437877;808.429168310795;879.468189767554;214.118559209185;285.155501716567;475.757386711244;696.362265007215\t22\t0.262893575628735\t0.0826446280991736\tNone\tUnknown\t\t41.4230894199432;4.75668724862449;3.9515580701967\tAAAAAAAAAPAAAATAPTTAATTAATAAQ;FHRGPPDKDDMVSVTQILQGK;PVTLWITVTHMQADEVSVWR\t_AAAAAAAAAPAAAATAPTTAATTAATAAQ_;_FHRGPPDKDDMVSVTQILQGK_;_PVTLWITVTHMQADEVSVWR_\t\t\t0\t1443\t0\t0\t0\t\n",
    "20190402_QX1_SeVW_MA_HeLa_500ng_LC11\t81391\t74010\tAAAAAAAAAAPAAAATAPTTAATTAATAAQ\t29\t0\tUnmodified\t_AAAAAAAAAPAAAATAPTTAATTAATAAQ_\t\t\t0\t0\tsp|P37108|SRP14_HUMAN\t2\tHCD\tFTMS\tMULTI-MSMS\t14\t0\t1184.6088\t2367.203\t0.037108\t4.3959E-05\t1.7026696\t70.287\t7.1474E-09\t118.21\t100.52\tNaN\tNaN\t1\t0\t0\t0\t81377\t9347701\t0.166790347889974\t-10\t0.12664794921875\ty1;y2;y3;y4;y5;y9;y12;y13;y14;y20;y13-H2O;y20-H2O;y1-NH3;y20-NH3;b3;b4;b5;b6;b7;b8;b9;b11;b12;b13;b14;b15;b16;b19;b15-H2O;b16-H2O\t500000;600000;200000;400000;200000;100000;200000;1000000;200000;300000;200000;100000;100000;70000;300000;900000;2000000;3000000;5000000;8000000;6000000;600000;800000;600000;200000;300000;200000;300000;300000;1000000\t-0.000194444760495571;0.000149986878682284;0.000774202587820128;-0.0002445094036716;0.000374520568641401;-0.00694293246522193;-0.0109837291331587;-0.0037745820627606;-0.000945546471939451;0.00152326440706929;0.00506054832726477;0.00996886361417637;6.25847393393997E-05;-0.024881067836759;-3.11821549132674E-05;-0.000183099230639527;0.000161332473453513;0.000265434980121881;0.000747070697229901;0.000975534518261156;0.00101513939785036;0.00651913000274362;0.0058584595163893;0.00579536744021425;0.00131097834105276;-0.0131378531671089;0.00472955218901916;-0.00161006322559842;-0.00201443239325272;0.0227149399370319\t-1.32206444236914;0.687655553213019;2.6775131607882;-0.626628140021726;0.811995006209331;-8.6203492854282;-10.1838066275079;-3.21078702288986;-0.758483069159249;0.881072738747222;4.37168212373889;5.82682888353564;0.481236695337485;-14.5343986203644;-0.145630261806375;-0.642102166533079;0.452935954800214;0.621293379181583;1.49934012872483;1.71355878380837;1.58531240493271;8.06399202403175;6.6614096214532;6.09718023739784;1.28333378040908;-11.7030234519348;3.96235146626144;-1.07856912288932;-1.82370619437775;19.3220953109188\t147.07661310906;218.113382465221;289.149872037312;390.198569223404;461.235063981231;805.411965958065;1078.54847749073;1175.59403219566;1246.62831694787;1728.87474561429;1157.57463237897;1710.85573532879;130.049806978061;1711.87460084504;214.118649012155;285.155914717031;356.192684073126;427.22969375842;498.266325910503;569.303211234482;640.340285417402;808.424659066597;879.462433524883;950.49961040476;1021.54120858166;1122.60333588727;1193.62258226971;1492.77704268533;1104.58164778019;1175.59403219566\t30\t0.474003002083763\t0.167630057803468\tNone\tUnknown\t\t118.209976573419;17.6937689289157;17.2534171481793\tAAAAAAAAAPAAAATAPTTAATTAATAAQ;SELKQEAMQSEQLQSVLYLK;VGSSVPSKASELVVMGDHDAARR\t_AAAAAAAAAPAAAATAPTTAATTAATAAQ_;_SELKQEAM(Oxidation (M))QSEQLQSVLYLK_;_VGSSVPSKASELVVMGDHDAARR_\t\t\t1\t1443\t0\t0\t1\t\n",
    "20190402_QX1_SeVW_MA_HeLa_500ng_LC11\t107307\t98306\tAAAAAAAGDSDSWDADAFSVEDPVRK\t26\t1\tAcetyl (Protein N-term)\t_(Acetyl (Protein N-term))AAAAAAAGDSDSWDADAFSVEDPVRK_\t\t\t1\t0\tsp|O75822|EIF3J_HUMAN\t3\tHCD\tFTMS\tMULTI-MSMS\t10\t2\t879.06841\t2634.1834\t-0.93926\t-0.00082567\t-3.2012471\t90.978\t2.1945E-12\t148.95\t141.24\tNaN\tNaN\t1\t0\t0\t0\t107297\t10193939\t0.267970762043589\t-8\t0.10211181640625\ty1;y2;y4;y5;y6;y7;y8;y9;y10;y11;y12;y13;y14;y15;y17;y18;y19;y20;y21;y23;y21-H2O;y1-NH3;y19-NH3;y14(2+);y16(2+);y22(2+);a2;b2;b3;b4;b5;b6;b7\t300000;200000;3000000;600000;1000000;500000;2000000;1000000;1000000;1000000;90000;1000000;400000;900000;1000000;400000;3000000;2000000;1000000;400000;100000;200000;200000;80000;100000;200000;200000;2000000;5000000;5000000;5000000;2000000;300000\t1.34859050149316E-07;-6.05140996867704E-06;2.27812602133781E-05;0.00128986659160546;-0.00934536073077652;0.000941953783126337;-0.00160424237344614;-0.00239257341399934;-0.00111053968612396;-0.00331340710044969;0.00330702864630439;0.000963683996815234;0.00596290290945944;-0.00662057038289277;-0.0117122701335575;0.00777853472800416;0.0021841542961738;0.000144322111736983;-0.00087403893667215;0.0197121595674616;-0.021204007680808;-0.000308954599830713;-0.026636719419912;-0.0137790992353075;0.00596067266928912;-0.0077053835773313;9.11402199221811E-06;-0.000142539300128419;-0.000251999832926231;1.90791054137662E-05;-0.00236430185879044;-9.54583337602344E-05;-0.000556959493223985\t0.000916705048437201;-0.0199575598103408;0.0456231928690862;2.09952637717462;-12.5708704058425;1.11808305811426;-1.72590731777249;-2.22239181008062;-0.967696370445928;-2.62418809422166;2.47964286628144;0.665205752892023;3.64753748704453;-3.84510115530963;-6.08782672045773;3.81508105974837;1.04209904973991;0.0666012719936656;-0.390545453668809;8.28224925531311;-9.55133250134922;-2.37499239179248;-12.8127653858411;-16.846761946123;6.48662354975264;-6.67117082062383;0.0580151981289049;-0.770098855873447;-0.983876895688683;0.0583162347158579;-5.93738717724506;-0.203431522818505;-1.03087538746314\t147.112804035741;303.21392125011;499.33507018564;614.360746132308;743.413974455831;842.472101057517;929.506675663573;1076.57587791081;1147.61170966489;1262.6408555643;1333.67134891635;1448.700635293;1634.77494902759;1721.81956091078;1923.88362405243;2038.89107627957;2095.9181343836;2166.95728800359;2237.99542015244;2380.04906152953;2220.00518543488;130.0865640237;2078.92040615582;817.907873297785;918.917619246831;1155.02717356753;157.097144992378;185.0922112678;256.129434516133;327.166277224995;398.205774393759;469.240619338034;540.278194626993\t33\t0.574496146107112\t0.14410480349345\tNone\tUnknown\t\t148.951235201399;7.71201258444522;7.36039532447559\tAAAAAAAGDSDSWDADAFSVEDPVRK;PSRQESELMWQWVDQRSDGER;HTLTSFWNFKAGCEEKCYSNR\t_(Acetyl (Protein N-term))AAAAAAAGDSDSWDADAFSVEDPVRK_;_PSRQESELM(Oxidation (M))WQWVDQRSDGER_;_HTLTSFWNFKAGCEEKCYSNR_\t\t\t2\t625\t1\t1\t2\t'''\n",
    ")\n",
    "\n",
    "mq_reader = psm_reader_provider.get_reader('maxquant')\n",
    "mq_reader.import_file(msms_txt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/zengwenfeng/opt/anaconda3/lib/python3.8/site-packages/pandas/util/_decorators.py:311: ParserWarning: Length of header or names does not match length of data. This leads to a loss of data with index_col=False.\n",
      "  return func(*args, **kwargs)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>charge</th>\n",
       "      <th>raw_name</th>\n",
       "      <th>query_id</th>\n",
       "      <th>scan_num</th>\n",
       "      <th>score</th>\n",
       "      <th>proteins</th>\n",
       "      <th>uniprot_ids</th>\n",
       "      <th>fdr</th>\n",
       "      <th>decoy</th>\n",
       "      <th>spec_idx</th>\n",
       "      <th>mods</th>\n",
       "      <th>mod_sites</th>\n",
       "      <th>nAA</th>\n",
       "      <th>precursor_mz</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMIEAGAAAVHFEDQLASVK</td>\n",
       "      <td>2</td>\n",
       "      <td>Ecoli-1to1to1-un-C13-N15-10mM-20150823</td>\n",
       "      <td>Ecoli-1to1to1-un-C13-N15-10mM-20150823.30507.3...</td>\n",
       "      <td>30507</td>\n",
       "      <td>28.293201</td>\n",
       "      <td>gi|16131841|ref|NP_418439.1|</td>\n",
       "      <td>gi|16131841|ref|NP_418439.1|</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>30506</td>\n",
       "      <td>Oxidation@M</td>\n",
       "      <td>2</td>\n",
       "      <td>20</td>\n",
       "      <td>1037.517322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SVFLIKGDKVWVYPPEKKEK</td>\n",
       "      <td>4</td>\n",
       "      <td>XXX</td>\n",
       "      <td>XXX.25802.25802.4.0.dta</td>\n",
       "      <td>25802</td>\n",
       "      <td>2.294746</td>\n",
       "      <td>sp|P02790|HEMO_HUMAN</td>\n",
       "      <td>sp|P02790|HEMO_HUMAN</td>\n",
       "      <td>0.003207</td>\n",
       "      <td>0</td>\n",
       "      <td>25801</td>\n",
       "      <td>Didehydro@K^Any C-term</td>\n",
       "      <td>-1</td>\n",
       "      <td>20</td>\n",
       "      <td>597.838602</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               sequence  charge                                raw_name  \\\n",
       "0  AMIEAGAAAVHFEDQLASVK       2  Ecoli-1to1to1-un-C13-N15-10mM-20150823   \n",
       "1  SVFLIKGDKVWVYPPEKKEK       4                                     XXX   \n",
       "\n",
       "                                            query_id  scan_num      score  \\\n",
       "0  Ecoli-1to1to1-un-C13-N15-10mM-20150823.30507.3...     30507  28.293201   \n",
       "1                            XXX.25802.25802.4.0.dta     25802   2.294746   \n",
       "\n",
       "                       proteins                   uniprot_ids       fdr  \\\n",
       "0  gi|16131841|ref|NP_418439.1|  gi|16131841|ref|NP_418439.1|  0.000000   \n",
       "1          sp|P02790|HEMO_HUMAN          sp|P02790|HEMO_HUMAN  0.003207   \n",
       "\n",
       "   decoy  spec_idx                    mods mod_sites  nAA  precursor_mz  \n",
       "0      0     30506             Oxidation@M         2   20   1037.517322  \n",
       "1      0     25801  Didehydro@K^Any C-term        -1   20    597.838602  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from io import StringIO\n",
    "txt = StringIO(\"\"\"File_Name\tScan_No\tExp.MH+\tCharge\tQ-value\tSequence\tCalc.MH+\tMass_Shift(Exp.-Calc.)\tRaw_Score\tFinal_Score\tModification\tSpecificity\tProteins\tPositions\tLabel\tTarget/Decoy\tMiss.Clv.Sites\tAvg.Frag.Mass.Shift\tOthers\n",
    "Ecoli-1to1to1-un-C13-N15-10mM-20150823.30507.30507.2.0.dta\t30507\t2074.030369\t2\t0\tAMIEAGAAAVHFEDQLASVK\t2074.027271\t0.003098\t35.299588\t5.15726e-013\t2,Oxidation[M];\t3\tgi|16131841|ref|NP_418439.1|/\t173,K,K/\t1|0|\ttarget\t0\t0.948977\t131070\t0\t0\t0\t262143\t0\t0\t0\t32\n",
    "Ecoli-1to1to1-un-C13-N15-150mM-20150823.41501.41501.3.0.dta\t41501\t2712.197421\t3\t0\tEGDNYVVLSDILGDEDHLGDMDFK\t2712.198013\t-0.000592\t27.073978\t9.82619e-010\t21,Unknown[M];\t3\tgi|145698316|ref|NP_417633.4|/\t470,K,V/\t1|0|\ttarget\t0\t0.814438\t65596\t0\t0\t0\t4194288\t0\t0\t0\t36\n",
    "XXX.25802.25802.4.0.dta\t25802\t2388.339186\t4\t0.0032066\tSVFLIKGDKVWVYPPEKKEK\t2388.332468\t0.006718\t17.822784\t0.100787\t21,Didehydro[AnyC-termK];\t0\tsp|P02790|HEMO_HUMAN/\t106,N,G/\t1|0|\ttarget\t0\t0.704714\t36\n",
    "\"\"\")\n",
    "psm_reader_provider.get_reader('pfind').import_file(txt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>raw_name</th>\n",
       "      <th>sequence</th>\n",
       "      <th>charge</th>\n",
       "      <th>rt</th>\n",
       "      <th>mobility</th>\n",
       "      <th>proteins</th>\n",
       "      <th>uniprot_ids</th>\n",
       "      <th>genes</th>\n",
       "      <th>scan_num</th>\n",
       "      <th>score</th>\n",
       "      <th>spec_idx</th>\n",
       "      <th>mods</th>\n",
       "      <th>mod_sites</th>\n",
       "      <th>nAA</th>\n",
       "      <th>rt_norm</th>\n",
       "      <th>precursor_mz</th>\n",
       "      <th>ccs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>20201218_tims03_Evo03_PS_SA_HeLa_200ng_high_sp...</td>\n",
       "      <td>AAAAAAALESWQAAAPR</td>\n",
       "      <td>2</td>\n",
       "      <td>19.9208</td>\n",
       "      <td>1.19708</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q9UH36</td>\n",
       "      <td>SRRD</td>\n",
       "      <td>30053</td>\n",
       "      <td>0.999997</td>\n",
       "      <td>30052</td>\n",
       "      <td>Acetyl@Protein N-term</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>1.0</td>\n",
       "      <td>834.428635</td>\n",
       "      <td>483.435307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>20201218_tims03_Evo03_PS_SA_HeLa_200ng_high_sp...</td>\n",
       "      <td>AAAAAAALESWQAAAPR</td>\n",
       "      <td>2</td>\n",
       "      <td>19.9050</td>\n",
       "      <td>1.19500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q9UH36</td>\n",
       "      <td>SRRD</td>\n",
       "      <td>30029</td>\n",
       "      <td>0.995505</td>\n",
       "      <td>30028</td>\n",
       "      <td>Acetyl@Protein N-term</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>1.0</td>\n",
       "      <td>834.428635</td>\n",
       "      <td>482.595308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20201218_tims03_Evo03_PS_SA_HeLa_200ng_high_sp...</td>\n",
       "      <td>AAAAAAALESWQAAAPR</td>\n",
       "      <td>2</td>\n",
       "      <td>19.8893</td>\n",
       "      <td>1.19409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q9UH36</td>\n",
       "      <td>SRRD</td>\n",
       "      <td>30005</td>\n",
       "      <td>0.997286</td>\n",
       "      <td>30004</td>\n",
       "      <td>Acetyl@Protein N-term</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>1.0</td>\n",
       "      <td>834.428635</td>\n",
       "      <td>482.227809</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            raw_name           sequence  \\\n",
       "0  20201218_tims03_Evo03_PS_SA_HeLa_200ng_high_sp...  AAAAAAALESWQAAAPR   \n",
       "1  20201218_tims03_Evo03_PS_SA_HeLa_200ng_high_sp...  AAAAAAALESWQAAAPR   \n",
       "2  20201218_tims03_Evo03_PS_SA_HeLa_200ng_high_sp...  AAAAAAALESWQAAAPR   \n",
       "\n",
       "   charge       rt  mobility  proteins uniprot_ids genes  scan_num     score  \\\n",
       "0       2  19.9208   1.19708       NaN      Q9UH36  SRRD     30053  0.999997   \n",
       "1       2  19.9050   1.19500       NaN      Q9UH36  SRRD     30029  0.995505   \n",
       "2       2  19.8893   1.19409       NaN      Q9UH36  SRRD     30005  0.997286   \n",
       "\n",
       "   spec_idx                   mods mod_sites  nAA  rt_norm  precursor_mz  \\\n",
       "0     30052  Acetyl@Protein N-term         0   17      1.0    834.428635   \n",
       "1     30028  Acetyl@Protein N-term         0   17      1.0    834.428635   \n",
       "2     30004  Acetyl@Protein N-term         0   17      1.0    834.428635   \n",
       "\n",
       "          ccs  \n",
       "0  483.435307  \n",
       "1  482.595308  \n",
       "2  482.227809  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from io import StringIO\n",
    "tsv = StringIO('''File.Name\tRun\tProtein.Group\tProtein.Ids\tProtein.Names\tGenes\tPG.Quantity\tPG.Normalised\tPG.MaxLFQ\tGenes.Quantity\tGenes.Normalised\tGenes.MaxLFQ\tGenes.MaxLFQ.Unique\tModified.Sequence\tStripped.Sequence\tPrecursor.Id\tPrecursor.Charge\tQ.Value\tGlobal.Q.Value\tProtein.Q.Value\tPG.Q.Value\tGlobal.PG.Q.Value\tGG.Q.Value\tTranslated.Q.Value\tProteotypic\tPrecursor.Quantity\tPrecursor.Normalised\tPrecursor.Translated\tQuantity.Quality\tRT\tRT.Start\tRT.Stop\tiRT\tPredicted.RT\tPredicted.iRT\tLib.Q.Value\tMs1.Profile.Corr\tMs1.Area\tEvidence\tSpectrum.Similarity\tMass.Evidence\tCScore\tDecoy.Evidence\tDecoy.CScore\tFragment.Quant.Raw\tFragment.Quant.Corrected\tFragment.Correlations\tMS2.Scan\tIM\tiIM\tPredicted.IM\tPredicted.iIM\n",
    "F:\\XXX\\20201218_tims03_Evo03_PS_SA_HeLa_200ng_high_speed_21min_8cm_S2-A2_1_22636.d\t20201218_tims03_Evo03_PS_SA_HeLa_200ng_high_speed_21min_8cm_S2-A2_1_22636\tQ9UH36\tQ9UH36\t\tSRRD\t3296.49\t3428.89\t3428.89\t3296.49\t3428.89\t3428.89\t3428.89\t(UniMod:1)AAAAAAALESWQAAAPR\tAAAAAAALESWQAAAPR\t(UniMod:1)AAAAAAALESWQAAAPR2\t2\t3.99074e-05\t1.96448e-05\t0.000159821\t0.000159821\t0.000146135\t0.000161212\t0\t1\t3296.49\t3428.89\t3296.49\t0.852479\t19.9208\t19.8731\t19.9685\t123.9\t19.8266\t128.292\t0\t0.960106\t5308.05\t1.96902\t0.683134\t0.362287\t0.999997\t1.23691\t3.43242e-05\t1212.01;2178.03;1390.01;1020.01;714.008;778.008;\t1212.01;1351.73;887.591;432.92;216.728;732.751;\t0.956668;0.757581;0.670497;0.592489;0.47072;0.855203;\t30053\t1.19708\t1.19328\t1.19453\t1.19469\n",
    "F:\\XXX\\20201218_tims03_Evo03_PS_SA_HeLa_200ng_high_speed_21min_8cm_S2-A8_1_22642.d\t20201218_tims03_Evo03_PS_SA_HeLa_200ng_high_speed_21min_8cm_S2-A8_1_22642\tQ9UH36\tQ9UH36\t\tSRRD\t2365\t2334.05\t2334.05\t2365\t2334.05\t2334.05\t2334.05\t(UniMod:1)AAAAAAALESWQAAAPR\tAAAAAAALESWQAAAPR\t(UniMod:1)AAAAAAALESWQAAAPR2\t2\t0.000184434\t1.96448e-05\t0.000596659\t0.000596659\t0.000146135\t0.000604961\t0\t1\t2365\t2334.05\t2365\t0.922581\t19.905\t19.8573\t19.9527\t123.9\t19.782\t128.535\t0\t0.940191\t4594.04\t1.31068\t0.758988\t0\t0.995505\t0.28633\t2.12584e-06\t1209.02;1210.02;1414.02;1051.01;236.003;130.002;\t1209.02;1109.89;732.154;735.384;0;46.0967;\t0.919244;0.937624;0.436748;0.639369;0.296736;0.647924;\t30029\t1.195\t1.19328\t1.19381\t1.19339\n",
    "F:\\XXX\\20201218_tims03_Evo03_PS_SA_HeLa_200ng_high_speed_21min_8cm_S2-B2_1_22648.d\t20201218_tims03_Evo03_PS_SA_HeLa_200ng_high_speed_21min_8cm_S2-B2_1_22648\tQ9UH36\tQ9UH36\t\tSRRD\t1664.51\t1635.46\t1635.47\t1664.51\t1635.46\t1635.47\t1635.47\t(UniMod:1)AAAAAAALESWQAAAPR\tAAAAAAALESWQAAAPR\t(UniMod:1)AAAAAAALESWQAAAPR2\t2\t0.000185123\t1.96448e-05\t0.000307409\t0.000307409\t0.000146135\t0.000311332\t0\t1\t1664.51\t1635.46\t1664.51\t0.811147\t19.8893\t19.8416\t19.937\t123.9\t19.7567\t128.896\t0\t0.458773\t6614.06\t1.7503\t0.491071\t0.00111683\t0.997286\t1.92753\t2.80543e-05\t744.01;1708.02;1630.02;1475.02;0;533.006;\t322.907;808.594;577.15;536.033;0;533.006;\t0.760181;0.764072;0.542005;0.415779;0;0.913438;\t30005\t1.19409\t1.19328\t1.19323\t1.19308\n",
    "''')\n",
    "psm_reader_provider.get_reader('diann').import_file(tsv)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that, to support both Bruker and Thermo data, we did not use `Scan Number` in the output dataframe but `spec_idx`. It is the scan number in thermo data. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also read spectronaut/openswath's precursor with out fragment information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>raw_name</th>\n",
       "      <th>sequence</th>\n",
       "      <th>charge</th>\n",
       "      <th>rt</th>\n",
       "      <th>precursor_mz</th>\n",
       "      <th>mobility</th>\n",
       "      <th>proteins</th>\n",
       "      <th>uniprot_ids</th>\n",
       "      <th>genes</th>\n",
       "      <th>mods</th>\n",
       "      <th>mod_sites</th>\n",
       "      <th>nAA</th>\n",
       "      <th>rt_norm</th>\n",
       "      <th>ccs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>202106018_TIMS03_EVO03_PaSk_SA_HeLa_EGF_Phosph...</td>\n",
       "      <td>ALVATPGK</td>\n",
       "      <td>2</td>\n",
       "      <td>-5.032703</td>\n",
       "      <td>418.717511</td>\n",
       "      <td>0.758</td>\n",
       "      <td>NUCL_HUMAN</td>\n",
       "      <td>P19338</td>\n",
       "      <td>NCL</td>\n",
       "      <td>Phospho@T</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>308.612143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>202106018_TIMS03_EVO03_PaSk_SA_HeLa_EGF_Phosph...</td>\n",
       "      <td>TLTPCPLR</td>\n",
       "      <td>2</td>\n",
       "      <td>27.716590</td>\n",
       "      <td>439.230786</td>\n",
       "      <td>0.818</td>\n",
       "      <td>ZC3HD_HUMAN</td>\n",
       "      <td>Q5T200</td>\n",
       "      <td>ZC3H13</td>\n",
       "      <td>Phospho@T;Carbamidomethyl@C</td>\n",
       "      <td>3;5</td>\n",
       "      <td>8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>332.788837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>202106018_TIMS03_EVO03_PaSk_SA_HeLa_EGF_Phosph...</td>\n",
       "      <td>LFVSPPEGSSR</td>\n",
       "      <td>2</td>\n",
       "      <td>38.050310</td>\n",
       "      <td>635.297385</td>\n",
       "      <td>0.917</td>\n",
       "      <td>MAP7_HUMAN</td>\n",
       "      <td>Q14244;Q14244-6;Q14244-7</td>\n",
       "      <td>MAP7</td>\n",
       "      <td>Acetyl@Protein N-term;Phospho@S</td>\n",
       "      <td>0;4</td>\n",
       "      <td>11</td>\n",
       "      <td>1.0</td>\n",
       "      <td>371.282739</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            raw_name     sequence  charge  \\\n",
       "0  202106018_TIMS03_EVO03_PaSk_SA_HeLa_EGF_Phosph...     ALVATPGK       2   \n",
       "1  202106018_TIMS03_EVO03_PaSk_SA_HeLa_EGF_Phosph...     TLTPCPLR       2   \n",
       "2  202106018_TIMS03_EVO03_PaSk_SA_HeLa_EGF_Phosph...  LFVSPPEGSSR       2   \n",
       "\n",
       "          rt  precursor_mz  mobility     proteins               uniprot_ids  \\\n",
       "0  -5.032703    418.717511     0.758   NUCL_HUMAN                    P19338   \n",
       "1  27.716590    439.230786     0.818  ZC3HD_HUMAN                    Q5T200   \n",
       "2  38.050310    635.297385     0.917   MAP7_HUMAN  Q14244;Q14244-6;Q14244-7   \n",
       "\n",
       "    genes                             mods mod_sites  nAA  rt_norm         ccs  \n",
       "0     NCL                        Phospho@T         5    8      0.0  308.612143  \n",
       "1  ZC3H13      Phospho@T;Carbamidomethyl@C       3;5    8      1.0  332.788837  \n",
       "2    MAP7  Acetyl@Protein N-term;Phospho@S       0;4   11      1.0  371.282739  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from io import StringIO\n",
    "tsv = StringIO('''ReferenceRun\tPrecursorCharge\tWorkflow\tIntModifiedPeptide\tCV\tAllowForNormalization\tModifiedPeptide\tStrippedPeptide\tiRT\tIonMobility\tiRTSourceSpecific\tBGSInferenceId\tIsProteotypic\tIntLabeledPeptide\tLabeledPeptide\tPrecursorMz\tReferenceRunQvalue\tReferenceRunMS1Response\tFragmentLossType\tFragmentNumber\tFragmentType\tFragmentCharge\tFragmentMz\tRelativeIntensity\tExcludeFromAssay\tDatabase\tProteinGroups\tUniProtIds\tProtein Name\tProteinDescription\tOrganisms\tOrganismId\tGenes\tProtein Existence\tSequence Version\tFASTAName\n",
    "202106018_TIMS03_EVO03_PaSk_SA_HeLa_EGF_Phospho_100ug_test_S4-A1_1_25843\t2\t\t_ALVAT[+80]PGK_\t\tTrue\t_ALVAT[Phospho (STY)]PGK_\tALVATPGK\t-5.032703\t0.758\t-5.032703\tP19338\tFalse\t_ALVAT[+80]PGK_\t_ALVAT[Phospho (STY)]PGK_\t418.717511324722\t0\t10352\tnoloss\t3\ty\t1\t301.187031733932\t53.1991\tFalse\tsp\tP19338\tP19338\tNUCL_HUMAN\tNucleolin\tHomo sapiens\t\tNCL\t1\t3\tMCT_human_UP000005640_9606\n",
    "202106018_TIMS03_EVO03_PaSk_SA_HeLa_EGF_Phospho_100ug_test_S4-A1_1_25843\t2\t\t_ALVAT[+80]PGK_\t\tTrue\t_ALVAT[Phospho (STY)]PGK_\tALVATPGK\t-5.032703\t0.758\t-5.032703\tP19338\tFalse\t_ALVAT[+80]PGK_\t_ALVAT[Phospho (STY)]PGK_\t418.717511324722\t0\t10352\tH3PO4\t4\ty\t1\t384.224142529733\t26.31595\tFalse\tsp\tP19338\tP19338\tNUCL_HUMAN\tNucleolin\tHomo sapiens\t\tNCL\t1\t3\tMCT_human_UP000005640_9606\n",
    "202106018_TIMS03_EVO03_PaSk_SA_HeLa_EGF_Phospho_100ug_test_S4-A1_1_25843\t2\t\t_TLT[+80]PCPLR_\t\tTrue\t_TLT[Phospho (STY)]PC[Carbamidomethyl (C)]PLR_\tTLTPCPLR\t27.71659\t0.818\t27.71659\tQ5T200\tFalse\t_TLT[+80]PPLR_\t_TLT[Phospho (STY)]PPLR_\t439.230785875227\t0.000138389150379226\t23117\tnoloss\t3\tb\t1\t396.153027901512\t6.3264\tFalse\tsp\tQ5T200\tQ5T200\tZC3HD_HUMAN\tZinc finger CCCH domain-containing protein 13\tHomo sapiens\t\tZC3H13\t1\t1\tMCT_human_UP000005640_9606\n",
    "202106018_TIMS03_EVO03_PaSk_SA_HeLa_EGF_Phospho_100ug_test_S4-A1_1_25843\t2\t\t_TLT[+80]PCPLR_\t\tTrue\t_TLT[Phospho (STY)]PC[Carbamidomethyl (C)]PLR_\tTLTPCPLR\t27.71659\t0.818\t27.71659\tQ5T200\tFalse\t_TLT[+80]PPLR_\t_TLT[Phospho (STY)]PPLR_\t439.230785875227\t0.000138389150379226\t23117\tnoloss\t3\ty\t1\t385.255780000092\t29.70625\tFalse\tsp\tQ5T200\tQ5T200\tZC3HD_HUMAN\tZinc finger CCCH domain-containing protein 13\tHomo sapiens\t\tZC3H13\t1\t1\tMCT_human_UP000005640_9606\n",
    "202106018_TIMS03_EVO03_PaSk_SA_HeLa_EGF_Phospho_library25_S4-C1_1_25867\t2\t\t_LFVT[+80]PPEGSSR_\t\tTrue\t_[Acetyl (Protein N-term)]LFVS[Phospho (STY)]PPEGSSR_\tLFVSPPEGSSR\t38.05031\t0.917\t38.05031\tQ14244;Q14244-6;Q14244-7\tFalse\t_LFVT[+80]PPEGSSR_\t_LFVT[Phospho (STY)]PPEGSSR_\t635.297385373987\t0\t14164\tH3PO4\t4\tb\t1\t443.265279065723\t12.24525\tFalse\tsp\tQ14244;Q14244-6;Q14244-7\tQ14244;Q14244-6;Q14244-7\tMAP7_HUMAN\tEnsconsin;Isoform of Q14244, Isoform 6 of Ensconsin;Isoform of Q14244, Isoform 7 of Ensconsin\tHomo sapiens\t\tMAP7\t1;;\t1;;\tMCT_human_UP000005640_9606;MCT_human2_UP000005640_9606_additional;MCT_human2_UP000005640_9606_additional\n",
    "202106018_TIMS03_EVO03_PaSk_SA_HeLa_EGF_Phospho_library25_S4-C1_1_25867\t2\t\t_LFVT[+80]PPEGSSR_\t\tTrue\t_[Acetyl (Protein N-term)]LFVS[Phospho (STY)]PPEGSSR_\tLFVSPPEGSSR\t38.05031\t0.917\t38.05031\tQ14244;Q14244-6;Q14244-7\tFalse\t_LFVT[+80]PPEGSSR_\t_LFVT[Phospho (STY)]PPEGSSR_\t635.297385373987\t0\t14164\tnoloss\t6\ty\t1\t632.299829640042\t46.07855\tFalse\tsp\tQ14244;Q14244-6;Q14244-7\tQ14244;Q14244-6;Q14244-7\tMAP7_HUMAN\tEnsconsin;Isoform of Q14244, Isoform 6 of Ensconsin;Isoform of Q14244, Isoform 7 of Ensconsin\tHomo sapiens\t\tMAP7\t1;;\t1;;\tMCT_human_UP000005640_9606;MCT_human2_UP000005640_9606_additional;MCT_human2_UP000005640_9606_additional\n",
    "202106018_TIMS03_EVO03_PaSk_SA_HeLa_EGF_Phospho_library25_S4-C1_1_25867\t2\t\t_LFVT[+80]PPEGSSR_\t\tTrue\t_[Acetyl (Protein N-term)]LFVS[Phospho (STY)]PPEGSSR_\tLFVSPPEGSSR\t38.05031\t0.917\t38.05031\tQ14244;Q14244-6;Q14244-7\tFalse\t_LFVT[+80]PPEGSSR_\t_LFVT[Phospho (STY)]PPEGSSR_\t635.297385373987\t0\t14164\tnoloss\t7\ty\t1\t729.352593488892\t100\tFalse\tsp\tQ14244;Q14244-6;Q14244-7\tQ14244;Q14244-6;Q14244-7\tMAP7_HUMAN\tEnsconsin;Isoform of Q14244, Isoform 6 of Ensconsin;Isoform of Q14244, Isoform 7 of Ensconsin\tHomo sapiens\t\tMAP7\t1;;\t1;;\tMCT_human_UP000005640_9606;MCT_human2_UP000005640_9606_additional;MCT_human2_UP000005640_9606_additional\n",
    "''')\n",
    "\n",
    "psm_reader_provider.get_reader('spectronaut').import_file(tsv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>charge</th>\n",
       "      <th>rt</th>\n",
       "      <th>precursor_mz</th>\n",
       "      <th>mods</th>\n",
       "      <th>mod_sites</th>\n",
       "      <th>nAA</th>\n",
       "      <th>rt_norm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAAAAAAAASGAAIPPLIPPRR</td>\n",
       "      <td>3</td>\n",
       "      <td>59.0</td>\n",
       "      <td>685.732240</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>23</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAAAAAAAASGAAIPPLIPPRR</td>\n",
       "      <td>4</td>\n",
       "      <td>59.2</td>\n",
       "      <td>514.550999</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>23</td>\n",
       "      <td>0.004673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAAAAAAAASGAAIPPLIPPRRVITLYQCFSVSQR</td>\n",
       "      <td>5</td>\n",
       "      <td>101.8</td>\n",
       "      <td>728.201724</td>\n",
       "      <td>Carbamidomethyl@C</td>\n",
       "      <td>30</td>\n",
       "      <td>36</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               sequence  charge     rt  precursor_mz  \\\n",
       "0               AAAAAAAAAASGAAIPPLIPPRR       3   59.0    685.732240   \n",
       "1               AAAAAAAAAASGAAIPPLIPPRR       4   59.2    514.550999   \n",
       "2  AAAAAAAAAASGAAIPPLIPPRRVITLYQCFSVSQR       5  101.8    728.201724   \n",
       "\n",
       "                mods mod_sites  nAA   rt_norm  \n",
       "0                                23  0.000000  \n",
       "1                                23  0.004673  \n",
       "2  Carbamidomethyl@C        30   36  1.000000  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from io import StringIO\n",
    "tsv = StringIO('''PrecursorMz\tProductMz\tTr_recalibrated\ttransition_name\tCE\tLibraryIntensity\ttransition_group_id\tdecoy\tPeptideSequence\tProteinName\tAnnotation\tFullUniModPeptideName\tPrecursorCharge\tGroupLabel\tUniprotID\tFragmentType\tFragmentCharge\tFragmentSeriesNumber\n",
    "685.732240417\t886.020494795\t59.0\t255_AAAAAAAAAASGAAIPPLIPPRR_3\t-1\t5257.9\t13_AAAAAAAAAASGAAIPPLIPPRR_3\t0\tAAAAAAAAAASGAAIPPLIPPRR\t1/O14654\ty19^2/0.002\tAAAAAAAAAASGAAIPPLIPPRR\t3\tlight\t1/O14654\ty\t2\t19\n",
    "514.550999438\t473.303261576\t59.2\t268_AAAAAAAAAASGAAIPPLIPPRR_4\t-1\t10000.0\t14_AAAAAAAAAASGAAIPPLIPPRR_4\t0\tAAAAAAAAAASGAAIPPLIPPRR\t1/O14654\ty8^2/0.002\tAAAAAAAAAASGAAIPPLIPPRR\t4\tlight\t1/O14654\ty\t2\t8\n",
    "514.550999438\t629.39313922\t59.2\t276_AAAAAAAAAASGAAIPPLIPPRR_4\t-1\t5923.1\t14_AAAAAAAAAASGAAIPPLIPPRR_4\t0\tAAAAAAAAAASGAAIPPLIPPRR\t1/O14654\ty12^2/0.001\tAAAAAAAAAASGAAIPPLIPPRR\t4\tlight\t1/O14654\ty\t2\t12\n",
    "514.550999438\t672.909153425\t59.2\t279_AAAAAAAAAASGAAIPPLIPPRR_4\t-1\t5249.8\t14_AAAAAAAAAASGAAIPPLIPPRR_4\t0\tAAAAAAAAAASGAAIPPLIPPRR\t1/O14654\ty13^2/0.001\tAAAAAAAAAASGAAIPPLIPPRR\t4\tlight\t1/O14654\ty\t2\t13\n",
    "514.550999438\t356.19284545\t59.2\t262_AAAAAAAAAASGAAIPPLIPPRR_4\t-1\t5233.6\t14_AAAAAAAAAASGAAIPPLIPPRR_4\t0\tAAAAAAAAAASGAAIPPLIPPRR\t1/O14654\tb5/0.001,b10^2/0.001,m6:10/0.001\tAAAAAAAAAASGAAIPPLIPPRR\t4\tlight\t1/O14654\tb\t1\t5\n",
    "514.550999438\t498.26707303\t59.2\t269_AAAAAAAAAASGAAIPPLIPPRR_4\t-1\t4976.0\t14_AAAAAAAAAASGAAIPPLIPPRR_4\t0\tAAAAAAAAAASGAAIPPLIPPRR\t1/O14654\tb7/0.001,m4:10/0.001\tAAAAAAAAAASGAAIPPLIPPRR\t4\tlight\t1/O14654\tb\t1\t7\n",
    "514.550999438\t427.22995924\t59.2\t265_AAAAAAAAAASGAAIPPLIPPRR_4\t-1\t4859.4\t14_AAAAAAAAAASGAAIPPLIPPRR_4\t0\tAAAAAAAAAASGAAIPPLIPPRR\t1/O14654\tb6/0.002,m5:10/0.002\tAAAAAAAAAASGAAIPPLIPPRR\t4\tlight\t1/O14654\tb\t1\t6\n",
    "728.201724416\t356.19284545\t101.8\t292_AAAAAAAAAASGAAIPPLIPPRRVITLYQC(UniMod:4)FSVSQR_5\t-1\t10000.0\t15_AAAAAAAAAASGAAIPPLIPPRRVITLYQC(UniMod:4)FSVSQR_5\t0\tAAAAAAAAAASGAAIPPLIPPRRVITLYQCFSVSQR\t1/O14654\tb5/0.003,b10^2/0.003,m6:10/0.003\tAAAAAAAAAASGAAIPPLIPPRRVITLYQC(UniMod:4)FSVSQR\t5\tlight\t1/O14654\tb\t1\t5\n",
    "728.201724416\t576.310000482\t101.8\t297_AAAAAAAAAASGAAIPPLIPPRRVITLYQC(UniMod:4)FSVSQR_5\t-1\t7611.0\t15_AAAAAAAAAASGAAIPPLIPPRRVITLYQC(UniMod:4)FSVSQR_5\t0\tAAAAAAAAAASGAAIPPLIPPRRVITLYQCFSVSQR\t1/O14654\ty5/0.002\tAAAAAAAAAASGAAIPPLIPPRRVITLYQC(UniMod:4)FSVSQR\t5\tlight\t1/O14654\ty\t1\t5\n",
    "728.201724416\t427.22995924\t101.8\t293_AAAAAAAAAASGAAIPPLIPPRRVITLYQC(UniMod:4)FSVSQR_5\t-1\t6805.1\t15_AAAAAAAAAASGAAIPPLIPPRRVITLYQC(UniMod:4)FSVSQR_5\t0\tAAAAAAAAAASGAAIPPLIPPRRVITLYQCFSVSQR\t1/O14654\tb6/-0.002,m5:10/-0.002\tAAAAAAAAAASGAAIPPLIPPRRVITLYQC(UniMod:4)FSVSQR\t5\tlight\t1/O14654\tb\t1\t6\n",
    "728.201724416\t569.30418682\t101.8\t296_AAAAAAAAAASGAAIPPLIPPRRVITLYQC(UniMod:4)FSVSQR_5\t-1\t6312.7\t15_AAAAAAAAAASGAAIPPLIPPRRVITLYQC(UniMod:4)FSVSQR_5\t0\tAAAAAAAAAASGAAIPPLIPPRRVITLYQCFSVSQR\t1/O14654\tb8/0.009,m3:10/0.009\tAAAAAAAAAASGAAIPPLIPPRRVITLYQC(UniMod:4)FSVSQR\t5\tlight\t1/O14654\tb\t1\t8\n",
    "''')\n",
    "\n",
    "osw_reader = psm_reader_provider.get_reader('openswath')\n",
    "osw_reader.import_file(tsv)\n",
    "osw_reader.psm_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.3 ('base')",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
