{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#default_exp protein.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "# for nbdev_build_docs\n",
    "# import os\n",
    "# __file__ = os.path.expanduser('~/Workspace/AlphaBase/alphabase/protein/fasta.py')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "# notebook does not have __file__\n",
    "import alphabase.protein.fasta\n",
    "__file__ = alphabase.protein.fasta.__file__"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fasta processing and peptide generation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "\n",
    "import regex as re\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import numba\n",
    "import os\n",
    "import itertools\n",
    "from Bio import SeqIO\n",
    "from typing import Union\n",
    "\n",
    "from alphabase.yaml_utils import load_yaml\n",
    "from alphabase.io.hdf import HDF_File\n",
    "from alphabase.utils import explode_multiple_columns\n",
    "\n",
    "protease_dict = load_yaml(\n",
    "    os.path.join(\n",
    "        os.path.dirname(\n",
    "            __file__\n",
    "        ), \n",
    "        'protease.yaml'\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "p = re.compile(protease_dict['trypsin'])\n",
    "\n",
    "idx = '0123456789012345678901234567890123456789012345'\n",
    "seq = 'ABDNGKENGLANGIXHGRKTNGLANGKVHNAKHNARKANGKPFAAT'\n",
    "cut_pos = np.array([m.start()+1 for m in p.finditer(seq)])\n",
    "assert np.all(cut_pos==np.array([6, 18, 19, 27, 32, 36, 37]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def get_uniprot_gene_name(description:str):\n",
    "    idx = description.find(' GN=')\n",
    "    if idx == -1: return ''\n",
    "    else: idx += 4\n",
    "    return description[idx:description.find(' ', idx)]\n",
    "\n",
    "def read_fasta_file(fasta_filename:str=\"\"):\n",
    "    \"\"\"\n",
    "    Read a FASTA file line by line\n",
    "    Args:\n",
    "        fasta_filename (str): fasta.\n",
    "    Yields:\n",
    "        dict {id:str, name:str, description:str, sequence:str}: protein information.\n",
    "    \"\"\"\n",
    "    with open(fasta_filename, \"rt\") as handle:\n",
    "        iterator = SeqIO.parse(handle, \"fasta\")\n",
    "        while iterator:\n",
    "            try:\n",
    "                record = next(iterator)\n",
    "                parts = record.id.split(\"|\")  # pipe char\n",
    "                if len(parts) > 1:\n",
    "                    id = parts[1]\n",
    "                else:\n",
    "                    id = record.name\n",
    "                sequence = str(record.seq)\n",
    "                entry = {\n",
    "                    \"protein_id\": id,\n",
    "                    \"full_name\": record.name,\n",
    "                    \"gene_name\": get_uniprot_gene_name(record.description),\n",
    "                    \"description\": record.description,\n",
    "                    \"sequence\": sequence,\n",
    "                }\n",
    "\n",
    "                yield entry\n",
    "            except StopIteration:\n",
    "                break\n",
    "\n",
    "def load_all_proteins(fasta_file_list:list):\n",
    "    protein_dict = {}\n",
    "    for fasta in fasta_file_list:\n",
    "        for protein in read_fasta_file(fasta):\n",
    "            protein_dict[protein['protein_id']] = protein\n",
    "    return protein_dict\n",
    "\n",
    "def concat_proteins(protein_dict:dict, sep='$')->str:\n",
    "    \"\"\"Concatenate all protein sequences into a single sequence, \n",
    "    seperated by `sep ($ by default)`.\n",
    "\n",
    "    Args:\n",
    "        protein_dict (dict): protein_dict by read_fasta_file()\n",
    "\n",
    "    Returns:\n",
    "        str: concatenated sequence seperated by `sep`.\n",
    "    \"\"\"\n",
    "    seq_list = ['']\n",
    "    seq_count = 1\n",
    "    for key in protein_dict:\n",
    "        protein_dict[key]['offset'] = seq_count\n",
    "        seq_list.append(protein_dict[key]['sequence'])\n",
    "        seq_count += protein_dict[key]['sequence']+1\n",
    "    seq_list.append('')\n",
    "    return '$'.join(seq_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "assert get_uniprot_gene_name('sp|Q9H9K5|MER34_HUMAN Endogenous retroviral envelope protein HEMO OS=Homo sapiens OX=9606 GN=ERVMER34-1 PE=1 SV=1') == 'ERVMER34-1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "@numba.njit\n",
    "def cleave_sequence_with_cut_pos(\n",
    "    sequence:str,\n",
    "    cut_pos:np.ndarray,\n",
    "    n_missed_cleavages:int=2,\n",
    "    pep_length_min:int=6,\n",
    "    pep_length_max:int=45,\n",
    ")->np.ndarray:\n",
    "    \"\"\"\n",
    "    Cleave a sequence with cut postions (cut_pos). \n",
    "    Filters to have a minimum and maximum length.\n",
    "    Args:\n",
    "        sequence (str): protein sequence\n",
    "        cut_pos (np.array): cut postions determined by a given protease.\n",
    "        n_missed_cleavages (int): the number of max missed cleavages.\n",
    "        pep_length_min (int): min peptide length.\n",
    "        pep_length_max (int): max peptide length.\n",
    "    Returns:\n",
    "        list (str): cleaved peptide sequences with missed cleavages.\n",
    "        list (int): number of miss cleavage of each peptide.\n",
    "        list (bool): if N-term peptide\n",
    "        list (bool): if C-term pepetide\n",
    "    \"\"\"\n",
    "    seq_list = []\n",
    "    miss_list = []\n",
    "    nterm_list = []\n",
    "    cterm_list = []\n",
    "    for i,start_pos in enumerate(cut_pos):\n",
    "        for n_miss,end_pos in enumerate(\n",
    "            cut_pos[i+1:i+2+n_missed_cleavages]\n",
    "        ):\n",
    "            if end_pos > start_pos + pep_length_max:\n",
    "                break\n",
    "            elif end_pos < start_pos + pep_length_min:\n",
    "                continue\n",
    "            else:\n",
    "                seq_list.append(sequence[start_pos:end_pos])\n",
    "                miss_list.append(n_miss)\n",
    "                if start_pos == 0:\n",
    "                    nterm_list.append(True)\n",
    "                else:\n",
    "                    nterm_list.append(False)\n",
    "                if end_pos == len(sequence):\n",
    "                    cterm_list.append(True)\n",
    "                else:\n",
    "                    cterm_list.append(False)\n",
    "    return seq_list, miss_list, nterm_list, cterm_list\n",
    "\n",
    "class Digest(object):\n",
    "    def __init__(self,\n",
    "        protease:str='trypsin',\n",
    "        max_missed_cleavages:int=2,\n",
    "        peptide_length_min:int=6,\n",
    "        peptide_length_max:int=45,\n",
    "    ):\n",
    "        self.n_miss_cleave = max_missed_cleavages\n",
    "        self.peptide_length_min = peptide_length_min\n",
    "        self.peptide_length_max = peptide_length_max\n",
    "        if protease in protease_dict:\n",
    "            self.regex_pattern = re.compile(\n",
    "                protease_dict[protease]\n",
    "            )\n",
    "        else:\n",
    "            self.regex_pattern = re.compile(\n",
    "                protease\n",
    "            )\n",
    "\n",
    "    def cleave_sequence(self,\n",
    "        sequence:str,\n",
    "    )->list:\n",
    "        \"\"\"\n",
    "        Cleave a sequence.\n",
    "        Args:\n",
    "            sequence (str): the given (protein) sequence.\n",
    "        Returns:\n",
    "            list (of str): cleaved peptide sequences with missed cleavages.\n",
    "        \"\"\"\n",
    "\n",
    "        cut_pos = [0]\n",
    "        cut_pos.extend([\n",
    "            m.start()+1 for m in \n",
    "            self.regex_pattern.finditer(sequence)\n",
    "        ])\n",
    "        cut_pos.append(len(sequence))\n",
    "        cut_pos = np.array(cut_pos, dtype=np.int64)\n",
    "\n",
    "        (\n",
    "            seq_list, miss_list, nterm_list, cterm_list\n",
    "        ) = cleave_sequence_with_cut_pos(\n",
    "            sequence, cut_pos, \n",
    "            self.n_miss_cleave,\n",
    "            self.peptide_length_min,\n",
    "            self.peptide_length_max,\n",
    "        )\n",
    "        # Consider M loss at protein N-term\n",
    "        if sequence.startswith('M'):\n",
    "            for seq,miss,cterm in zip(\n",
    "                seq_list,miss_list,cterm_list\n",
    "            ):\n",
    "                if (\n",
    "                    sequence.startswith(seq) \n",
    "                    and len(seq)>self.peptide_length_min\n",
    "                ):\n",
    "                    seq_list.append(seq[1:])\n",
    "                    miss_list.append(miss)\n",
    "                    nterm_list.append(True)\n",
    "                    cterm_list.append(cterm)\n",
    "        return seq_list, miss_list, nterm_list, cterm_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "seq = 'MABCDEKHIJKLNOPQRST'\n",
    "digest = Digest()\n",
    "seq_list, miss_list, nterm_list, cterm_list = digest.cleave_sequence(seq)\n",
    "assert len(seq_list) == len(miss_list) == len(nterm_list) == len(cterm_list)\n",
    "M_start_seqs = [seq for seq in seq_list if seq.startswith('M')]\n",
    "assert len(M_start_seqs)*2 == len([_ for _ in nterm_list if _])\n",
    "assert np.all(nterm_list[-len(M_start_seqs):])\n",
    "T_end_seqs = [seq for seq in seq_list if seq.endswith('T')]\n",
    "assert len(T_end_seqs) == len([_ for _ in cterm_list if _])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def get_fix_mods(\n",
    "    sequence:str,\n",
    "    fix_mod_aas:str,\n",
    "    fix_mod_dict:dict\n",
    ")->tuple:\n",
    "    \"\"\"\n",
    "    Generate fix modifications for the sequence\n",
    "    \"\"\"\n",
    "    mods = []\n",
    "    mod_sites = []\n",
    "    for i,aa in enumerate(sequence):\n",
    "        if aa in fix_mod_aas:\n",
    "            mod_sites.append(i+1)\n",
    "            mods.append(fix_mod_dict[aa])\n",
    "    return ';'.join(mods), ';'.join(str(i) for i in mod_sites)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('mod@C;mod@C;mod@C', '2;4;6')"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq = 'ACBCDCK'\n",
    "_fix_mod_dict = {}\n",
    "_fix_mod_dict['C'] = 'mod@C'\n",
    "mods, mod_sites = get_fix_mods(seq, 'C', _fix_mod_dict)\n",
    "assert mods==';'.join(['mod@C']*3)\n",
    "assert mod_sites=='2;4;6'\n",
    "get_fix_mods(seq, 'C', _fix_mod_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def get_candidate_sites(\n",
    "    sequence:str, target_mod_aas:str\n",
    ")->list:\n",
    "    \"\"\"get candidate modification sites\n",
    "\n",
    "    Args:\n",
    "        sequence (str): peptide sequence\n",
    "        target_mod_aas (str): AAs that may have modifications\n",
    "\n",
    "    Returns:\n",
    "        list: candiadte mod sites in alphabase format (0: N-term, -1: C-term, 1-n:others)\n",
    "    \"\"\"\n",
    "    candidate_sites = []\n",
    "    for i,aa in enumerate(sequence):\n",
    "        if aa in target_mod_aas:\n",
    "            candidate_sites.append(i+1) #alphabase mod sites\n",
    "    return candidate_sites\n",
    "\n",
    "def get_var_mod_sites(\n",
    "    sequence:str,\n",
    "    target_mod_aas:str,\n",
    "    max_var_mod: int,\n",
    "    max_combs: int,\n",
    ")->list:\n",
    "    \"\"\"get all combinations of variable modification sites\n",
    "\n",
    "    Args:\n",
    "        sequence (str): peptide sequence\n",
    "        target_mod_aas (str): AAs that may have modifications\n",
    "        max_var_mod (int): max number of mods in a sequence\n",
    "        max_combs (int): max number of combinations for a sequence\n",
    "\n",
    "    Returns:\n",
    "        list: list of combinations of (tuple) modification sites \n",
    "    \"\"\"\n",
    "    candidate_sites = get_candidate_sites(\n",
    "        sequence, target_mod_aas\n",
    "    )\n",
    "    mod_sites = [(s,) for s in candidate_sites]\n",
    "    for n_var_mod in range(2, max_var_mod+1):\n",
    "        if len(mod_sites)>=max_combs: break\n",
    "        mod_sites.extend(\n",
    "            itertools.islice(\n",
    "                itertools.combinations(\n",
    "                    candidate_sites, n_var_mod\n",
    "                ),\n",
    "                max_combs-len(mod_sites)\n",
    "            )\n",
    "        )\n",
    "    return mod_sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(2,),\n",
       " (4,),\n",
       " (5,),\n",
       " (6,),\n",
       " (7,),\n",
       " (2, 4),\n",
       " (2, 5),\n",
       " (2, 6),\n",
       " (2, 7),\n",
       " (4, 5),\n",
       " (4, 6),\n",
       " (4, 7),\n",
       " (5, 6),\n",
       " (5, 7),\n",
       " (6, 7),\n",
       " (2, 4, 5),\n",
       " (2, 4, 6),\n",
       " (2, 4, 7),\n",
       " (2, 5, 6),\n",
       " (2, 5, 7)]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq = 'AMCMSTYK'\n",
    "candidate_sites = get_candidate_sites(seq, 'MSTY')\n",
    "assert np.all(np.array(candidate_sites)==np.array([2,4,5,6,7]))\n",
    "get_var_mod_sites(seq, 'MSTY', 3, 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import copy\n",
    "def get_var_mods_per_sites_multi_mods_on_aa(\n",
    "    sequence:str,\n",
    "    mod_sites:tuple,\n",
    "    var_mod_dict:dict\n",
    ")->list:\n",
    "    \"\"\"\n",
    "    Used only when the var mod list contains \n",
    "    more than one mods on the same AA, for example:\n",
    "    Mod1@A, Mod2@A ...\n",
    "    \"\"\"\n",
    "    mods_str_list = ['']\n",
    "    for i,site in enumerate(mod_sites):\n",
    "        if len(var_mod_dict[sequence[site-1]]) == 1:\n",
    "            for i in range(len(mods_str_list)):\n",
    "                mods_str_list[i] += var_mod_dict[sequence[site-1]][0]+';'\n",
    "        else:\n",
    "            _new_list = []\n",
    "            for mod in var_mod_dict[sequence[site-1]]:\n",
    "                _lst = copy.deepcopy(mods_str_list)\n",
    "                for i in range(len(_lst)):\n",
    "                    _lst[i] += mod+';'\n",
    "                _new_list.extend(_lst)\n",
    "            mods_str_list = _new_list\n",
    "    return [mod[:-1] for mod in mods_str_list]\n",
    "\n",
    "def get_var_mods_per_sites_single_mod_on_aa(\n",
    "    sequence:str,\n",
    "    mod_sites:tuple,\n",
    "    var_mod_dict:dict\n",
    ")->list:\n",
    "    \"\"\"\n",
    "    Used when the var mod list contains \n",
    "    only one mods on the each AA, for example:\n",
    "    Mod1@A, Mod2@D ...\n",
    "    \"\"\"\n",
    "    mod_str = ''\n",
    "    for site in mod_sites:\n",
    "            mod_str += var_mod_dict[sequence[site-1]]+';'\n",
    "    return [mod_str[:-1]]\n",
    "\n",
    "get_var_mods_per_sites = get_var_mods_per_sites_single_mod_on_aa\n",
    "\n",
    "def get_var_mods(\n",
    "    sequence:str,\n",
    "    var_mod_aas:str,\n",
    "    mod_dict:dict,\n",
    "    max_var_mod:int,\n",
    "    max_combs:int,\n",
    "    keep_unmodified:bool=False,\n",
    ")->tuple:\n",
    "    \"\"\"\n",
    "    Generate all modification combinations and associated sites\n",
    "    for the sequence.\n",
    "    \"\"\"\n",
    "    mod_sites_list = get_var_mod_sites(\n",
    "        sequence, var_mod_aas, \n",
    "        max_var_mod, max_combs\n",
    "    )\n",
    "    ret_mods = []\n",
    "    ret_sites_list = []\n",
    "    for mod_sites in mod_sites_list:\n",
    "        _mods = get_var_mods_per_sites(\n",
    "            sequence,mod_sites,mod_dict\n",
    "        )\n",
    "        mod_sites_str = ';'.join([str(i) for i in mod_sites])\n",
    "        ret_mods.extend(_mods)\n",
    "        ret_sites_list.extend([mod_sites_str]*len(_mods))\n",
    "    if keep_unmodified:\n",
    "        ret_mods.append('')\n",
    "        ret_sites_list.append('')\n",
    "    return ret_mods, ret_sites_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(['mod@M',\n",
       "  'mod@M',\n",
       "  'mod@S',\n",
       "  'modX@S',\n",
       "  'mod@T',\n",
       "  'mod@Y',\n",
       "  'mod@M;mod@M',\n",
       "  'mod@M;mod@S',\n",
       "  'mod@M;modX@S',\n",
       "  'mod@M;mod@T',\n",
       "  'mod@M;mod@Y',\n",
       "  'mod@M;mod@S',\n",
       "  'mod@M;modX@S',\n",
       "  'mod@M;mod@T',\n",
       "  'mod@M;mod@Y',\n",
       "  'mod@S;mod@T',\n",
       "  'modX@S;mod@T',\n",
       "  'mod@S;mod@Y',\n",
       "  'modX@S;mod@Y',\n",
       "  'mod@T;mod@Y',\n",
       "  'mod@M;mod@M;mod@S',\n",
       "  'mod@M;mod@M;modX@S'],\n",
       " ['2',\n",
       "  '4',\n",
       "  '5',\n",
       "  '5',\n",
       "  '6',\n",
       "  '7',\n",
       "  '2;4',\n",
       "  '2;5',\n",
       "  '2;5',\n",
       "  '2;6',\n",
       "  '2;7',\n",
       "  '4;5',\n",
       "  '4;5',\n",
       "  '4;6',\n",
       "  '4;7',\n",
       "  '5;6',\n",
       "  '5;6',\n",
       "  '5;7',\n",
       "  '5;7',\n",
       "  '6;7',\n",
       "  '2;4;5',\n",
       "  '2;4;5'])"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#hide\n",
    "get_var_mods_per_sites = get_var_mods_per_sites_multi_mods_on_aa\n",
    "seq = 'AMCMSTYK'\n",
    "candidate_sites = get_candidate_sites(seq, 'MSTY')\n",
    "mod_sites_list = get_var_mod_sites(seq, 'MSTY', 3, 20)\n",
    "_mod_dict = {\n",
    "    'M':['mod@M'],\n",
    "    'S':['mod@S','modX@S'],\n",
    "    'T':['mod@T'],\n",
    "    'Y':['mod@Y'],\n",
    "}\n",
    "_mods, _sites = get_var_mods(seq, 'MSTY', _mod_dict, 3, 16)\n",
    "\n",
    "assert 'mod@M;mod@M;mod@S' in _mods\n",
    "assert _sites[_mods.index('mod@M;mod@M;mod@S')] == '2;4;5'\n",
    "assert 'mod@M;mod@M;modX@S' in _mods\n",
    "assert _sites[_mods.index('mod@M;mod@M;modX@S')] == '2;4;5'\n",
    "assert 'mod@M;mod@S' in _mods\n",
    "assert 'mod@M;modX@S' in _mods\n",
    "get_var_mods(seq, 'MSTY', _mod_dict, 3, 16)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(['mod@M',\n",
       "  'mod@M',\n",
       "  'mod@S',\n",
       "  'mod@T',\n",
       "  'mod@Y',\n",
       "  'mod@M;mod@M',\n",
       "  'mod@M;mod@S',\n",
       "  'mod@M;mod@T',\n",
       "  'mod@M;mod@Y',\n",
       "  'mod@M;mod@S',\n",
       "  'mod@M;mod@T',\n",
       "  'mod@M;mod@Y',\n",
       "  'mod@S;mod@T',\n",
       "  'mod@S;mod@Y',\n",
       "  'mod@T;mod@Y',\n",
       "  'mod@M;mod@M;mod@S'],\n",
       " ['2',\n",
       "  '4',\n",
       "  '5',\n",
       "  '6',\n",
       "  '7',\n",
       "  '2;4',\n",
       "  '2;5',\n",
       "  '2;6',\n",
       "  '2;7',\n",
       "  '4;5',\n",
       "  '4;6',\n",
       "  '4;7',\n",
       "  '5;6',\n",
       "  '5;7',\n",
       "  '6;7',\n",
       "  '2;4;5'])"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#hide\n",
    "get_var_mods_per_sites = get_var_mods_per_sites_single_mod_on_aa\n",
    "seq = 'AMCMSTYK'\n",
    "candidate_sites = get_candidate_sites(seq, 'MSTY')\n",
    "mod_sites_list = get_var_mod_sites(seq, 'MSTY', 3, 20)\n",
    "_mod_dict = {\n",
    "    'M':'mod@M',\n",
    "    'S':'mod@S',\n",
    "    'T':'mod@T',\n",
    "    'Y':'mod@Y',\n",
    "}\n",
    "_mods, _sites = get_var_mods(seq, 'MSTY', _mod_dict, 3, 16)\n",
    "assert len(_mods) == len(_sites) == 16\n",
    "assert _sites[_mods.index('mod@M;mod@M;mod@S')] == '2;4;5'\n",
    "get_var_mods(seq, 'MSTY', _mod_dict, 3, 16)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def parse_term_mod(term_mod_name:str):\n",
    "    _mod, term = term_mod_name.split('@')\n",
    "    if '^' in term:\n",
    "        return tuple(term.split('^'))\n",
    "    else:\n",
    "        return '', term"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "assert parse_term_mod('Acetyl@Protein N-term') == ('', 'Protein N-term')\n",
    "assert parse_term_mod('Gln->pyro-Glu@Q^Any N-term') == ('Q', 'Any N-term')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def add_single_peptide_labeling(\n",
    "    seq:str,\n",
    "    mods:str,\n",
    "    mod_sites:str, \n",
    "    label_aas:str, \n",
    "    label_mod_dict:dict, \n",
    "    nterm_label_mod:str, \n",
    "    cterm_label_mod:str\n",
    "):\n",
    "    add_nterm_label = True if nterm_label_mod else False\n",
    "    add_cterm_label = True if cterm_label_mod else False\n",
    "    if mod_sites:\n",
    "        _sites = mod_sites.split(';')\n",
    "        if '0' in _sites: add_nterm_label = False\n",
    "        if '-1' in _sites: add_cterm_label = False\n",
    "        mod_list = [mods]\n",
    "        mod_site_list = [mod_sites]\n",
    "    else:\n",
    "        mod_list = []\n",
    "        mod_site_list = []\n",
    "    if add_nterm_label:\n",
    "        mod_list.append(nterm_label_mod)\n",
    "        mod_site_list.append('0')\n",
    "    if add_cterm_label:\n",
    "        mod_list.append(cterm_label_mod)\n",
    "        mod_site_list.append('-1')\n",
    "    aa_labels, aa_label_sites = get_fix_mods(seq, label_aas, label_mod_dict)\n",
    "    if aa_labels:\n",
    "        mod_list.append(aa_labels)\n",
    "        mod_site_list.append(aa_label_sites)\n",
    "    return ';'.join(mod_list), ';'.join(mod_site_list)\n",
    "\n",
    "def parse_labels(labels:list):\n",
    "    label_aas = ''\n",
    "    label_mod_dict = {}\n",
    "    nterm_label_mod = ''\n",
    "    cterm_label_mod = ''\n",
    "    for label in labels:\n",
    "        _, aa = label.split('@')\n",
    "        if len(aa) == 1:\n",
    "            label_aas += aa\n",
    "            label_mod_dict[aa] = label\n",
    "        elif aa == 'Any N-term':\n",
    "            nterm_label_mod = label\n",
    "        elif aa == 'Any C-term':\n",
    "            cterm_label_mod = label\n",
    "    return label_aas, label_mod_dict, nterm_label_mod, cterm_label_mod\n",
    "        \n",
    "def create_labeling_peptide_df(peptide_df:pd.DataFrame, labels:list):\n",
    "    df = peptide_df.copy()\n",
    "    (\n",
    "        label_aas, label_mod_dict, \n",
    "        nterm_label_mod, cterm_label_mod\n",
    "    ) = parse_labels(labels)\n",
    "\n",
    "    (\n",
    "        df['mods'],\n",
    "        df['mod_sites']\n",
    "    ) = zip(*df[\n",
    "        ['sequence','mods','mod_sites']\n",
    "    ].apply(lambda x:\n",
    "        add_single_peptide_labeling(\n",
    "            *x, label_aas, label_mod_dict, \n",
    "            nterm_label_mod, cterm_label_mod\n",
    "        ), axis=1,\n",
    "    ))\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "labels = ['label@Any N-term','label@K']\n",
    "(\n",
    "    label_aas, label_mod_dict, \n",
    "    nterm_label_mod, cterm_label_mod\n",
    ") = parse_labels(labels)\n",
    "assert add_single_peptide_labeling(\n",
    "    'ABCK','','', label_aas, label_mod_dict, \n",
    "    nterm_label_mod, cterm_label_mod\n",
    ") == ('label@Any N-term;label@K', '0;4')\n",
    "assert add_single_peptide_labeling(\n",
    "    'ABCK','Mod@Any N-term','0', label_aas, label_mod_dict, \n",
    "    nterm_label_mod, cterm_label_mod\n",
    ") == ('Mod@Any N-term;label@K', '0;4')\n",
    "assert add_single_peptide_labeling(\n",
    "    'KBCK','','', label_aas, label_mod_dict, \n",
    "    nterm_label_mod, cterm_label_mod\n",
    ") == ('label@Any N-term;label@K;label@K', '0;1;4')\n",
    "assert add_single_peptide_labeling(\n",
    "    'KBCK','Mod@Any N-term','0', label_aas, label_mod_dict, \n",
    "    nterm_label_mod, cterm_label_mod\n",
    ") == ('Mod@Any N-term;label@K;label@K', '0;1;4')\n",
    "pep_df = pd.DataFrame({\n",
    "    'sequence': ['ABCD','ABCK','KABK','EFGK'],\n",
    "    'mods': ['']*3+['Mod@Any N-term'],\n",
    "    'mod_sites': ['']*3+['0']\n",
    "})\n",
    "df = create_labeling_peptide_df(pep_df, labels)\n",
    "assert np.all(df.mods.values!=pep_df.mods.values)\n",
    "assert df[df.sequence=='ABCD'].mods.values[0] == 'label@Any N-term'\n",
    "assert df[df.sequence=='ABCD'].mod_sites.values[0] == '0'\n",
    "assert df[df.sequence=='ABCK'].mods.values[0] == 'label@Any N-term;label@K'\n",
    "assert df[df.sequence=='ABCK'].mod_sites.values[0] == '0;4'\n",
    "assert df[df.sequence=='KABK'].mods.values[0] == 'label@Any N-term;label@K;label@K'\n",
    "assert df[df.sequence=='KABK'].mod_sites.values[0] == '0;1;4'\n",
    "assert df[df.sequence=='EFGK'].mods.values[0] == 'Mod@Any N-term;label@K'\n",
    "assert df[df.sequence=='EFGK'].mod_sites.values[0] == '0;4'\n",
    "df = create_labeling_peptide_df(pep_df, [])\n",
    "assert np.all(df.mods.values==pep_df.mods.values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def protein_idxes_to_names(protein_idxes:str, protein_names:list):\n",
    "    if len(protein_idxes) == 0: return ''\n",
    "    proteins = [protein_names[int(i)] for i in protein_idxes.split(';')]\n",
    "    proteins = [protein for protein in proteins if protein]\n",
    "    return ';'.join(proteins)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "assert protein_idxes_to_names('0;1', ['A','','B'])=='A'\n",
    "assert protein_idxes_to_names('0;1', ['A','C','B'])=='A;C'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def append_regular_modifications(df:pd.DataFrame, \n",
    "    var_mods = ['Phospho@S','Phospho@T','Phospho@Y'], \n",
    "    max_mod_num=1, max_combs=100,\n",
    "    keep_unmodified=True,\n",
    ")->pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Append regular (not N/C-term) variable modifications to the \n",
    "    exsiting modifications of each sequence in `df`.\n",
    "\n",
    "    Args:\n",
    "        df (pd.DataFrame): Precursor dataframe\n",
    "        var_mods (list, optional): Considered varialbe modification list. \n",
    "            Defaults to ['Phospho@S','Phospho@T','Phospho@Y'].\n",
    "        max_mod_num (int, optional): Maximal modification number for \n",
    "            each sequence of the `var_mods`. Defaults to 1.\n",
    "        max_combs (int, optional): One sequence is only allowed to explode \n",
    "            to `max_combs` number of modified peptides. Defaults to 100.\n",
    "        keep_unmodified (bool, optional): If unmodified (only refered to `var_mods`)\n",
    "        peptides are also remained in the returned dataframe. Defaults to True.\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame: The precursor_df with `var_mods` appended.\n",
    "    \"\"\"\n",
    "    mod_dict = dict([(mod[-1],mod) for mod in var_mods])\n",
    "    var_mod_aas = ''.join(mod_dict.keys())\n",
    "    \n",
    "    (\n",
    "        df['mods_app'],\n",
    "        df['mod_sites_app']\n",
    "    ) = zip(*df.sequence.apply(get_var_mods, \n",
    "            var_mod_aas=var_mod_aas, mod_dict=mod_dict, \n",
    "            max_var_mod=max_mod_num, max_combs=max_combs,\n",
    "            keep_unmodified=keep_unmodified\n",
    "        )\n",
    "    )\n",
    "    \n",
    "    if keep_unmodified:\n",
    "        df = df.explode(['mods_app','mod_sites_app'])\n",
    "        df.fillna('', inplace=True)\n",
    "    else:\n",
    "        df.drop(df[df.mods_app.apply(lambda x: len(x)==0)].index, inplace=True)\n",
    "        df = df.explode(['mods_app','mod_sites_app'])\n",
    "    df['mods'] = df[['mods','mods_app']].apply(\n",
    "        lambda x: ';'.join(i for i in x if i), axis=1\n",
    "    )\n",
    "    df['mod_sites'] = df[['mod_sites','mod_sites_app']].apply(\n",
    "        lambda x: ';'.join(i for i in x if i), axis=1\n",
    "    )\n",
    "    df.drop(columns=['mods_app', 'mod_sites_app'], inplace=True)\n",
    "    df.reset_index(drop=True, inplace=True)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "from alphabase.spectral_library.library_base import SpecLibBase\n",
    "\n",
    "class FastaLib(SpecLibBase):\n",
    "    def __init__(self,\n",
    "        charged_frag_types:list = [\n",
    "            'b_z1','b_z2','y_z1', 'y_z2'\n",
    "        ],\n",
    "        protease:str = 'trypsin',\n",
    "        max_missed_cleavages:int = 2,\n",
    "        peptide_length_min:int = 7,\n",
    "        peptide_length_max:int = 35,\n",
    "        precursor_charge_min:int = 2,\n",
    "        precursor_charge_max:int = 4,\n",
    "        precursor_mz_min:float = 200.0, \n",
    "        precursor_mz_max:float = 2000.0,\n",
    "        var_mods:list = ['Acetyl@Protein N-term','Oxidation@M'],\n",
    "        max_var_mod_num:int = 2,\n",
    "        fix_mods:list = ['Carbamidomethyl@C'],\n",
    "        decoy: str = None, # or pseudo_reverse or diann\n",
    "        I_to_L=False,\n",
    "    ):\n",
    "        super().__init__(\n",
    "            charged_frag_types=charged_frag_types,\n",
    "            precursor_mz_min=precursor_mz_min,\n",
    "            precursor_mz_max=precursor_mz_max,\n",
    "            decoy=decoy\n",
    "        )\n",
    "        self.protein_df:pd.DataFrame() = pd.DataFrame()\n",
    "        self.I_to_L = I_to_L\n",
    "        self.max_mod_combinations = 100\n",
    "        self._digest = Digest(\n",
    "            protease, max_missed_cleavages,\n",
    "            peptide_length_min, peptide_length_max\n",
    "        )\n",
    "        self.min_precursor_charge = precursor_charge_min\n",
    "        self.max_precursor_charge = precursor_charge_max\n",
    "\n",
    "        self.var_mods = var_mods\n",
    "        self.fix_mods = fix_mods\n",
    "        self.max_var_mod_num = max_var_mod_num\n",
    "\n",
    "        self.fix_mod_aas = ''\n",
    "        self.fix_mod_prot_nterm_dict = {}\n",
    "        self.fix_mod_prot_cterm_dict = {}\n",
    "        self.fix_mod_pep_nterm_dict = {}\n",
    "        self.fix_mod_pep_cterm_dict = {}\n",
    "        self.fix_mod_dict = {}\n",
    "\n",
    "        def _set_term_mod(term_mod,\n",
    "            prot_nterm, prot_cterm, pep_nterm, pep_cterm,\n",
    "            allow_conflicts\n",
    "        ):\n",
    "            def _set_dict(term_dict,site,mod,\n",
    "                allow_conflicts\n",
    "            ):\n",
    "                if allow_conflicts:\n",
    "                    if site in term_dict:\n",
    "                        term_dict[site].append(term_mod)\n",
    "                    else:\n",
    "                        term_dict[site] = [term_mod]\n",
    "                else:\n",
    "                    term_dict[site] = term_mod\n",
    "            site, term = parse_term_mod(term_mod)\n",
    "            if term == \"Any N-term\":\n",
    "                _set_dict(pep_nterm, site, term_mod, \n",
    "                    allow_conflicts\n",
    "                )\n",
    "            elif term == 'Protein N-term':\n",
    "                _set_dict(prot_nterm, site, term_mod, \n",
    "                    allow_conflicts\n",
    "                )\n",
    "            elif term == 'Any C-term':\n",
    "                _set_dict(pep_cterm, site, term_mod, \n",
    "                    allow_conflicts\n",
    "                )\n",
    "            elif term == 'Protein C-term':\n",
    "                _set_dict(prot_cterm, site, term_mod, \n",
    "                    allow_conflicts\n",
    "                )\n",
    "        \n",
    "        for mod in fix_mods:\n",
    "            if mod.find('@')+2 == len(mod):\n",
    "                self.fix_mod_aas += mod[-1]\n",
    "                self.fix_mod_dict[mod[-1]] = mod\n",
    "            else:\n",
    "                _set_term_mod(\n",
    "                    mod, \n",
    "                    self.fix_mod_prot_nterm_dict,\n",
    "                    self.fix_mod_prot_cterm_dict,\n",
    "                    self.fix_mod_pep_nterm_dict,\n",
    "                    self.fix_mod_pep_cterm_dict,\n",
    "                    allow_conflicts=False\n",
    "                )\n",
    "\n",
    "        self.var_mod_aas = ''\n",
    "        self.var_mod_prot_nterm_dict = {}\n",
    "        self.var_mod_prot_cterm_dict = {}\n",
    "        self.var_mod_pep_nterm_dict = {}\n",
    "        self.var_mod_pep_cterm_dict = {}\n",
    "        self.var_mod_dict = {}\n",
    "\n",
    "        global get_var_mods_per_sites\n",
    "        if self._check_if_multi_mods_on_aa(var_mods):\n",
    "            for mod in var_mods:\n",
    "                if mod.find('@')+2 == len(mod):\n",
    "                    if mod[-1] in self.fix_mod_dict: continue\n",
    "                    self.var_mod_aas += mod[-1]\n",
    "                    if mod[-1] in self.var_mod_dict:\n",
    "                        self.var_mod_dict[mod[-1]].append(mod)\n",
    "                    else:\n",
    "                        self.var_mod_dict[mod[-1]] = [mod]\n",
    "            get_var_mods_per_sites = get_var_mods_per_sites_multi_mods_on_aa\n",
    "        else:\n",
    "            for mod in var_mods:\n",
    "                if mod.find('@')+2 == len(mod):\n",
    "                    if mod[-1] in self.fix_mod_dict: continue\n",
    "                    self.var_mod_aas += mod[-1]\n",
    "                    self.var_mod_dict[mod[-1]] = mod\n",
    "            get_var_mods_per_sites = get_var_mods_per_sites_single_mod_on_aa\n",
    "        \n",
    "        for mod in var_mods:\n",
    "            if mod.find('@')+2 < len(mod):\n",
    "                _set_term_mod(\n",
    "                    mod, \n",
    "                    self.var_mod_prot_nterm_dict,\n",
    "                    self.var_mod_prot_cterm_dict,\n",
    "                    self.var_mod_pep_nterm_dict,\n",
    "                    self.var_mod_pep_cterm_dict,\n",
    "                    allow_conflicts=True\n",
    "                )\n",
    "\n",
    "    def _check_if_multi_mods_on_aa(self, var_mods):\n",
    "        mod_set = set()\n",
    "        for mod in var_mods:\n",
    "            if mod.find('@')+2 == len(mod):\n",
    "                if mod[-1] in mod_set: return True\n",
    "                mod_set.add(mod[-1])\n",
    "        return False\n",
    "\n",
    "    def import_and_process_fasta(self, fasta_files:Union[str,list]):\n",
    "        protein_dict = load_all_proteins(fasta_files)\n",
    "        self.import_and_process_protein_dict(protein_dict)\n",
    "\n",
    "    def import_and_process_protein_dict(self, protein_dict:dict):\n",
    "        self.get_peptides_from_protein_dict(protein_dict)\n",
    "        self._process_after_load_pep_seqs()\n",
    "\n",
    "    def import_and_process_peptide_sequences(self, \n",
    "        pep_seq_list:list, protein_list=None,\n",
    "    ):\n",
    "        self.get_peptides_from_peptide_sequence_list(\n",
    "            pep_seq_list, protein_list\n",
    "        )\n",
    "        self._process_after_load_pep_seqs()\n",
    "\n",
    "    def _process_after_load_pep_seqs(self):\n",
    "        \"\"\"\n",
    "        Called by `import_and_process_...` methods. \n",
    "        \"\"\"\n",
    "        self.append_decoy_sequence()\n",
    "        self.add_modifications()\n",
    "        self.add_charge()\n",
    "\n",
    "    def get_peptides_from_fasta(self, fasta_file:Union[str,list]):\n",
    "        \"\"\"Load peptide sequence from fasta file.\n",
    "\n",
    "        Args:\n",
    "            fasta_path (Union[str,list]): could be a fasta path or a list of fasta paths\n",
    "              or a list of fasta paths\n",
    "        \"\"\"\n",
    "        if isinstance(fasta_file, str):\n",
    "            self.get_peptides_from_fasta_list([fasta_file])\n",
    "        else:\n",
    "            self.get_peptides_from_fasta_list(fasta_file)\n",
    "\n",
    "    def get_peptides_from_fasta_list(self, fasta_files:list):\n",
    "        \"\"\"Load peptide sequences from fasta file list\n",
    "\n",
    "        Args:\n",
    "            fasta_files (list): fasta file list\n",
    "        \"\"\"\n",
    "        protein_dict = load_all_proteins(fasta_files)\n",
    "        self.get_peptides_from_protein_dict(protein_dict)\n",
    "\n",
    "    def get_peptides_from_protein_dict(self, protein_dict:dict):\n",
    "\n",
    "        self.protein_df = pd.DataFrame.from_dict(\n",
    "            protein_dict, orient='index'\n",
    "        ).reset_index(drop=True)\n",
    "\n",
    "        if self.I_to_L:\n",
    "            self.protein_df[\n",
    "                'sequence_I2L'\n",
    "            ] = self.protein_df.sequence.str.replace('I','L')\n",
    "            digest_seq = 'sequence_I2L'\n",
    "        else:\n",
    "            digest_seq = 'sequence'\n",
    "        self._cleave_to_peptides(\n",
    "            self.protein_df,\n",
    "            protein_seq_column=digest_seq\n",
    "        )\n",
    "\n",
    "    def _cleave_to_peptides(self, \n",
    "        protein_df:pd.DataFrame,\n",
    "        protein_seq_column:str='sequence'\n",
    "    ):\n",
    "        pep_dict = {}\n",
    "\n",
    "        for i,prot_seq in enumerate(\n",
    "            protein_df[protein_seq_column].values\n",
    "        ):\n",
    "            (\n",
    "                seq_list, miss_list, nterm_list, cterm_list\n",
    "            ) = self._digest.cleave_sequence(prot_seq)\n",
    "            for seq,miss,nterm,cterm in zip(\n",
    "                seq_list,miss_list,nterm_list, cterm_list\n",
    "            ):\n",
    "                prot_id = str(i)\n",
    "                if seq in pep_dict:\n",
    "                    if not pep_dict[seq][0].endswith(prot_id):\n",
    "                        pep_dict[seq][0] += ';'+prot_id\n",
    "                    if nterm:\n",
    "                        pep_dict[seq][2] = nterm\n",
    "                    if cterm:\n",
    "                        pep_dict[seq][3] = cterm\n",
    "                else:\n",
    "                    pep_dict[seq] = [prot_id,miss,nterm,cterm]\n",
    "        self._precursor_df = pd.DataFrame().from_dict(\n",
    "            pep_dict, orient='index', columns = [\n",
    "                'protein_idxes','miss_cleavage',\n",
    "                'is_prot_nterm','is_prot_cterm'\n",
    "            ]\n",
    "        )\n",
    "        self._precursor_df.reset_index(drop=False, inplace=True)\n",
    "        self._precursor_df.rename(\n",
    "            columns={'index':'sequence'}, inplace=True\n",
    "        )\n",
    "        self._precursor_df['mods'] = ''\n",
    "        self._precursor_df['mod_sites'] = ''\n",
    "        self.refine_df()\n",
    "\n",
    "    def append_protein_name(self):\n",
    "        if (\n",
    "            'protein_id' not in self.protein_df or \n",
    "            'protein_idxes' not in self._precursor_df\n",
    "        ): \n",
    "            return\n",
    "\n",
    "        self._precursor_df['proteins'] = self._precursor_df['protein_idxes'].apply(\n",
    "            protein_idxes_to_names,\n",
    "            protein_names=self.protein_df['protein_id'].values\n",
    "        )\n",
    "\n",
    "        if 'gene_name' in self.protein_df.columns:\n",
    "            self._precursor_df['genes'] = self._precursor_df['protein_idxes'].apply(\n",
    "                protein_idxes_to_names,\n",
    "                protein_names=self.protein_df['gene_name'].values\n",
    "            )\n",
    "\n",
    "    def get_peptides_from_peptide_sequence_list(self, \n",
    "        pep_seq_list:list,\n",
    "        protein_list:list = None\n",
    "    ):\n",
    "        self._precursor_df = pd.DataFrame()\n",
    "        self._precursor_df['sequence'] = pep_seq_list\n",
    "        if protein_list is not None:\n",
    "            self._precursor_df['protein_name'] = protein_list\n",
    "        self._precursor_df['is_prot_nterm'] = False\n",
    "        self._precursor_df['is_prot_cterm'] = False\n",
    "        self.refine_df()\n",
    "\n",
    "    def add_mods_for_one_seq(self, sequence:str, \n",
    "        is_prot_nterm, is_prot_cterm\n",
    "    ):\n",
    "        fix_mods, fix_mod_sites = get_fix_mods(\n",
    "            sequence, self.fix_mod_aas, self.fix_mod_dict\n",
    "        )\n",
    "        #TODO add prot and pep C-term fix mods\n",
    "        #TODO add prot and pep N-term fix mods\n",
    "\n",
    "        if len(fix_mods) == 0:\n",
    "            fix_mods = ['']\n",
    "            fix_mod_sites = ['']\n",
    "        else:\n",
    "            fix_mods = [fix_mods]\n",
    "            fix_mod_sites = [fix_mod_sites]\n",
    "\n",
    "        var_mods_list, var_mod_sites_list = get_var_mods(\n",
    "            sequence, self.var_mod_aas, self.var_mod_dict, \n",
    "            self.max_var_mod_num, self.max_mod_combinations-1, # 1 for unmodified\n",
    "            keep_unmodified=True\n",
    "        )\n",
    "\n",
    "        nterm_var_mods = ['']\n",
    "        nterm_var_mod_sites = ['']\n",
    "        if is_prot_nterm and len(self.var_mod_prot_nterm_dict)>0:\n",
    "            if '' in self.var_mod_prot_nterm_dict:\n",
    "                nterm_var_mods.extend(self.var_mod_prot_nterm_dict[''])\n",
    "            if sequence[0] in self.var_mod_prot_nterm_dict:\n",
    "                nterm_var_mods.extend(self.var_mod_prot_nterm_dict[sequence[0]])\n",
    "        if len(self.var_mod_pep_nterm_dict)>0:\n",
    "            if '' in self.var_mod_pep_nterm_dict:\n",
    "                nterm_var_mods.extend(self.var_mod_pep_nterm_dict[''])\n",
    "            if sequence[0] in self.var_mod_pep_nterm_dict:\n",
    "                nterm_var_mods.extend(self.var_mod_pep_nterm_dict[sequence[0]])\n",
    "        nterm_var_mod_sites.extend(['0']*(len(nterm_var_mods)-1))\n",
    "\n",
    "        #TODO add prot and pep C-term var mods\n",
    "\n",
    "        return (\n",
    "            list(\n",
    "                ';'.join([i for i in items if i]) for items in itertools.product(\n",
    "                    fix_mods, nterm_var_mods, var_mods_list\n",
    "                )\n",
    "            ),\n",
    "            list(\n",
    "                ';'.join([i for i in items if i]) for items in itertools.product(\n",
    "                    fix_mod_sites, nterm_var_mod_sites, var_mod_sites_list\n",
    "                )\n",
    "            ),\n",
    "        )\n",
    "\n",
    "    def add_modifications(self):\n",
    "        if 'is_prot_nterm' not in self._precursor_df.columns:\n",
    "            self._precursor_df['is_prot_nterm'] = False\n",
    "        if 'is_prot_cterm' not in self._precursor_df.columns:\n",
    "            self._precursor_df['is_prot_cterm'] = False\n",
    "        \n",
    "        (\n",
    "            self._precursor_df['mods'],\n",
    "            self._precursor_df['mod_sites']\n",
    "        ) = zip(*self._precursor_df[\n",
    "            ['sequence','is_prot_nterm','is_prot_cterm']\n",
    "        ].apply(lambda x:\n",
    "            self.add_mods_for_one_seq(*x), axis=1\n",
    "        ))\n",
    "        self._precursor_df = explode_multiple_columns(\n",
    "            self._precursor_df,\n",
    "            ['mods','mod_sites']\n",
    "        )\n",
    "        self._precursor_df.reset_index(drop=True, inplace=True)\n",
    "\n",
    "    def add_additional_modifications(self,\n",
    "        var_mods = ['Phospho@S','Phospho@T','Phospho@Y'], \n",
    "        max_mod_num=1, max_combs=100,\n",
    "        keep_unmodified=True,\n",
    "    ):\n",
    "        self._precursor_df = append_regular_modifications(\n",
    "            self.precursor_df,\n",
    "            var_mods=var_mods,\n",
    "            max_mod_num=max_mod_num,\n",
    "            max_combs=max_combs,\n",
    "            keep_unmodified=keep_unmodified\n",
    "        )\n",
    "\n",
    "    def add_peptide_labeling(self, labeling_channel_dict:dict):\n",
    "        \"\"\" \n",
    "        Add labeling onto peptides inplace of self._precursor_df\n",
    "\n",
    "        Args:\n",
    "            labeling_channel_dict (dict of list): for example:\n",
    "              {\n",
    "                  'reference': [], # not labelled for reference\n",
    "                  'light': ['Dimethyl@Any N-term','Dimethyl@K'],\n",
    "                  'median': ['Dimethyl:2H(4)@Any N-term','Dimethyl:2H(4)@K'],\n",
    "                  'heavy': ['Dimethyl:2H(6)13C(2)@Any N-term','Dimethyl:2H(6)13C(2)@K'],\n",
    "              }.\n",
    "              The key name could be arbitrary distinguished strings, and value must be a list of string.\n",
    "        \n",
    "        \"\"\"\n",
    "        df_list = []\n",
    "        for channel, labels in labeling_channel_dict.items():\n",
    "            df = create_labeling_peptide_df(self._precursor_df, labels)\n",
    "            df['label_channel'] = channel\n",
    "            df_list.append(df)\n",
    "        self._precursor_df = pd.concat(df_list)\n",
    "        self._precursor_df.reset_index(drop=True, inplace=True)\n",
    "\n",
    "    def add_charge(self):\n",
    "        self._precursor_df['charge'] = [\n",
    "            np.arange(\n",
    "                self.min_precursor_charge, \n",
    "                self.max_precursor_charge+1\n",
    "            )\n",
    "        ]*len(self._precursor_df)\n",
    "        self._precursor_df = self._precursor_df.explode('charge')\n",
    "        self._precursor_df['charge'] = self._precursor_df.charge.astype(np.int8)\n",
    "        self._precursor_df.reset_index(drop=True, inplace=True)\n",
    "\n",
    "    def save_hdf(self, hdf_file):\n",
    "        super().save_hdf(hdf_file)\n",
    "        _hdf = HDF_File(\n",
    "            hdf_file,\n",
    "            read_only=False,\n",
    "            truncate=True,\n",
    "            delete_existing=False\n",
    "        )\n",
    "        _hdf.library.protein_df = self.protein_df\n",
    "\n",
    "    def load_hdf(self, hdf_file, load_mod_seq=False):\n",
    "        super().load_hdf(hdf_file, load_mod_seq=load_mod_seq)\n",
    "        try:\n",
    "            _hdf = HDF_File(\n",
    "                hdf_file,\n",
    "            )\n",
    "            self.protein_df = _hdf.library.protein_df.values\n",
    "        except (AttributeError, KeyError, ValueError, TypeError):\n",
    "            print(f\"No protein_df in {hdf_file}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>mods</th>\n",
       "      <th>mod_sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABSTY</td>\n",
       "      <td>Phospho@S</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABSTY</td>\n",
       "      <td>Phospho@T</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABSTY</td>\n",
       "      <td>Phospho@Y</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABSTY</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACXSX</td>\n",
       "      <td>Acetyl@Protein N-term;Phospho@S</td>\n",
       "      <td>0;4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ACXSX</td>\n",
       "      <td>Acetyl@Protein N-term</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ACDEFG</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sequence                             mods mod_sites\n",
       "0    ABSTY                        Phospho@S         3\n",
       "1    ABSTY                        Phospho@T         4\n",
       "2    ABSTY                        Phospho@Y         5\n",
       "3    ABSTY                                           \n",
       "4    ACXSX  Acetyl@Protein N-term;Phospho@S       0;4\n",
       "5    ACXSX            Acetyl@Protein N-term         0\n",
       "6   ACDEFG                                           "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame(\n",
    "    {\n",
    "        'sequence': ['ABSTY','ACXSX','ACDEFG'],\n",
    "        'mods': ['', 'Acetyl@Protein N-term', ''],\n",
    "        'mod_sites': ['', '0', '']\n",
    "    }\n",
    ")\n",
    "df = append_regular_modifications(df, keep_unmodified=True)\n",
    "assert np.sum(df.sequence=='ABSTY')==4\n",
    "assert np.sum(df.sequence=='ACXSX')==2\n",
    "assert np.sum(df.sequence=='ACDEFG')==1\n",
    "assert all(df[df.sequence=='ABSTY'].mods.values == np.array(['Phospho@S','Phospho@T','Phospho@Y','']))\n",
    "assert all(df[df.sequence=='ABSTY'].mod_sites.values == np.array(['3','4','5','']))\n",
    "assert all(df[df.sequence=='ACXSX'].mods.values == np.array(['Acetyl@Protein N-term;Phospho@S','Acetyl@Protein N-term']))\n",
    "assert all(df[df.sequence=='ACXSX'].mod_sites.values == np.array(['0;4','0']))\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>mods</th>\n",
       "      <th>mod_sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABSTY</td>\n",
       "      <td>Phospho@S</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABSTY</td>\n",
       "      <td>Phospho@T</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABSTY</td>\n",
       "      <td>Phospho@Y</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACXSX</td>\n",
       "      <td>Acetyl@Protein N-term;Phospho@S</td>\n",
       "      <td>0;4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sequence                             mods mod_sites\n",
       "0    ABSTY                        Phospho@S         3\n",
       "1    ABSTY                        Phospho@T         4\n",
       "2    ABSTY                        Phospho@Y         5\n",
       "3    ACXSX  Acetyl@Protein N-term;Phospho@S       0;4"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame(\n",
    "    {\n",
    "        'sequence': ['ABSTY','ACXSX','ACDEFG'],\n",
    "        'mods': ['', 'Acetyl@Protein N-term', ''],\n",
    "        'mod_sites': ['', '0', '']\n",
    "    }\n",
    ")\n",
    "df = append_regular_modifications(df, keep_unmodified=False)\n",
    "assert np.sum(df.sequence=='ABSTY')==3\n",
    "assert np.sum(df.sequence=='ACXSX')==1\n",
    "assert np.sum(df.sequence=='ACDEFG')==0\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>protein_idxes</th>\n",
       "      <th>miss_cleavage</th>\n",
       "      <th>is_prot_nterm</th>\n",
       "      <th>is_prot_cterm</th>\n",
       "      <th>mods</th>\n",
       "      <th>mod_sites</th>\n",
       "      <th>nAA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AFGHIJK</td>\n",
       "      <td>0;1</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LMNOPQR</td>\n",
       "      <td>0;1</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABCDESTK</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MABCDESTK</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AFGHIJKLMNOPQR</td>\n",
       "      <td>0;1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>LMNOPQRAFGHIJK</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ABCDESTKAFGHIJK</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MABCDESTKAFGHIJK</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AFGHIJKLMNOPQRAFGHIJK</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ABCDESTKAFGHIJKLMNOPQR</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>MABCDESTKAFGHIJKLMNOPQR</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   sequence protein_idxes  miss_cleavage  is_prot_nterm  \\\n",
       "0                   AFGHIJK           0;1              0           True   \n",
       "1                   LMNOPQR           0;1              0          False   \n",
       "2                  ABCDESTK             0              0           True   \n",
       "3                 MABCDESTK             0              0           True   \n",
       "4            AFGHIJKLMNOPQR           0;1              1           True   \n",
       "5            LMNOPQRAFGHIJK             0              1          False   \n",
       "6           ABCDESTKAFGHIJK             0              1           True   \n",
       "7          MABCDESTKAFGHIJK             0              1           True   \n",
       "8     AFGHIJKLMNOPQRAFGHIJK             0              2          False   \n",
       "9    ABCDESTKAFGHIJKLMNOPQR             0              2           True   \n",
       "10  MABCDESTKAFGHIJKLMNOPQR             0              2           True   \n",
       "\n",
       "    is_prot_cterm mods mod_sites  nAA  \n",
       "0            True                   7  \n",
       "1            True                   7  \n",
       "2           False                   8  \n",
       "3           False                   9  \n",
       "4            True                  14  \n",
       "5            True                  14  \n",
       "6           False                  15  \n",
       "7           False                  16  \n",
       "8            True                  21  \n",
       "9           False                  22  \n",
       "10          False                  23  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_lib = FastaLib(None, I_to_L=False, decoy='pseudo_reverse')\n",
    "prot1 = 'MABCDESTKAFGHIJKLMNOPQRAFGHIJK'\n",
    "prot2 = 'AFGHIJKLMNOPQR'\n",
    "protein_dict = {\n",
    "    'xx': {\n",
    "        'protein_id': 'xx',\n",
    "        'gene_name': '',\n",
    "        'sequence': prot1\n",
    "    },\n",
    "    'yy': {\n",
    "        'protein_id': 'yy',\n",
    "        'gene_name': 'gene',\n",
    "        'sequence': prot2\n",
    "    }\n",
    "}\n",
    "_lib.get_peptides_from_protein_dict(protein_dict)\n",
    "_lib.precursor_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "      <td>MABCDESTKAFGHIJKLMNOPQRAFGHIJK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>yy</td>\n",
       "      <td>gene</td>\n",
       "      <td>AFGHIJKLMNOPQR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  protein_id gene_name                        sequence\n",
       "0         xx            MABCDESTKAFGHIJKLMNOPQRAFGHIJK\n",
       "1         yy      gene                  AFGHIJKLMNOPQR"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_lib.protein_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>protein_idxes</th>\n",
       "      <th>miss_cleavage</th>\n",
       "      <th>is_prot_nterm</th>\n",
       "      <th>is_prot_cterm</th>\n",
       "      <th>mods</th>\n",
       "      <th>mod_sites</th>\n",
       "      <th>nAA</th>\n",
       "      <th>proteins</th>\n",
       "      <th>genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AFGHIJK</td>\n",
       "      <td>0;1</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>7</td>\n",
       "      <td>xx;yy</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LMNOPQR</td>\n",
       "      <td>0;1</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>7</td>\n",
       "      <td>xx;yy</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABCDESTK</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>8</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MABCDESTK</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>9</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AFGHIJKLMNOPQR</td>\n",
       "      <td>0;1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>14</td>\n",
       "      <td>xx;yy</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>LMNOPQRAFGHIJK</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>14</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ABCDESTKAFGHIJK</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>15</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MABCDESTKAFGHIJK</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>16</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AFGHIJKLMNOPQRAFGHIJK</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>21</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ABCDESTKAFGHIJKLMNOPQR</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>22</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>MABCDESTKAFGHIJKLMNOPQR</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>23</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   sequence protein_idxes  miss_cleavage  is_prot_nterm  \\\n",
       "0                   AFGHIJK           0;1              0           True   \n",
       "1                   LMNOPQR           0;1              0          False   \n",
       "2                  ABCDESTK             0              0           True   \n",
       "3                 MABCDESTK             0              0           True   \n",
       "4            AFGHIJKLMNOPQR           0;1              1           True   \n",
       "5            LMNOPQRAFGHIJK             0              1          False   \n",
       "6           ABCDESTKAFGHIJK             0              1           True   \n",
       "7          MABCDESTKAFGHIJK             0              1           True   \n",
       "8     AFGHIJKLMNOPQRAFGHIJK             0              2          False   \n",
       "9    ABCDESTKAFGHIJKLMNOPQR             0              2           True   \n",
       "10  MABCDESTKAFGHIJKLMNOPQR             0              2           True   \n",
       "\n",
       "    is_prot_cterm mods mod_sites  nAA proteins genes  \n",
       "0            True                   7    xx;yy  gene  \n",
       "1            True                   7    xx;yy  gene  \n",
       "2           False                   8       xx        \n",
       "3           False                   9       xx        \n",
       "4            True                  14    xx;yy  gene  \n",
       "5            True                  14       xx        \n",
       "6           False                  15       xx        \n",
       "7           False                  16       xx        \n",
       "8            True                  21       xx        \n",
       "9           False                  22       xx        \n",
       "10          False                  23       xx        "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_lib.append_protein_name()\n",
    "assert 'proteins' in _lib.precursor_df.columns\n",
    "_lib.precursor_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "for i in range(len(_lib.precursor_df)):\n",
    "    seq = _lib.precursor_df.sequence[i]\n",
    "    # test is_prot_nterm\n",
    "    if prot1.startswith(seq) or prot2.startswith(seq):\n",
    "        assert _lib.precursor_df.is_prot_nterm[i], seq\n",
    "    elif prot1[1:].startswith(seq): # M.xxxxx\n",
    "        assert _lib.precursor_df.is_prot_nterm[i], seq\n",
    "    else:\n",
    "        assert not _lib.precursor_df.is_prot_nterm[i], seq\n",
    "    # test is_prot_cterm\n",
    "    if prot1.endswith(seq) or prot2.endswith(seq):\n",
    "        assert _lib.precursor_df.is_prot_cterm[i], seq\n",
    "    else:\n",
    "        assert not _lib.precursor_df.is_prot_cterm[i], seq\n",
    "    # test protein_idxes\n",
    "    if seq in prot1 and seq in prot2:\n",
    "        assert _lib.precursor_df.protein_idxes[i] == '0;1'\n",
    "        assert _lib.precursor_df.proteins[i] == 'xx;yy'\n",
    "        assert _lib.precursor_df.genes[i] == 'gene'\n",
    "    else:\n",
    "        assert ';' not in _lib.precursor_df.protein_idxes[i]\n",
    "        assert ';' not in _lib.precursor_df.proteins[i]\n",
    "        assert _lib.precursor_df.genes[i] == ''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>protein_idxes</th>\n",
       "      <th>miss_cleavage</th>\n",
       "      <th>is_prot_nterm</th>\n",
       "      <th>is_prot_cterm</th>\n",
       "      <th>mods</th>\n",
       "      <th>mod_sites</th>\n",
       "      <th>nAA</th>\n",
       "      <th>proteins</th>\n",
       "      <th>genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AFGHIJK</td>\n",
       "      <td>0;1</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>7</td>\n",
       "      <td>xx;yy</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AFGHIJK</td>\n",
       "      <td>0;1</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Acetyl@Protein N-term</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>xx;yy</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LMNOPQR</td>\n",
       "      <td>0;1</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Oxidation@M</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>xx;yy</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LMNOPQR</td>\n",
       "      <td>0;1</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>7</td>\n",
       "      <td>xx;yy</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABCDESTK</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ABCDESTK</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Acetyl@Protein N-term</td>\n",
       "      <td>3;0</td>\n",
       "      <td>8</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MABCDESTK</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Oxidation@M</td>\n",
       "      <td>4;1</td>\n",
       "      <td>9</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MABCDESTK</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MABCDESTK</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...</td>\n",
       "      <td>4;0;1</td>\n",
       "      <td>9</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MABCDESTK</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Acetyl@Protein N-term</td>\n",
       "      <td>4;0</td>\n",
       "      <td>9</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>AFGHIJKLMNOPQR</td>\n",
       "      <td>0;1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Oxidation@M</td>\n",
       "      <td>9</td>\n",
       "      <td>14</td>\n",
       "      <td>xx;yy</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>AFGHIJKLMNOPQR</td>\n",
       "      <td>0;1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>14</td>\n",
       "      <td>xx;yy</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>AFGHIJKLMNOPQR</td>\n",
       "      <td>0;1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Acetyl@Protein N-term;Oxidation@M</td>\n",
       "      <td>0;9</td>\n",
       "      <td>14</td>\n",
       "      <td>xx;yy</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AFGHIJKLMNOPQR</td>\n",
       "      <td>0;1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Acetyl@Protein N-term</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>xx;yy</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>LMNOPQRAFGHIJK</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Oxidation@M</td>\n",
       "      <td>2</td>\n",
       "      <td>14</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>LMNOPQRAFGHIJK</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>14</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ABCDESTKAFGHIJK</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C</td>\n",
       "      <td>3</td>\n",
       "      <td>15</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ABCDESTKAFGHIJK</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Acetyl@Protein N-term</td>\n",
       "      <td>3;0</td>\n",
       "      <td>15</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>MABCDESTKAFGHIJK</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Oxidation@M</td>\n",
       "      <td>4;1</td>\n",
       "      <td>16</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MABCDESTKAFGHIJK</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C</td>\n",
       "      <td>4</td>\n",
       "      <td>16</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>MABCDESTKAFGHIJK</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...</td>\n",
       "      <td>4;0;1</td>\n",
       "      <td>16</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>MABCDESTKAFGHIJK</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Acetyl@Protein N-term</td>\n",
       "      <td>4;0</td>\n",
       "      <td>16</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>AFGHIJKLMNOPQRAFGHIJK</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Oxidation@M</td>\n",
       "      <td>9</td>\n",
       "      <td>21</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>AFGHIJKLMNOPQRAFGHIJK</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>21</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ABCDESTKAFGHIJKLMNOPQR</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Oxidation@M</td>\n",
       "      <td>3;17</td>\n",
       "      <td>22</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ABCDESTKAFGHIJKLMNOPQR</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C</td>\n",
       "      <td>3</td>\n",
       "      <td>22</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ABCDESTKAFGHIJKLMNOPQR</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...</td>\n",
       "      <td>3;0;17</td>\n",
       "      <td>22</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ABCDESTKAFGHIJKLMNOPQR</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Acetyl@Protein N-term</td>\n",
       "      <td>3;0</td>\n",
       "      <td>22</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>MABCDESTKAFGHIJKLMNOPQR</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Oxidation@M</td>\n",
       "      <td>4;1</td>\n",
       "      <td>23</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>MABCDESTKAFGHIJKLMNOPQR</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Oxidation@M</td>\n",
       "      <td>4;18</td>\n",
       "      <td>23</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>MABCDESTKAFGHIJKLMNOPQR</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Oxidation@M;Oxidation@M</td>\n",
       "      <td>4;1;18</td>\n",
       "      <td>23</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>MABCDESTKAFGHIJKLMNOPQR</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C</td>\n",
       "      <td>4</td>\n",
       "      <td>23</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>MABCDESTKAFGHIJKLMNOPQR</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...</td>\n",
       "      <td>4;0;1</td>\n",
       "      <td>23</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>MABCDESTKAFGHIJKLMNOPQR</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...</td>\n",
       "      <td>4;0;18</td>\n",
       "      <td>23</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>MABCDESTKAFGHIJKLMNOPQR</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...</td>\n",
       "      <td>4;0;1;18</td>\n",
       "      <td>23</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>MABCDESTKAFGHIJKLMNOPQR</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Acetyl@Protein N-term</td>\n",
       "      <td>4;0</td>\n",
       "      <td>23</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   sequence protein_idxes  miss_cleavage  is_prot_nterm  \\\n",
       "0                   AFGHIJK           0;1              0           True   \n",
       "1                   AFGHIJK           0;1              0           True   \n",
       "2                   LMNOPQR           0;1              0          False   \n",
       "3                   LMNOPQR           0;1              0          False   \n",
       "4                  ABCDESTK             0              0           True   \n",
       "5                  ABCDESTK             0              0           True   \n",
       "6                 MABCDESTK             0              0           True   \n",
       "7                 MABCDESTK             0              0           True   \n",
       "8                 MABCDESTK             0              0           True   \n",
       "9                 MABCDESTK             0              0           True   \n",
       "10           AFGHIJKLMNOPQR           0;1              1           True   \n",
       "11           AFGHIJKLMNOPQR           0;1              1           True   \n",
       "12           AFGHIJKLMNOPQR           0;1              1           True   \n",
       "13           AFGHIJKLMNOPQR           0;1              1           True   \n",
       "14           LMNOPQRAFGHIJK             0              1          False   \n",
       "15           LMNOPQRAFGHIJK             0              1          False   \n",
       "16          ABCDESTKAFGHIJK             0              1           True   \n",
       "17          ABCDESTKAFGHIJK             0              1           True   \n",
       "18         MABCDESTKAFGHIJK             0              1           True   \n",
       "19         MABCDESTKAFGHIJK             0              1           True   \n",
       "20         MABCDESTKAFGHIJK             0              1           True   \n",
       "21         MABCDESTKAFGHIJK             0              1           True   \n",
       "22    AFGHIJKLMNOPQRAFGHIJK             0              2          False   \n",
       "23    AFGHIJKLMNOPQRAFGHIJK             0              2          False   \n",
       "24   ABCDESTKAFGHIJKLMNOPQR             0              2           True   \n",
       "25   ABCDESTKAFGHIJKLMNOPQR             0              2           True   \n",
       "26   ABCDESTKAFGHIJKLMNOPQR             0              2           True   \n",
       "27   ABCDESTKAFGHIJKLMNOPQR             0              2           True   \n",
       "28  MABCDESTKAFGHIJKLMNOPQR             0              2           True   \n",
       "29  MABCDESTKAFGHIJKLMNOPQR             0              2           True   \n",
       "30  MABCDESTKAFGHIJKLMNOPQR             0              2           True   \n",
       "31  MABCDESTKAFGHIJKLMNOPQR             0              2           True   \n",
       "32  MABCDESTKAFGHIJKLMNOPQR             0              2           True   \n",
       "33  MABCDESTKAFGHIJKLMNOPQR             0              2           True   \n",
       "34  MABCDESTKAFGHIJKLMNOPQR             0              2           True   \n",
       "35  MABCDESTKAFGHIJKLMNOPQR             0              2           True   \n",
       "\n",
       "    is_prot_cterm                                               mods  \\\n",
       "0            True                                                      \n",
       "1            True                              Acetyl@Protein N-term   \n",
       "2            True                                        Oxidation@M   \n",
       "3            True                                                      \n",
       "4           False                                  Carbamidomethyl@C   \n",
       "5           False            Carbamidomethyl@C;Acetyl@Protein N-term   \n",
       "6           False                      Carbamidomethyl@C;Oxidation@M   \n",
       "7           False                                  Carbamidomethyl@C   \n",
       "8           False  Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...   \n",
       "9           False            Carbamidomethyl@C;Acetyl@Protein N-term   \n",
       "10           True                                        Oxidation@M   \n",
       "11           True                                                      \n",
       "12           True                  Acetyl@Protein N-term;Oxidation@M   \n",
       "13           True                              Acetyl@Protein N-term   \n",
       "14           True                                        Oxidation@M   \n",
       "15           True                                                      \n",
       "16          False                                  Carbamidomethyl@C   \n",
       "17          False            Carbamidomethyl@C;Acetyl@Protein N-term   \n",
       "18          False                      Carbamidomethyl@C;Oxidation@M   \n",
       "19          False                                  Carbamidomethyl@C   \n",
       "20          False  Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...   \n",
       "21          False            Carbamidomethyl@C;Acetyl@Protein N-term   \n",
       "22           True                                        Oxidation@M   \n",
       "23           True                                                      \n",
       "24          False                      Carbamidomethyl@C;Oxidation@M   \n",
       "25          False                                  Carbamidomethyl@C   \n",
       "26          False  Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...   \n",
       "27          False            Carbamidomethyl@C;Acetyl@Protein N-term   \n",
       "28          False                      Carbamidomethyl@C;Oxidation@M   \n",
       "29          False                      Carbamidomethyl@C;Oxidation@M   \n",
       "30          False          Carbamidomethyl@C;Oxidation@M;Oxidation@M   \n",
       "31          False                                  Carbamidomethyl@C   \n",
       "32          False  Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...   \n",
       "33          False  Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...   \n",
       "34          False  Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...   \n",
       "35          False            Carbamidomethyl@C;Acetyl@Protein N-term   \n",
       "\n",
       "   mod_sites  nAA proteins genes  \n",
       "0               7    xx;yy  gene  \n",
       "1          0    7    xx;yy  gene  \n",
       "2          2    7    xx;yy  gene  \n",
       "3               7    xx;yy  gene  \n",
       "4          3    8       xx        \n",
       "5        3;0    8       xx        \n",
       "6        4;1    9       xx        \n",
       "7          4    9       xx        \n",
       "8      4;0;1    9       xx        \n",
       "9        4;0    9       xx        \n",
       "10         9   14    xx;yy  gene  \n",
       "11             14    xx;yy  gene  \n",
       "12       0;9   14    xx;yy  gene  \n",
       "13         0   14    xx;yy  gene  \n",
       "14         2   14       xx        \n",
       "15             14       xx        \n",
       "16         3   15       xx        \n",
       "17       3;0   15       xx        \n",
       "18       4;1   16       xx        \n",
       "19         4   16       xx        \n",
       "20     4;0;1   16       xx        \n",
       "21       4;0   16       xx        \n",
       "22         9   21       xx        \n",
       "23             21       xx        \n",
       "24      3;17   22       xx        \n",
       "25         3   22       xx        \n",
       "26    3;0;17   22       xx        \n",
       "27       3;0   22       xx        \n",
       "28       4;1   23       xx        \n",
       "29      4;18   23       xx        \n",
       "30    4;1;18   23       xx        \n",
       "31         4   23       xx        \n",
       "32     4;0;1   23       xx        \n",
       "33    4;0;18   23       xx        \n",
       "34  4;0;1;18   23       xx        \n",
       "35       4;0   23       xx        "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_lib.add_modifications()\n",
    "_lib.precursor_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "for i in range(len(_lib.precursor_df)):\n",
    "    seq = _lib.precursor_df.sequence[i]\n",
    "    mods = _lib.precursor_df.mods[i]\n",
    "    sites = _lib.precursor_df.mod_sites[i]\n",
    "    # test fix mods\n",
    "    if 'C' in seq:\n",
    "        assert str(seq.find('C')+1) in sites\n",
    "        assert 'Carbamidomethyl@C' in mods\n",
    "    else:\n",
    "        assert 'Carbamidomethyl@C' not in mods\n",
    "    # test Acetyl@Protein N-term\n",
    "    if 'Acetyl@Protein N-term' in mods:\n",
    "        assert _lib.precursor_df.is_prot_nterm[i]\n",
    "        assert '0' in sites\n",
    "    if '0' in mods:\n",
    "        assert _lib.precursor_df.is_prot_nterm[i]\n",
    "        assert 'Acetyl@Protein N-term' in mods\n",
    "    if not _lib.precursor_df.is_prot_nterm[i]:\n",
    "        assert 'Acetyl@Protein N-term' not in mods\n",
    "    # test Oxidation@M\n",
    "    if 'Oxidation@M' in mods:\n",
    "        assert 'M' in seq\n",
    "        assert str(seq.find('M')+1) in sites\n",
    "    # test unmodified\n",
    "    if mods == '':\n",
    "        assert sites == ''\n",
    "    if sites == '':\n",
    "        assert mods == ''\n",
    "df = _lib.precursor_df\n",
    "# at least one nterm peptide does not contain Acetyl@Protein N-term\n",
    "assert not df[df.is_prot_nterm].mod_sites.str.contains('0').all()\n",
    "# at least one nterm peptide contains Acetyl@Protein N-term\n",
    "assert df[df.is_prot_nterm].mod_sites.str.contains('0').any()\n",
    "# test var mod Oxidation@M\n",
    "assert not df[df.sequence.str.contains('M')].mods.str.contains('Oxidation@M').all()\n",
    "assert df[df.sequence.str.contains('M')].mods.str.contains('Oxidation@M').any()\n",
    "assert '' in df.mods.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>protein_idxes</th>\n",
       "      <th>miss_cleavage</th>\n",
       "      <th>is_prot_nterm</th>\n",
       "      <th>is_prot_cterm</th>\n",
       "      <th>mods</th>\n",
       "      <th>mod_sites</th>\n",
       "      <th>nAA</th>\n",
       "      <th>proteins</th>\n",
       "      <th>genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AFGHIJK</td>\n",
       "      <td>0;1</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>7</td>\n",
       "      <td>xx;yy</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AFGHIJK</td>\n",
       "      <td>0;1</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Acetyl@Protein N-term</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>xx;yy</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LMNOPQR</td>\n",
       "      <td>0;1</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Oxidation@M</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>xx;yy</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LMNOPQR</td>\n",
       "      <td>0;1</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>7</td>\n",
       "      <td>xx;yy</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABCDESTK</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Phospho@S</td>\n",
       "      <td>3;6</td>\n",
       "      <td>8</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>MABCDESTKAFGHIJKLMNOPQR</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...</td>\n",
       "      <td>4;0;1;18;8</td>\n",
       "      <td>23</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>MABCDESTKAFGHIJKLMNOPQR</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...</td>\n",
       "      <td>4;0;1;18</td>\n",
       "      <td>23</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>MABCDESTKAFGHIJKLMNOPQR</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Acetyl@Protein N-term;Phospho@S</td>\n",
       "      <td>4;0;7</td>\n",
       "      <td>23</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>MABCDESTKAFGHIJKLMNOPQR</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Acetyl@Protein N-term;Phospho@T</td>\n",
       "      <td>4;0;8</td>\n",
       "      <td>23</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>MABCDESTKAFGHIJKLMNOPQR</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Acetyl@Protein N-term</td>\n",
       "      <td>4;0</td>\n",
       "      <td>23</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>84 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   sequence protein_idxes  miss_cleavage  is_prot_nterm  \\\n",
       "0                   AFGHIJK           0;1              0           True   \n",
       "1                   AFGHIJK           0;1              0           True   \n",
       "2                   LMNOPQR           0;1              0          False   \n",
       "3                   LMNOPQR           0;1              0          False   \n",
       "4                  ABCDESTK             0              0           True   \n",
       "..                      ...           ...            ...            ...   \n",
       "79  MABCDESTKAFGHIJKLMNOPQR             0              2           True   \n",
       "80  MABCDESTKAFGHIJKLMNOPQR             0              2           True   \n",
       "81  MABCDESTKAFGHIJKLMNOPQR             0              2           True   \n",
       "82  MABCDESTKAFGHIJKLMNOPQR             0              2           True   \n",
       "83  MABCDESTKAFGHIJKLMNOPQR             0              2           True   \n",
       "\n",
       "    is_prot_cterm                                               mods  \\\n",
       "0            True                                                      \n",
       "1            True                              Acetyl@Protein N-term   \n",
       "2            True                                        Oxidation@M   \n",
       "3            True                                                      \n",
       "4           False                        Carbamidomethyl@C;Phospho@S   \n",
       "..            ...                                                ...   \n",
       "79          False  Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...   \n",
       "80          False  Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...   \n",
       "81          False  Carbamidomethyl@C;Acetyl@Protein N-term;Phospho@S   \n",
       "82          False  Carbamidomethyl@C;Acetyl@Protein N-term;Phospho@T   \n",
       "83          False            Carbamidomethyl@C;Acetyl@Protein N-term   \n",
       "\n",
       "     mod_sites  nAA proteins genes  \n",
       "0                 7    xx;yy  gene  \n",
       "1            0    7    xx;yy  gene  \n",
       "2            2    7    xx;yy  gene  \n",
       "3                 7    xx;yy  gene  \n",
       "4          3;6    8       xx        \n",
       "..         ...  ...      ...   ...  \n",
       "79  4;0;1;18;8   23       xx        \n",
       "80    4;0;1;18   23       xx        \n",
       "81       4;0;7   23       xx        \n",
       "82       4;0;8   23       xx        \n",
       "83         4;0   23       xx        \n",
       "\n",
       "[84 rows x 10 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_lib.add_additional_modifications(['Phospho@S','Phospho@T'])\n",
    "assert _lib.precursor_df.mods.str.contains('Phospho').any()\n",
    "_lib.precursor_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>protein_idxes</th>\n",
       "      <th>miss_cleavage</th>\n",
       "      <th>is_prot_nterm</th>\n",
       "      <th>is_prot_cterm</th>\n",
       "      <th>mods</th>\n",
       "      <th>mod_sites</th>\n",
       "      <th>nAA</th>\n",
       "      <th>proteins</th>\n",
       "      <th>genes</th>\n",
       "      <th>label_channel</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AFGHIJK</td>\n",
       "      <td>0;1</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>7</td>\n",
       "      <td>xx;yy</td>\n",
       "      <td>gene</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AFGHIJK</td>\n",
       "      <td>0;1</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Acetyl@Protein N-term</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>xx;yy</td>\n",
       "      <td>gene</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LMNOPQR</td>\n",
       "      <td>0;1</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Oxidation@M</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>xx;yy</td>\n",
       "      <td>gene</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LMNOPQR</td>\n",
       "      <td>0;1</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>7</td>\n",
       "      <td>xx;yy</td>\n",
       "      <td>gene</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABCDESTK</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Phospho@S</td>\n",
       "      <td>3;6</td>\n",
       "      <td>8</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>MABCDESTKAFGHIJKLMNOPQR</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...</td>\n",
       "      <td>4;0;1;18;8;9;16</td>\n",
       "      <td>23</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "      <td>heavy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248</th>\n",
       "      <td>MABCDESTKAFGHIJKLMNOPQR</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...</td>\n",
       "      <td>4;0;1;18;9;16</td>\n",
       "      <td>23</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "      <td>heavy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249</th>\n",
       "      <td>MABCDESTKAFGHIJKLMNOPQR</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Acetyl@Protein N-term;Phosph...</td>\n",
       "      <td>4;0;7;9;16</td>\n",
       "      <td>23</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "      <td>heavy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250</th>\n",
       "      <td>MABCDESTKAFGHIJKLMNOPQR</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Acetyl@Protein N-term;Phosph...</td>\n",
       "      <td>4;0;8;9;16</td>\n",
       "      <td>23</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "      <td>heavy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251</th>\n",
       "      <td>MABCDESTKAFGHIJKLMNOPQR</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Carbamidomethyl@C;Acetyl@Protein N-term;Dimeth...</td>\n",
       "      <td>4;0;9;16</td>\n",
       "      <td>23</td>\n",
       "      <td>xx</td>\n",
       "      <td></td>\n",
       "      <td>heavy</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>252 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    sequence protein_idxes  miss_cleavage  is_prot_nterm  \\\n",
       "0                    AFGHIJK           0;1              0           True   \n",
       "1                    AFGHIJK           0;1              0           True   \n",
       "2                    LMNOPQR           0;1              0          False   \n",
       "3                    LMNOPQR           0;1              0          False   \n",
       "4                   ABCDESTK             0              0           True   \n",
       "..                       ...           ...            ...            ...   \n",
       "247  MABCDESTKAFGHIJKLMNOPQR             0              2           True   \n",
       "248  MABCDESTKAFGHIJKLMNOPQR             0              2           True   \n",
       "249  MABCDESTKAFGHIJKLMNOPQR             0              2           True   \n",
       "250  MABCDESTKAFGHIJKLMNOPQR             0              2           True   \n",
       "251  MABCDESTKAFGHIJKLMNOPQR             0              2           True   \n",
       "\n",
       "     is_prot_cterm                                               mods  \\\n",
       "0             True                                                      \n",
       "1             True                              Acetyl@Protein N-term   \n",
       "2             True                                        Oxidation@M   \n",
       "3             True                                                      \n",
       "4            False                        Carbamidomethyl@C;Phospho@S   \n",
       "..             ...                                                ...   \n",
       "247          False  Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...   \n",
       "248          False  Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...   \n",
       "249          False  Carbamidomethyl@C;Acetyl@Protein N-term;Phosph...   \n",
       "250          False  Carbamidomethyl@C;Acetyl@Protein N-term;Phosph...   \n",
       "251          False  Carbamidomethyl@C;Acetyl@Protein N-term;Dimeth...   \n",
       "\n",
       "           mod_sites  nAA proteins genes label_channel  \n",
       "0                       7    xx;yy  gene          none  \n",
       "1                  0    7    xx;yy  gene          none  \n",
       "2                  2    7    xx;yy  gene          none  \n",
       "3                       7    xx;yy  gene          none  \n",
       "4                3;6    8       xx                none  \n",
       "..               ...  ...      ...   ...           ...  \n",
       "247  4;0;1;18;8;9;16   23       xx               heavy  \n",
       "248    4;0;1;18;9;16   23       xx               heavy  \n",
       "249       4;0;7;9;16   23       xx               heavy  \n",
       "250       4;0;8;9;16   23       xx               heavy  \n",
       "251         4;0;9;16   23       xx               heavy  \n",
       "\n",
       "[252 rows x 11 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#hide\n",
    "_lib.add_peptide_labeling({\n",
    "    'none': [], # not labelled for reference\n",
    "    'light': ['Dimethyl@Any N-term','Dimethyl@K'],\n",
    "    'heavy': ['Dimethyl:2H(6)13C(2)@Any N-term','Dimethyl:2H(6)13C(2)@K'],\n",
    "})\n",
    "_lib.precursor_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>protein_idxes</th>\n",
       "      <th>miss_cleavage</th>\n",
       "      <th>is_prot_nterm</th>\n",
       "      <th>is_prot_cterm</th>\n",
       "      <th>mods</th>\n",
       "      <th>mod_sites</th>\n",
       "      <th>nAA</th>\n",
       "      <th>decoy</th>\n",
       "      <th>charge</th>\n",
       "      <th>precursor_mz</th>\n",
       "      <th>isotope_intensity_m1</th>\n",
       "      <th>isotope_intensity_m2</th>\n",
       "      <th>isotope_apex_intensity</th>\n",
       "      <th>isotope_apex_index</th>\n",
       "      <th>isotope_mz_m1</th>\n",
       "      <th>isotope_mz_m2</th>\n",
       "      <th>isotope_apex_mz</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LMNPQRST</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Oxidation@M</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>481.739834</td>\n",
       "      <td>0.478814</td>\n",
       "      <td>0.183746</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>482.241484</td>\n",
       "      <td>482.743134</td>\n",
       "      <td>481.739834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LMNPQRST</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Oxidation@M</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>321.495648</td>\n",
       "      <td>0.478814</td>\n",
       "      <td>0.183746</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>321.830081</td>\n",
       "      <td>322.164515</td>\n",
       "      <td>321.495648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LMNPQRST</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Oxidation@M</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>241.373555</td>\n",
       "      <td>0.478814</td>\n",
       "      <td>0.183746</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>241.624380</td>\n",
       "      <td>241.875205</td>\n",
       "      <td>241.373555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LMNPQRST</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>473.742377</td>\n",
       "      <td>0.478433</td>\n",
       "      <td>0.181509</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>474.244027</td>\n",
       "      <td>474.745677</td>\n",
       "      <td>473.742377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LMNPQRST</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>316.164010</td>\n",
       "      <td>0.478433</td>\n",
       "      <td>0.181509</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>316.498443</td>\n",
       "      <td>316.832877</td>\n",
       "      <td>316.164010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>FGHIKLMNPQRST</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Oxidation@M</td>\n",
       "      <td>7</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>515.604920</td>\n",
       "      <td>0.828432</td>\n",
       "      <td>0.420789</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>515.939354</td>\n",
       "      <td>516.273787</td>\n",
       "      <td>515.604920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>FGHIKLMNPQRST</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Oxidation@M</td>\n",
       "      <td>7</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>386.955509</td>\n",
       "      <td>0.828432</td>\n",
       "      <td>0.420789</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>387.206334</td>\n",
       "      <td>387.457159</td>\n",
       "      <td>386.955509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>FGHIKLMNPQRST</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>764.906285</td>\n",
       "      <td>0.828051</td>\n",
       "      <td>0.418418</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>765.407935</td>\n",
       "      <td>765.909585</td>\n",
       "      <td>764.906285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>FGHIKLMNPQRST</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>510.273282</td>\n",
       "      <td>0.828051</td>\n",
       "      <td>0.418418</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>510.607715</td>\n",
       "      <td>510.942149</td>\n",
       "      <td>510.273282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>FGHIKLMNPQRST</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>382.956781</td>\n",
       "      <td>0.828051</td>\n",
       "      <td>0.418418</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>383.207606</td>\n",
       "      <td>383.458431</td>\n",
       "      <td>382.956781</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>84 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         sequence protein_idxes  miss_cleavage  is_prot_nterm  is_prot_cterm  \\\n",
       "0        LMNPQRST             0              1          False           True   \n",
       "1        LMNPQRST             0              1          False           True   \n",
       "2        LMNPQRST             0              1          False           True   \n",
       "3        LMNPQRST             0              1          False           True   \n",
       "4        LMNPQRST             0              1          False           True   \n",
       "..            ...           ...            ...            ...            ...   \n",
       "79  FGHIKLMNPQRST             0              2          False           True   \n",
       "80  FGHIKLMNPQRST             0              2          False           True   \n",
       "81  FGHIKLMNPQRST             0              2          False           True   \n",
       "82  FGHIKLMNPQRST             0              2          False           True   \n",
       "83  FGHIKLMNPQRST             0              2          False           True   \n",
       "\n",
       "           mods mod_sites  nAA  decoy  charge  precursor_mz  \\\n",
       "0   Oxidation@M         2    8      0       2    481.739834   \n",
       "1   Oxidation@M         2    8      0       3    321.495648   \n",
       "2   Oxidation@M         2    8      0       4    241.373555   \n",
       "3                            8      0       2    473.742377   \n",
       "4                            8      0       3    316.164010   \n",
       "..          ...       ...  ...    ...     ...           ...   \n",
       "79  Oxidation@M         7   13      0       3    515.604920   \n",
       "80  Oxidation@M         7   13      0       4    386.955509   \n",
       "81                          13      0       2    764.906285   \n",
       "82                          13      0       3    510.273282   \n",
       "83                          13      0       4    382.956781   \n",
       "\n",
       "    isotope_intensity_m1  isotope_intensity_m2  isotope_apex_intensity  \\\n",
       "0               0.478814              0.183746                     1.0   \n",
       "1               0.478814              0.183746                     1.0   \n",
       "2               0.478814              0.183746                     1.0   \n",
       "3               0.478433              0.181509                     1.0   \n",
       "4               0.478433              0.181509                     1.0   \n",
       "..                   ...                   ...                     ...   \n",
       "79              0.828432              0.420789                     1.0   \n",
       "80              0.828432              0.420789                     1.0   \n",
       "81              0.828051              0.418418                     1.0   \n",
       "82              0.828051              0.418418                     1.0   \n",
       "83              0.828051              0.418418                     1.0   \n",
       "\n",
       "    isotope_apex_index  isotope_mz_m1  isotope_mz_m2  isotope_apex_mz  \n",
       "0                    0     482.241484     482.743134       481.739834  \n",
       "1                    0     321.830081     322.164515       321.495648  \n",
       "2                    0     241.624380     241.875205       241.373555  \n",
       "3                    0     474.244027     474.745677       473.742377  \n",
       "4                    0     316.498443     316.832877       316.164010  \n",
       "..                 ...            ...            ...              ...  \n",
       "79                   0     515.939354     516.273787       515.604920  \n",
       "80                   0     387.206334     387.457159       386.955509  \n",
       "81                   0     765.407935     765.909585       764.906285  \n",
       "82                   0     510.607715     510.942149       510.273282  \n",
       "83                   0     383.207606     383.458431       382.956781  \n",
       "\n",
       "[84 rows x 18 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_lib = FastaLib(\n",
    "    ['b_z1','y_z1'], I_to_L=False, \n",
    "    decoy='pseudo_reverse'\n",
    ")\n",
    "prot1 = 'MACDESTYKBKFGHIKLMNPQRST'\n",
    "prot2 = 'FGHIKLMNPQR'\n",
    "protein_dict = {\n",
    "    'xx': {\n",
    "        'protein_id': 'xx',\n",
    "        'sequence': prot1\n",
    "    },\n",
    "    'yy': {\n",
    "        'protein_id': 'yy',\n",
    "        'sequence': prot2\n",
    "    }\n",
    "}\n",
    "_lib.import_and_process_protein_dict(protein_dict)\n",
    "_lib.calc_precursor_isotope()\n",
    "assert (_lib.precursor_df.charge == _lib.min_precursor_charge).any()\n",
    "assert (_lib.precursor_df.charge == _lib.max_precursor_charge).any()\n",
    "assert (_lib.precursor_df.decoy==1).any()\n",
    "assert ('MACDESTY'[::-1]+'K') in _lib.precursor_df.sequence.values\n",
    "assert 'isotope_apex_index' in _lib.precursor_df.columns\n",
    "assert 'isotope_apex_intensity' in _lib.precursor_df.columns\n",
    "assert ~_lib.precursor_df.sequence.str.contains('B').any()\n",
    "_lib.precursor_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>protein_idxes</th>\n",
       "      <th>miss_cleavage</th>\n",
       "      <th>is_prot_nterm</th>\n",
       "      <th>is_prot_cterm</th>\n",
       "      <th>mods</th>\n",
       "      <th>mod_sites</th>\n",
       "      <th>nAA</th>\n",
       "      <th>decoy</th>\n",
       "      <th>charge</th>\n",
       "      <th>label_channel</th>\n",
       "      <th>precursor_mz</th>\n",
       "      <th>isotope_intensity_m1</th>\n",
       "      <th>isotope_intensity_m2</th>\n",
       "      <th>isotope_apex_intensity</th>\n",
       "      <th>isotope_apex_index</th>\n",
       "      <th>isotope_mz_m1</th>\n",
       "      <th>isotope_mz_m2</th>\n",
       "      <th>isotope_apex_mz</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LMNPQRST</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Oxidation@M;Dimethyl@Any N-term</td>\n",
       "      <td>2;0</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>light</td>\n",
       "      <td>495.755484</td>\n",
       "      <td>0.500906</td>\n",
       "      <td>0.194451</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>496.257134</td>\n",
       "      <td>496.758784</td>\n",
       "      <td>495.755484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LMNPQRST</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Oxidation@M;Dimethyl@Any N-term</td>\n",
       "      <td>2;0</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>light</td>\n",
       "      <td>330.839415</td>\n",
       "      <td>0.500906</td>\n",
       "      <td>0.194451</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>331.173848</td>\n",
       "      <td>331.508282</td>\n",
       "      <td>330.839415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LMNPQRST</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Oxidation@M;Dimethyl@Any N-term</td>\n",
       "      <td>2;0</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>light</td>\n",
       "      <td>248.381380</td>\n",
       "      <td>0.500906</td>\n",
       "      <td>0.194451</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>248.632205</td>\n",
       "      <td>248.883030</td>\n",
       "      <td>248.381380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LMNPQRST</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Dimethyl@Any N-term</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>light</td>\n",
       "      <td>487.758027</td>\n",
       "      <td>0.500525</td>\n",
       "      <td>0.192205</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>488.259677</td>\n",
       "      <td>488.761327</td>\n",
       "      <td>487.758027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LMNPQRST</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Dimethyl@Any N-term</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>light</td>\n",
       "      <td>325.507777</td>\n",
       "      <td>0.500525</td>\n",
       "      <td>0.192205</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>325.842210</td>\n",
       "      <td>326.176643</td>\n",
       "      <td>325.507777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>FGHIKLMNPQRST</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Oxidation@M;Dimethyl:2H(6)13C(2)@Any N-term;Di...</td>\n",
       "      <td>7;0;5</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>heavy</td>\n",
       "      <td>539.655367</td>\n",
       "      <td>0.788273</td>\n",
       "      <td>0.392103</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>539.989801</td>\n",
       "      <td>540.324234</td>\n",
       "      <td>539.655367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>FGHIKLMNPQRST</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Oxidation@M;Dimethyl:2H(6)13C(2)@Any N-term;Di...</td>\n",
       "      <td>7;0;5</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>heavy</td>\n",
       "      <td>404.993344</td>\n",
       "      <td>0.788273</td>\n",
       "      <td>0.392103</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>405.244169</td>\n",
       "      <td>405.494994</td>\n",
       "      <td>404.993344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>FGHIKLMNPQRST</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Dimethyl:2H(6)13C(2)@Any N-term;Dimethyl:2H(6)...</td>\n",
       "      <td>0;5</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>heavy</td>\n",
       "      <td>800.981955</td>\n",
       "      <td>0.787646</td>\n",
       "      <td>0.389779</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>801.483605</td>\n",
       "      <td>801.985255</td>\n",
       "      <td>800.981955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>FGHIKLMNPQRST</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Dimethyl:2H(6)13C(2)@Any N-term;Dimethyl:2H(6)...</td>\n",
       "      <td>0;5</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>heavy</td>\n",
       "      <td>534.323729</td>\n",
       "      <td>0.787646</td>\n",
       "      <td>0.389779</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>534.658162</td>\n",
       "      <td>534.992596</td>\n",
       "      <td>534.323729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>FGHIKLMNPQRST</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Dimethyl:2H(6)13C(2)@Any N-term;Dimethyl:2H(6)...</td>\n",
       "      <td>0;5</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>heavy</td>\n",
       "      <td>400.994616</td>\n",
       "      <td>0.787646</td>\n",
       "      <td>0.389779</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>401.245441</td>\n",
       "      <td>401.496266</td>\n",
       "      <td>400.994616</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>168 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          sequence protein_idxes  miss_cleavage  is_prot_nterm  is_prot_cterm  \\\n",
       "0         LMNPQRST             0              1          False           True   \n",
       "1         LMNPQRST             0              1          False           True   \n",
       "2         LMNPQRST             0              1          False           True   \n",
       "3         LMNPQRST             0              1          False           True   \n",
       "4         LMNPQRST             0              1          False           True   \n",
       "..             ...           ...            ...            ...            ...   \n",
       "163  FGHIKLMNPQRST             0              2          False           True   \n",
       "164  FGHIKLMNPQRST             0              2          False           True   \n",
       "165  FGHIKLMNPQRST             0              2          False           True   \n",
       "166  FGHIKLMNPQRST             0              2          False           True   \n",
       "167  FGHIKLMNPQRST             0              2          False           True   \n",
       "\n",
       "                                                  mods mod_sites  nAA  decoy  \\\n",
       "0                      Oxidation@M;Dimethyl@Any N-term       2;0    8      0   \n",
       "1                      Oxidation@M;Dimethyl@Any N-term       2;0    8      0   \n",
       "2                      Oxidation@M;Dimethyl@Any N-term       2;0    8      0   \n",
       "3                                  Dimethyl@Any N-term         0    8      0   \n",
       "4                                  Dimethyl@Any N-term         0    8      0   \n",
       "..                                                 ...       ...  ...    ...   \n",
       "163  Oxidation@M;Dimethyl:2H(6)13C(2)@Any N-term;Di...     7;0;5   13      0   \n",
       "164  Oxidation@M;Dimethyl:2H(6)13C(2)@Any N-term;Di...     7;0;5   13      0   \n",
       "165  Dimethyl:2H(6)13C(2)@Any N-term;Dimethyl:2H(6)...       0;5   13      0   \n",
       "166  Dimethyl:2H(6)13C(2)@Any N-term;Dimethyl:2H(6)...       0;5   13      0   \n",
       "167  Dimethyl:2H(6)13C(2)@Any N-term;Dimethyl:2H(6)...       0;5   13      0   \n",
       "\n",
       "     charge label_channel  precursor_mz  isotope_intensity_m1  \\\n",
       "0         2         light    495.755484              0.500906   \n",
       "1         3         light    330.839415              0.500906   \n",
       "2         4         light    248.381380              0.500906   \n",
       "3         2         light    487.758027              0.500525   \n",
       "4         3         light    325.507777              0.500525   \n",
       "..      ...           ...           ...                   ...   \n",
       "163       3         heavy    539.655367              0.788273   \n",
       "164       4         heavy    404.993344              0.788273   \n",
       "165       2         heavy    800.981955              0.787646   \n",
       "166       3         heavy    534.323729              0.787646   \n",
       "167       4         heavy    400.994616              0.787646   \n",
       "\n",
       "     isotope_intensity_m2  isotope_apex_intensity  isotope_apex_index  \\\n",
       "0                0.194451                     1.0                   0   \n",
       "1                0.194451                     1.0                   0   \n",
       "2                0.194451                     1.0                   0   \n",
       "3                0.192205                     1.0                   0   \n",
       "4                0.192205                     1.0                   0   \n",
       "..                    ...                     ...                 ...   \n",
       "163              0.392103                     1.0                   0   \n",
       "164              0.392103                     1.0                   0   \n",
       "165              0.389779                     1.0                   0   \n",
       "166              0.389779                     1.0                   0   \n",
       "167              0.389779                     1.0                   0   \n",
       "\n",
       "     isotope_mz_m1  isotope_mz_m2  isotope_apex_mz  \n",
       "0       496.257134     496.758784       495.755484  \n",
       "1       331.173848     331.508282       330.839415  \n",
       "2       248.632205     248.883030       248.381380  \n",
       "3       488.259677     488.761327       487.758027  \n",
       "4       325.842210     326.176643       325.507777  \n",
       "..             ...            ...              ...  \n",
       "163     539.989801     540.324234       539.655367  \n",
       "164     405.244169     405.494994       404.993344  \n",
       "165     801.483605     801.985255       800.981955  \n",
       "166     534.658162     534.992596       534.323729  \n",
       "167     401.245441     401.496266       400.994616  \n",
       "\n",
       "[168 rows x 19 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_lib.import_and_process_protein_dict(protein_dict)\n",
    "_lib.add_peptide_labeling({\n",
    "    'light': ['Dimethyl@Any N-term','Dimethyl@K'],\n",
    "    'heavy': ['Dimethyl:2H(6)13C(2)@Any N-term','Dimethyl:2H(6)13C(2)@K'],\n",
    "})\n",
    "_lib.calc_precursor_isotope()\n",
    "assert (_lib.precursor_df.decoy==1).any()\n",
    "assert ('MACDESTY'[::-1]+'K') in _lib.precursor_df.sequence.values\n",
    "assert 'isotope_apex_index' in _lib.precursor_df.columns\n",
    "assert 'isotope_apex_intensity' in _lib.precursor_df.columns\n",
    "assert ~_lib.precursor_df.sequence.str.contains('B').any()\n",
    "_lib.precursor_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.3 ('base')",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
