---

title: Fasta processing and peptide generation


keywords: fastai
sidebar: home_sidebar



nb_path: "nbdev_nbs/protein/fasta.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: nbdev_nbs/protein/fasta.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_uniprot_gene_name" class="doc_header"><code>get_uniprot_gene_name</code><a href="https://github.com/MannLabs/alphabase/tree/main/alphabase/protein/fasta.py#L39" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_uniprot_gene_name</code>(<strong><code>description</code></strong>:<code>str</code>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="read_fasta_file" class="doc_header"><code>read_fasta_file</code><a href="https://github.com/MannLabs/alphabase/tree/main/alphabase/protein/fasta.py#L45" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>read_fasta_file</code>(<strong><code>fasta_filename</code></strong>:<code>str</code>=<em><code>''</code></em>)</p>
</blockquote>
<p>Read a FASTA file line by line
Args:
    fasta_filename (str): fasta.
Yields:
    dict {id:str, name:str, description:str, sequence:str}: protein information.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="load_all_proteins" class="doc_header"><code>load_all_proteins</code><a href="https://github.com/MannLabs/alphabase/tree/main/alphabase/protein/fasta.py#L76" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>load_all_proteins</code>(<strong><code>fasta_file_list</code></strong>:<code>list</code>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="concat_proteins" class="doc_header"><code>concat_proteins</code><a href="https://github.com/MannLabs/alphabase/tree/main/alphabase/protein/fasta.py#L83" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>concat_proteins</code>(<strong><code>protein_dict</code></strong>:<code>dict</code>, <strong><code>sep</code></strong>=<em><code>'$'</code></em>)</p>
</blockquote>
<p>Concatenate all protein sequences into a single sequence,
seperated by <code>sep ($ by default)</code>.</p>
<p>Args:
    protein_dict (dict): protein_dict by read_fasta_file()</p>
<p>Returns:
    str: concatenated sequence seperated by <code>sep</code>.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="cleave_sequence_with_cut_pos" class="doc_header"><code>cleave_sequence_with_cut_pos</code><a href="https://github.com/MannLabs/alphabase/tree/main/alphabase/protein/fasta.py#L103" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>cleave_sequence_with_cut_pos</code>(<strong><code>sequence</code></strong>:<code>str</code>, <strong><code>cut_pos</code></strong>:<code>ndarray</code>, <strong><code>n_missed_cleavages</code></strong>:<code>int</code>=<em><code>2</code></em>, <strong><code>pep_length_min</code></strong>:<code>int</code>=<em><code>6</code></em>, <strong><code>pep_length_max</code></strong>:<code>int</code>=<em><code>45</code></em>)</p>
</blockquote>
<p>Cleave a sequence with cut postions (cut_pos).
Filters to have a minimum and maximum length.
Args:
    sequence (str): protein sequence
    cut_pos (np.array): cut postions determined by a given protease.
    n_missed_cleavages (int): the number of max missed cleavages.
    pep_length_min (int): min peptide length.
    pep_length_max (int): max peptide length.
Returns:
    list (str): cleaved peptide sequences with missed cleavages.
    list (int): number of miss cleavage of each peptide.
    list (bool): if N-term peptide
    list (bool): if C-term pepetide</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h2 id="Digest" class="doc_header"><code>class</code> <code>Digest</code><a href="https://github.com/MannLabs/alphabase/tree/main/alphabase/protein/fasta.py#L151" class="source_link" style="float:right">[source]</a></h2><blockquote><p><code>Digest</code>(<strong><code>protease</code></strong>:<code>str</code>=<em><code>'trypsin'</code></em>, <strong><code>max_missed_cleavages</code></strong>:<code>int</code>=<em><code>2</code></em>, <strong><code>peptide_length_min</code></strong>:<code>int</code>=<em><code>6</code></em>, <strong><code>peptide_length_max</code></strong>:<code>int</code>=<em><code>45</code></em>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_fix_mods" class="doc_header"><code>get_fix_mods</code><a href="https://github.com/MannLabs/alphabase/tree/main/alphabase/protein/fasta.py#L213" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_fix_mods</code>(<strong><code>sequence</code></strong>:<code>str</code>, <strong><code>fix_mod_aas</code></strong>:<code>str</code>, <strong><code>fix_mod_dict</code></strong>:<code>dict</code>)</p>
</blockquote>
<p>Generate fix modifications for the sequence</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">seq</span> <span class="o">=</span> <span class="s1">&#39;ACBCDCK&#39;</span>
<span class="n">_fix_mod_dict</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">_fix_mod_dict</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;mod@C&#39;</span>
<span class="n">mods</span><span class="p">,</span> <span class="n">mod_sites</span> <span class="o">=</span> <span class="n">get_fix_mods</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="n">_fix_mod_dict</span><span class="p">)</span>
<span class="k">assert</span> <span class="n">mods</span><span class="o">==</span><span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s1">&#39;mod@C&#39;</span><span class="p">]</span><span class="o">*</span><span class="mi">3</span><span class="p">)</span>
<span class="k">assert</span> <span class="n">mod_sites</span><span class="o">==</span><span class="s1">&#39;2;4;6&#39;</span>
<span class="n">get_fix_mods</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="n">_fix_mod_dict</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>(&#39;mod@C;mod@C;mod@C&#39;, &#39;2;4;6&#39;)</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_candidate_sites" class="doc_header"><code>get_candidate_sites</code><a href="https://github.com/MannLabs/alphabase/tree/main/alphabase/protein/fasta.py#L230" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_candidate_sites</code>(<strong><code>sequence</code></strong>:<code>str</code>, <strong><code>target_mod_aas</code></strong>:<code>str</code>)</p>
</blockquote>
<p>get candidate modification sites</p>
<p>Args:
    sequence (str): peptide sequence
    target_mod_aas (str): AAs that may have modifications</p>
<p>Returns:
    list: candiadte mod sites in alphabase format (0: N-term, -1: C-term, 1-n:others)</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_var_mod_sites" class="doc_header"><code>get_var_mod_sites</code><a href="https://github.com/MannLabs/alphabase/tree/main/alphabase/protein/fasta.py#L248" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_var_mod_sites</code>(<strong><code>sequence</code></strong>:<code>str</code>, <strong><code>target_mod_aas</code></strong>:<code>str</code>, <strong><code>max_var_mod</code></strong>:<code>int</code>, <strong><code>max_combs</code></strong>:<code>int</code>)</p>
</blockquote>
<p>get all combinations of variable modification sites</p>
<p>Args:
    sequence (str): peptide sequence
    target_mod_aas (str): AAs that may have modifications
    max_var_mod (int): max number of mods in a sequence
    max_combs (int): max number of combinations for a sequence</p>
<p>Returns:
    list: list of combinations of (tuple) modification sites</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">seq</span> <span class="o">=</span> <span class="s1">&#39;AMCMSTYK&#39;</span>
<span class="n">candidate_sites</span> <span class="o">=</span> <span class="n">get_candidate_sites</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="s1">&#39;MSTY&#39;</span><span class="p">)</span>
<span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">candidate_sites</span><span class="p">)</span><span class="o">==</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">7</span><span class="p">]))</span>
<span class="n">get_var_mod_sites</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="s1">&#39;MSTY&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[(2,),
 (4,),
 (5,),
 (6,),
 (7,),
 (2, 4),
 (2, 5),
 (2, 6),
 (2, 7),
 (4, 5),
 (4, 6),
 (4, 7),
 (5, 6),
 (5, 7),
 (6, 7),
 (2, 4, 5),
 (2, 4, 6),
 (2, 4, 7),
 (2, 5, 6),
 (2, 5, 7)]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_var_mods_per_sites_multi_mods_on_aa" class="doc_header"><code>get_var_mods_per_sites_multi_mods_on_aa</code><a href="https://github.com/MannLabs/alphabase/tree/main/alphabase/protein/fasta.py#L283" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_var_mods_per_sites_multi_mods_on_aa</code>(<strong><code>sequence</code></strong>:<code>str</code>, <strong><code>mod_sites</code></strong>:<code>tuple</code>, <strong><code>var_mod_dict</code></strong>:<code>dict</code>)</p>
</blockquote>
<p>Used only when the var mod list contains
more than one mods on the same AA, for example:
Mod1@A, Mod2@A ...</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_var_mods_per_sites_single_mod_on_aa" class="doc_header"><code>get_var_mods_per_sites_single_mod_on_aa</code><a href="https://github.com/MannLabs/alphabase/tree/main/alphabase/protein/fasta.py#L308" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_var_mods_per_sites_single_mod_on_aa</code>(<strong><code>sequence</code></strong>:<code>str</code>, <strong><code>mod_sites</code></strong>:<code>tuple</code>, <strong><code>var_mod_dict</code></strong>:<code>dict</code>)</p>
</blockquote>
<p>Used when the var mod list contains
only one mods on the each AA, for example:
Mod1@A, Mod2@D ...</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_var_mods" class="doc_header"><code>get_var_mods</code><a href="https://github.com/MannLabs/alphabase/tree/main/alphabase/protein/fasta.py#L325" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_var_mods</code>(<strong><code>sequence</code></strong>:<code>str</code>, <strong><code>var_mod_aas</code></strong>:<code>str</code>, <strong><code>mod_dict</code></strong>:<code>dict</code>, <strong><code>max_var_mod</code></strong>:<code>int</code>, <strong><code>max_combs</code></strong>:<code>int</code>, <strong><code>keep_unmodified</code></strong>:<code>bool</code>=<em><code>False</code></em>)</p>
</blockquote>
<p>Generate all modification combinations and associated sites
for the sequence.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="parse_term_mod" class="doc_header"><code>parse_term_mod</code><a href="https://github.com/MannLabs/alphabase/tree/main/alphabase/protein/fasta.py#L356" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>parse_term_mod</code>(<strong><code>term_mod_name</code></strong>:<code>str</code>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="add_single_peptide_labeling" class="doc_header"><code>add_single_peptide_labeling</code><a href="https://github.com/MannLabs/alphabase/tree/main/alphabase/protein/fasta.py#L364" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>add_single_peptide_labeling</code>(<strong><code>seq</code></strong>:<code>str</code>, <strong><code>mods</code></strong>:<code>str</code>, <strong><code>mod_sites</code></strong>:<code>str</code>, <strong><code>label_aas</code></strong>:<code>str</code>, <strong><code>label_mod_dict</code></strong>:<code>dict</code>, <strong><code>nterm_label_mod</code></strong>:<code>str</code>, <strong><code>cterm_label_mod</code></strong>:<code>str</code>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="parse_labels" class="doc_header"><code>parse_labels</code><a href="https://github.com/MannLabs/alphabase/tree/main/alphabase/protein/fasta.py#L396" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>parse_labels</code>(<strong><code>labels</code></strong>:<code>list</code>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="create_labeling_peptide_df" class="doc_header"><code>create_labeling_peptide_df</code><a href="https://github.com/MannLabs/alphabase/tree/main/alphabase/protein/fasta.py#L412" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>create_labeling_peptide_df</code>(<strong><code>peptide_df</code></strong>:<code>DataFrame</code>, <strong><code>labels</code></strong>:<code>list</code>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="protein_idxes_to_names" class="doc_header"><code>protein_idxes_to_names</code><a href="https://github.com/MannLabs/alphabase/tree/main/alphabase/protein/fasta.py#L434" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>protein_idxes_to_names</code>(<strong><code>protein_idxes</code></strong>:<code>str</code>, <strong><code>protein_names</code></strong>:<code>list</code>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="append_regular_modifications" class="doc_header"><code>append_regular_modifications</code><a href="https://github.com/MannLabs/alphabase/tree/main/alphabase/protein/fasta.py#L441" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>append_regular_modifications</code>(<strong><code>df</code></strong>:<code>DataFrame</code>, <strong><code>var_mods</code></strong>=<em><code>['Phospho@S', 'Phospho@T', 'Phospho@Y']</code></em>, <strong><code>max_mod_num</code></strong>=<em><code>1</code></em>, <strong><code>max_combs</code></strong>=<em><code>100</code></em>, <strong><code>keep_unmodified</code></strong>=<em><code>True</code></em>)</p>
</blockquote>
<p>Append regular (not N/C-term) variable modifications to the
exsiting modifications of each sequence in <code>df</code>.</p>
<p>Args:
    df (pd.DataFrame): Precursor dataframe
    var_mods (list, optional): Considered varialbe modification list.
        Defaults to ['Phospho@S','Phospho@T','Phospho@Y'].
    max_mod_num (int, optional): Maximal modification number for
        each sequence of the <code>var_mods</code>. Defaults to 1.
    max_combs (int, optional): One sequence is only allowed to explode
        to <code>max_combs</code> number of modified peptides. Defaults to 100.
    keep_unmodified (bool, optional): If unmodified (only refered to <code>var_mods</code>)
    peptides are also remained in the returned dataframe. Defaults to True.</p>
<p>Returns:
    pd.DataFrame: The precursor_df with <code>var_mods</code> appended.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h2 id="FastaLib" class="doc_header"><code>class</code> <code>FastaLib</code><a href="https://github.com/MannLabs/alphabase/tree/main/alphabase/protein/fasta.py#L496" class="source_link" style="float:right">[source]</a></h2><blockquote><p><code>FastaLib</code>(<strong><code>charged_frag_types</code></strong>:<code>list</code>=<em><code>['b_z1', 'b_z2', 'y_z1', 'y_z2']</code></em>, <strong><code>protease</code></strong>:<code>str</code>=<em><code>'trypsin'</code></em>, <strong><code>max_missed_cleavages</code></strong>:<code>int</code>=<em><code>2</code></em>, <strong><code>peptide_length_min</code></strong>:<code>int</code>=<em><code>7</code></em>, <strong><code>peptide_length_max</code></strong>:<code>int</code>=<em><code>35</code></em>, <strong><code>precursor_charge_min</code></strong>:<code>int</code>=<em><code>2</code></em>, <strong><code>precursor_charge_max</code></strong>:<code>int</code>=<em><code>4</code></em>, <strong><code>precursor_mz_min</code></strong>:<code>float</code>=<em><code>200.0</code></em>, <strong><code>precursor_mz_max</code></strong>:<code>float</code>=<em><code>2000.0</code></em>, <strong><code>var_mods</code></strong>:<code>list</code>=<em><code>['Acetyl@Protein N-term', 'Oxidation@M']</code></em>, <strong><code>max_var_mod_num</code></strong>:<code>int</code>=<em><code>2</code></em>, <strong><code>fix_mods</code></strong>:<code>list</code>=<em><code>['Carbamidomethyl@C']</code></em>, <strong><code>decoy</code></strong>:<code>str</code>=<em><code>None</code></em>, <strong><code>I_to_L</code></strong>=<em><code>False</code></em>) :: <a href="/alphabase/library_base.html#SpecLibBase"><code>SpecLibBase</code></a></p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="p">{</span>
        <span class="s1">&#39;sequence&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;ABSTY&#39;</span><span class="p">,</span><span class="s1">&#39;ACXSX&#39;</span><span class="p">,</span><span class="s1">&#39;ACDEFG&#39;</span><span class="p">],</span>
        <span class="s1">&#39;mods&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;Acetyl@Protein N-term&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">],</span>
        <span class="s1">&#39;mod_sites&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;0&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">]</span>
    <span class="p">}</span>
<span class="p">)</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">append_regular_modifications</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">keep_unmodified</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">sequence</span><span class="o">==</span><span class="s1">&#39;ABSTY&#39;</span><span class="p">)</span><span class="o">==</span><span class="mi">4</span>
<span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">sequence</span><span class="o">==</span><span class="s1">&#39;ACXSX&#39;</span><span class="p">)</span><span class="o">==</span><span class="mi">2</span>
<span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">sequence</span><span class="o">==</span><span class="s1">&#39;ACDEFG&#39;</span><span class="p">)</span><span class="o">==</span><span class="mi">1</span>
<span class="k">assert</span> <span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">sequence</span><span class="o">==</span><span class="s1">&#39;ABSTY&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mods</span><span class="o">.</span><span class="n">values</span> <span class="o">==</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="s1">&#39;Phospho@S&#39;</span><span class="p">,</span><span class="s1">&#39;Phospho@T&#39;</span><span class="p">,</span><span class="s1">&#39;Phospho@Y&#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">]))</span>
<span class="k">assert</span> <span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">sequence</span><span class="o">==</span><span class="s1">&#39;ABSTY&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mod_sites</span><span class="o">.</span><span class="n">values</span> <span class="o">==</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="s1">&#39;3&#39;</span><span class="p">,</span><span class="s1">&#39;4&#39;</span><span class="p">,</span><span class="s1">&#39;5&#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">]))</span>
<span class="k">assert</span> <span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">sequence</span><span class="o">==</span><span class="s1">&#39;ACXSX&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mods</span><span class="o">.</span><span class="n">values</span> <span class="o">==</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="s1">&#39;Acetyl@Protein N-term;Phospho@S&#39;</span><span class="p">,</span><span class="s1">&#39;Acetyl@Protein N-term&#39;</span><span class="p">]))</span>
<span class="k">assert</span> <span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">sequence</span><span class="o">==</span><span class="s1">&#39;ACXSX&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mod_sites</span><span class="o">.</span><span class="n">values</span> <span class="o">==</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="s1">&#39;0;4&#39;</span><span class="p">,</span><span class="s1">&#39;0&#39;</span><span class="p">]))</span>
<span class="n">df</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sequence</th>
      <th>mods</th>
      <th>mod_sites</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>ABSTY</td>
      <td>Phospho@S</td>
      <td>3</td>
    </tr>
    <tr>
      <th>1</th>
      <td>ABSTY</td>
      <td>Phospho@T</td>
      <td>4</td>
    </tr>
    <tr>
      <th>2</th>
      <td>ABSTY</td>
      <td>Phospho@Y</td>
      <td>5</td>
    </tr>
    <tr>
      <th>3</th>
      <td>ABSTY</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>4</th>
      <td>ACXSX</td>
      <td>Acetyl@Protein N-term;Phospho@S</td>
      <td>0;4</td>
    </tr>
    <tr>
      <th>5</th>
      <td>ACXSX</td>
      <td>Acetyl@Protein N-term</td>
      <td>0</td>
    </tr>
    <tr>
      <th>6</th>
      <td>ACDEFG</td>
      <td></td>
      <td></td>
    </tr>
  </tbody>
</table>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="p">{</span>
        <span class="s1">&#39;sequence&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;ABSTY&#39;</span><span class="p">,</span><span class="s1">&#39;ACXSX&#39;</span><span class="p">,</span><span class="s1">&#39;ACDEFG&#39;</span><span class="p">],</span>
        <span class="s1">&#39;mods&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;Acetyl@Protein N-term&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">],</span>
        <span class="s1">&#39;mod_sites&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;0&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">]</span>
    <span class="p">}</span>
<span class="p">)</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">append_regular_modifications</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">keep_unmodified</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">sequence</span><span class="o">==</span><span class="s1">&#39;ABSTY&#39;</span><span class="p">)</span><span class="o">==</span><span class="mi">3</span>
<span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">sequence</span><span class="o">==</span><span class="s1">&#39;ACXSX&#39;</span><span class="p">)</span><span class="o">==</span><span class="mi">1</span>
<span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">sequence</span><span class="o">==</span><span class="s1">&#39;ACDEFG&#39;</span><span class="p">)</span><span class="o">==</span><span class="mi">0</span>
<span class="n">df</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sequence</th>
      <th>mods</th>
      <th>mod_sites</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>ABSTY</td>
      <td>Phospho@S</td>
      <td>3</td>
    </tr>
    <tr>
      <th>1</th>
      <td>ABSTY</td>
      <td>Phospho@T</td>
      <td>4</td>
    </tr>
    <tr>
      <th>2</th>
      <td>ABSTY</td>
      <td>Phospho@Y</td>
      <td>5</td>
    </tr>
    <tr>
      <th>3</th>
      <td>ACXSX</td>
      <td>Acetyl@Protein N-term;Phospho@S</td>
      <td>0;4</td>
    </tr>
  </tbody>
</table>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">_lib</span> <span class="o">=</span> <span class="n">FastaLib</span><span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="n">I_to_L</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">decoy</span><span class="o">=</span><span class="s1">&#39;pseudo_reverse&#39;</span><span class="p">)</span>
<span class="n">prot1</span> <span class="o">=</span> <span class="s1">&#39;MABCDEKFGHIJKLMNOPQRST&#39;</span>
<span class="n">prot2</span> <span class="o">=</span> <span class="s1">&#39;FGHIJKLMNOPQR&#39;</span>
<span class="n">protein_dict</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;xx&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;protein_id&#39;</span><span class="p">:</span> <span class="s1">&#39;xx&#39;</span><span class="p">,</span>
        <span class="s1">&#39;gene_name&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
        <span class="s1">&#39;sequence&#39;</span><span class="p">:</span> <span class="n">prot1</span>
    <span class="p">},</span>
    <span class="s1">&#39;yy&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;protein_id&#39;</span><span class="p">:</span> <span class="s1">&#39;yy&#39;</span><span class="p">,</span>
        <span class="s1">&#39;gene_name&#39;</span><span class="p">:</span> <span class="s1">&#39;gene&#39;</span><span class="p">,</span>
        <span class="s1">&#39;sequence&#39;</span><span class="p">:</span> <span class="n">prot2</span>
    <span class="p">}</span>
<span class="p">}</span>
<span class="n">_lib</span><span class="o">.</span><span class="n">get_peptides_from_protein_dict</span><span class="p">(</span><span class="n">protein_dict</span><span class="p">)</span>
<span class="n">_lib</span><span class="o">.</span><span class="n">precursor_df</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sequence</th>
      <th>protein_idxes</th>
      <th>miss_cleavage</th>
      <th>is_prot_nterm</th>
      <th>is_prot_cterm</th>
      <th>mods</th>
      <th>mod_sites</th>
      <th>nAA</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>MABCDEK</td>
      <td>0</td>
      <td>0</td>
      <td>True</td>
      <td>False</td>
      <td></td>
      <td></td>
      <td>7</td>
    </tr>
    <tr>
      <th>1</th>
      <td>LMNOPQR</td>
      <td>0;1</td>
      <td>0</td>
      <td>False</td>
      <td>True</td>
      <td></td>
      <td></td>
      <td>7</td>
    </tr>
    <tr>
      <th>2</th>
      <td>LMNOPQRST</td>
      <td>0</td>
      <td>1</td>
      <td>False</td>
      <td>True</td>
      <td></td>
      <td></td>
      <td>9</td>
    </tr>
    <tr>
      <th>3</th>
      <td>ABCDEKFGHIJK</td>
      <td>0</td>
      <td>1</td>
      <td>True</td>
      <td>False</td>
      <td></td>
      <td></td>
      <td>12</td>
    </tr>
    <tr>
      <th>4</th>
      <td>MABCDEKFGHIJK</td>
      <td>0</td>
      <td>1</td>
      <td>True</td>
      <td>False</td>
      <td></td>
      <td></td>
      <td>13</td>
    </tr>
    <tr>
      <th>5</th>
      <td>FGHIJKLMNOPQR</td>
      <td>0;1</td>
      <td>1</td>
      <td>True</td>
      <td>True</td>
      <td></td>
      <td></td>
      <td>13</td>
    </tr>
    <tr>
      <th>6</th>
      <td>FGHIJKLMNOPQRST</td>
      <td>0</td>
      <td>2</td>
      <td>False</td>
      <td>True</td>
      <td></td>
      <td></td>
      <td>15</td>
    </tr>
    <tr>
      <th>7</th>
      <td>ABCDEKFGHIJKLMNOPQR</td>
      <td>0</td>
      <td>2</td>
      <td>True</td>
      <td>False</td>
      <td></td>
      <td></td>
      <td>19</td>
    </tr>
    <tr>
      <th>8</th>
      <td>MABCDEKFGHIJKLMNOPQR</td>
      <td>0</td>
      <td>2</td>
      <td>True</td>
      <td>False</td>
      <td></td>
      <td></td>
      <td>20</td>
    </tr>
  </tbody>
</table>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">_lib</span><span class="o">.</span><span class="n">protein_df</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>protein_id</th>
      <th>gene_name</th>
      <th>sequence</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>xx</td>
      <td></td>
      <td>MABCDEKFGHIJKLMNOPQRST</td>
    </tr>
    <tr>
      <th>1</th>
      <td>yy</td>
      <td>gene</td>
      <td>FGHIJKLMNOPQR</td>
    </tr>
  </tbody>
</table>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">_lib</span><span class="o">.</span><span class="n">append_protein_name</span><span class="p">()</span>
<span class="k">assert</span> <span class="s1">&#39;proteins&#39;</span> <span class="ow">in</span> <span class="n">_lib</span><span class="o">.</span><span class="n">precursor_df</span><span class="o">.</span><span class="n">columns</span>
<span class="n">_lib</span><span class="o">.</span><span class="n">precursor_df</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sequence</th>
      <th>protein_idxes</th>
      <th>miss_cleavage</th>
      <th>is_prot_nterm</th>
      <th>is_prot_cterm</th>
      <th>mods</th>
      <th>mod_sites</th>
      <th>nAA</th>
      <th>proteins</th>
      <th>genes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>MABCDEK</td>
      <td>0</td>
      <td>0</td>
      <td>True</td>
      <td>False</td>
      <td></td>
      <td></td>
      <td>7</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>1</th>
      <td>LMNOPQR</td>
      <td>0;1</td>
      <td>0</td>
      <td>False</td>
      <td>True</td>
      <td></td>
      <td></td>
      <td>7</td>
      <td>xx;yy</td>
      <td>gene</td>
    </tr>
    <tr>
      <th>2</th>
      <td>LMNOPQRST</td>
      <td>0</td>
      <td>1</td>
      <td>False</td>
      <td>True</td>
      <td></td>
      <td></td>
      <td>9</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>3</th>
      <td>ABCDEKFGHIJK</td>
      <td>0</td>
      <td>1</td>
      <td>True</td>
      <td>False</td>
      <td></td>
      <td></td>
      <td>12</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>4</th>
      <td>MABCDEKFGHIJK</td>
      <td>0</td>
      <td>1</td>
      <td>True</td>
      <td>False</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>5</th>
      <td>FGHIJKLMNOPQR</td>
      <td>0;1</td>
      <td>1</td>
      <td>True</td>
      <td>True</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>xx;yy</td>
      <td>gene</td>
    </tr>
    <tr>
      <th>6</th>
      <td>FGHIJKLMNOPQRST</td>
      <td>0</td>
      <td>2</td>
      <td>False</td>
      <td>True</td>
      <td></td>
      <td></td>
      <td>15</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>7</th>
      <td>ABCDEKFGHIJKLMNOPQR</td>
      <td>0</td>
      <td>2</td>
      <td>True</td>
      <td>False</td>
      <td></td>
      <td></td>
      <td>19</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>8</th>
      <td>MABCDEKFGHIJKLMNOPQR</td>
      <td>0</td>
      <td>2</td>
      <td>True</td>
      <td>False</td>
      <td></td>
      <td></td>
      <td>20</td>
      <td>xx</td>
      <td></td>
    </tr>
  </tbody>
</table>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">_lib</span><span class="o">.</span><span class="n">add_modifications</span><span class="p">()</span>
<span class="n">_lib</span><span class="o">.</span><span class="n">precursor_df</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sequence</th>
      <th>protein_idxes</th>
      <th>miss_cleavage</th>
      <th>is_prot_nterm</th>
      <th>is_prot_cterm</th>
      <th>mods</th>
      <th>mod_sites</th>
      <th>nAA</th>
      <th>proteins</th>
      <th>genes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>MABCDEK</td>
      <td>0</td>
      <td>0</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Oxidation@M</td>
      <td>4;1</td>
      <td>7</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>1</th>
      <td>MABCDEK</td>
      <td>0</td>
      <td>0</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C</td>
      <td>4</td>
      <td>7</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>2</th>
      <td>MABCDEK</td>
      <td>0</td>
      <td>0</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...</td>
      <td>4;0;1</td>
      <td>7</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>3</th>
      <td>MABCDEK</td>
      <td>0</td>
      <td>0</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Acetyl@Protein N-term</td>
      <td>4;0</td>
      <td>7</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>4</th>
      <td>LMNOPQR</td>
      <td>0;1</td>
      <td>0</td>
      <td>False</td>
      <td>True</td>
      <td>Oxidation@M</td>
      <td>2</td>
      <td>7</td>
      <td>xx;yy</td>
      <td>gene</td>
    </tr>
    <tr>
      <th>5</th>
      <td>LMNOPQR</td>
      <td>0;1</td>
      <td>0</td>
      <td>False</td>
      <td>True</td>
      <td></td>
      <td></td>
      <td>7</td>
      <td>xx;yy</td>
      <td>gene</td>
    </tr>
    <tr>
      <th>6</th>
      <td>LMNOPQRST</td>
      <td>0</td>
      <td>1</td>
      <td>False</td>
      <td>True</td>
      <td>Oxidation@M</td>
      <td>2</td>
      <td>9</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>7</th>
      <td>LMNOPQRST</td>
      <td>0</td>
      <td>1</td>
      <td>False</td>
      <td>True</td>
      <td></td>
      <td></td>
      <td>9</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>8</th>
      <td>ABCDEKFGHIJK</td>
      <td>0</td>
      <td>1</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C</td>
      <td>3</td>
      <td>12</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>9</th>
      <td>ABCDEKFGHIJK</td>
      <td>0</td>
      <td>1</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Acetyl@Protein N-term</td>
      <td>3;0</td>
      <td>12</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>10</th>
      <td>MABCDEKFGHIJK</td>
      <td>0</td>
      <td>1</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Oxidation@M</td>
      <td>4;1</td>
      <td>13</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>11</th>
      <td>MABCDEKFGHIJK</td>
      <td>0</td>
      <td>1</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C</td>
      <td>4</td>
      <td>13</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>12</th>
      <td>MABCDEKFGHIJK</td>
      <td>0</td>
      <td>1</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...</td>
      <td>4;0;1</td>
      <td>13</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>13</th>
      <td>MABCDEKFGHIJK</td>
      <td>0</td>
      <td>1</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Acetyl@Protein N-term</td>
      <td>4;0</td>
      <td>13</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>14</th>
      <td>FGHIJKLMNOPQR</td>
      <td>0;1</td>
      <td>1</td>
      <td>True</td>
      <td>True</td>
      <td>Oxidation@M</td>
      <td>8</td>
      <td>13</td>
      <td>xx;yy</td>
      <td>gene</td>
    </tr>
    <tr>
      <th>15</th>
      <td>FGHIJKLMNOPQR</td>
      <td>0;1</td>
      <td>1</td>
      <td>True</td>
      <td>True</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>xx;yy</td>
      <td>gene</td>
    </tr>
    <tr>
      <th>16</th>
      <td>FGHIJKLMNOPQR</td>
      <td>0;1</td>
      <td>1</td>
      <td>True</td>
      <td>True</td>
      <td>Acetyl@Protein N-term;Oxidation@M</td>
      <td>0;8</td>
      <td>13</td>
      <td>xx;yy</td>
      <td>gene</td>
    </tr>
    <tr>
      <th>17</th>
      <td>FGHIJKLMNOPQR</td>
      <td>0;1</td>
      <td>1</td>
      <td>True</td>
      <td>True</td>
      <td>Acetyl@Protein N-term</td>
      <td>0</td>
      <td>13</td>
      <td>xx;yy</td>
      <td>gene</td>
    </tr>
    <tr>
      <th>18</th>
      <td>FGHIJKLMNOPQRST</td>
      <td>0</td>
      <td>2</td>
      <td>False</td>
      <td>True</td>
      <td>Oxidation@M</td>
      <td>8</td>
      <td>15</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>19</th>
      <td>FGHIJKLMNOPQRST</td>
      <td>0</td>
      <td>2</td>
      <td>False</td>
      <td>True</td>
      <td></td>
      <td></td>
      <td>15</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>20</th>
      <td>ABCDEKFGHIJKLMNOPQR</td>
      <td>0</td>
      <td>2</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Oxidation@M</td>
      <td>3;14</td>
      <td>19</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>21</th>
      <td>ABCDEKFGHIJKLMNOPQR</td>
      <td>0</td>
      <td>2</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C</td>
      <td>3</td>
      <td>19</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>22</th>
      <td>ABCDEKFGHIJKLMNOPQR</td>
      <td>0</td>
      <td>2</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...</td>
      <td>3;0;14</td>
      <td>19</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>23</th>
      <td>ABCDEKFGHIJKLMNOPQR</td>
      <td>0</td>
      <td>2</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Acetyl@Protein N-term</td>
      <td>3;0</td>
      <td>19</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>24</th>
      <td>MABCDEKFGHIJKLMNOPQR</td>
      <td>0</td>
      <td>2</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Oxidation@M</td>
      <td>4;1</td>
      <td>20</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>25</th>
      <td>MABCDEKFGHIJKLMNOPQR</td>
      <td>0</td>
      <td>2</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Oxidation@M</td>
      <td>4;15</td>
      <td>20</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>26</th>
      <td>MABCDEKFGHIJKLMNOPQR</td>
      <td>0</td>
      <td>2</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Oxidation@M;Oxidation@M</td>
      <td>4;1;15</td>
      <td>20</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>27</th>
      <td>MABCDEKFGHIJKLMNOPQR</td>
      <td>0</td>
      <td>2</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C</td>
      <td>4</td>
      <td>20</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>28</th>
      <td>MABCDEKFGHIJKLMNOPQR</td>
      <td>0</td>
      <td>2</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...</td>
      <td>4;0;1</td>
      <td>20</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>29</th>
      <td>MABCDEKFGHIJKLMNOPQR</td>
      <td>0</td>
      <td>2</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...</td>
      <td>4;0;15</td>
      <td>20</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>30</th>
      <td>MABCDEKFGHIJKLMNOPQR</td>
      <td>0</td>
      <td>2</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...</td>
      <td>4;0;1;15</td>
      <td>20</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>31</th>
      <td>MABCDEKFGHIJKLMNOPQR</td>
      <td>0</td>
      <td>2</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Acetyl@Protein N-term</td>
      <td>4;0</td>
      <td>20</td>
      <td>xx</td>
      <td></td>
    </tr>
  </tbody>
</table>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">_lib</span><span class="o">.</span><span class="n">add_additional_modifications</span><span class="p">([</span><span class="s1">&#39;Phospho@S&#39;</span><span class="p">,</span><span class="s1">&#39;Phospho@T&#39;</span><span class="p">])</span>
<span class="k">assert</span> <span class="n">_lib</span><span class="o">.</span><span class="n">precursor_df</span><span class="o">.</span><span class="n">mods</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;Phospho&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">()</span>
<span class="n">_lib</span><span class="o">.</span><span class="n">precursor_df</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sequence</th>
      <th>protein_idxes</th>
      <th>miss_cleavage</th>
      <th>is_prot_nterm</th>
      <th>is_prot_cterm</th>
      <th>mods</th>
      <th>mod_sites</th>
      <th>nAA</th>
      <th>proteins</th>
      <th>genes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>MABCDEK</td>
      <td>0</td>
      <td>0</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Oxidation@M</td>
      <td>4;1</td>
      <td>7</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>1</th>
      <td>MABCDEK</td>
      <td>0</td>
      <td>0</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C</td>
      <td>4</td>
      <td>7</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>2</th>
      <td>MABCDEK</td>
      <td>0</td>
      <td>0</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...</td>
      <td>4;0;1</td>
      <td>7</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>3</th>
      <td>MABCDEK</td>
      <td>0</td>
      <td>0</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Acetyl@Protein N-term</td>
      <td>4;0</td>
      <td>7</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>4</th>
      <td>LMNOPQR</td>
      <td>0;1</td>
      <td>0</td>
      <td>False</td>
      <td>True</td>
      <td>Oxidation@M</td>
      <td>2</td>
      <td>7</td>
      <td>xx;yy</td>
      <td>gene</td>
    </tr>
    <tr>
      <th>5</th>
      <td>LMNOPQR</td>
      <td>0;1</td>
      <td>0</td>
      <td>False</td>
      <td>True</td>
      <td></td>
      <td></td>
      <td>7</td>
      <td>xx;yy</td>
      <td>gene</td>
    </tr>
    <tr>
      <th>6</th>
      <td>LMNOPQRST</td>
      <td>0</td>
      <td>1</td>
      <td>False</td>
      <td>True</td>
      <td>Oxidation@M;Phospho@S</td>
      <td>2;8</td>
      <td>9</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>7</th>
      <td>LMNOPQRST</td>
      <td>0</td>
      <td>1</td>
      <td>False</td>
      <td>True</td>
      <td>Oxidation@M;Phospho@T</td>
      <td>2;9</td>
      <td>9</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>8</th>
      <td>LMNOPQRST</td>
      <td>0</td>
      <td>1</td>
      <td>False</td>
      <td>True</td>
      <td>Oxidation@M</td>
      <td>2</td>
      <td>9</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>9</th>
      <td>LMNOPQRST</td>
      <td>0</td>
      <td>1</td>
      <td>False</td>
      <td>True</td>
      <td>Phospho@S</td>
      <td>8</td>
      <td>9</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>10</th>
      <td>LMNOPQRST</td>
      <td>0</td>
      <td>1</td>
      <td>False</td>
      <td>True</td>
      <td>Phospho@T</td>
      <td>9</td>
      <td>9</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>11</th>
      <td>LMNOPQRST</td>
      <td>0</td>
      <td>1</td>
      <td>False</td>
      <td>True</td>
      <td></td>
      <td></td>
      <td>9</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>12</th>
      <td>ABCDEKFGHIJK</td>
      <td>0</td>
      <td>1</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C</td>
      <td>3</td>
      <td>12</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>13</th>
      <td>ABCDEKFGHIJK</td>
      <td>0</td>
      <td>1</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Acetyl@Protein N-term</td>
      <td>3;0</td>
      <td>12</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>14</th>
      <td>MABCDEKFGHIJK</td>
      <td>0</td>
      <td>1</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Oxidation@M</td>
      <td>4;1</td>
      <td>13</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>15</th>
      <td>MABCDEKFGHIJK</td>
      <td>0</td>
      <td>1</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C</td>
      <td>4</td>
      <td>13</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>16</th>
      <td>MABCDEKFGHIJK</td>
      <td>0</td>
      <td>1</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...</td>
      <td>4;0;1</td>
      <td>13</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>17</th>
      <td>MABCDEKFGHIJK</td>
      <td>0</td>
      <td>1</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Acetyl@Protein N-term</td>
      <td>4;0</td>
      <td>13</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>18</th>
      <td>FGHIJKLMNOPQR</td>
      <td>0;1</td>
      <td>1</td>
      <td>True</td>
      <td>True</td>
      <td>Oxidation@M</td>
      <td>8</td>
      <td>13</td>
      <td>xx;yy</td>
      <td>gene</td>
    </tr>
    <tr>
      <th>19</th>
      <td>FGHIJKLMNOPQR</td>
      <td>0;1</td>
      <td>1</td>
      <td>True</td>
      <td>True</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>xx;yy</td>
      <td>gene</td>
    </tr>
    <tr>
      <th>20</th>
      <td>FGHIJKLMNOPQR</td>
      <td>0;1</td>
      <td>1</td>
      <td>True</td>
      <td>True</td>
      <td>Acetyl@Protein N-term;Oxidation@M</td>
      <td>0;8</td>
      <td>13</td>
      <td>xx;yy</td>
      <td>gene</td>
    </tr>
    <tr>
      <th>21</th>
      <td>FGHIJKLMNOPQR</td>
      <td>0;1</td>
      <td>1</td>
      <td>True</td>
      <td>True</td>
      <td>Acetyl@Protein N-term</td>
      <td>0</td>
      <td>13</td>
      <td>xx;yy</td>
      <td>gene</td>
    </tr>
    <tr>
      <th>22</th>
      <td>FGHIJKLMNOPQRST</td>
      <td>0</td>
      <td>2</td>
      <td>False</td>
      <td>True</td>
      <td>Oxidation@M;Phospho@S</td>
      <td>8;14</td>
      <td>15</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>23</th>
      <td>FGHIJKLMNOPQRST</td>
      <td>0</td>
      <td>2</td>
      <td>False</td>
      <td>True</td>
      <td>Oxidation@M;Phospho@T</td>
      <td>8;15</td>
      <td>15</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>24</th>
      <td>FGHIJKLMNOPQRST</td>
      <td>0</td>
      <td>2</td>
      <td>False</td>
      <td>True</td>
      <td>Oxidation@M</td>
      <td>8</td>
      <td>15</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>25</th>
      <td>FGHIJKLMNOPQRST</td>
      <td>0</td>
      <td>2</td>
      <td>False</td>
      <td>True</td>
      <td>Phospho@S</td>
      <td>14</td>
      <td>15</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>26</th>
      <td>FGHIJKLMNOPQRST</td>
      <td>0</td>
      <td>2</td>
      <td>False</td>
      <td>True</td>
      <td>Phospho@T</td>
      <td>15</td>
      <td>15</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>27</th>
      <td>FGHIJKLMNOPQRST</td>
      <td>0</td>
      <td>2</td>
      <td>False</td>
      <td>True</td>
      <td></td>
      <td></td>
      <td>15</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>28</th>
      <td>ABCDEKFGHIJKLMNOPQR</td>
      <td>0</td>
      <td>2</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Oxidation@M</td>
      <td>3;14</td>
      <td>19</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>29</th>
      <td>ABCDEKFGHIJKLMNOPQR</td>
      <td>0</td>
      <td>2</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C</td>
      <td>3</td>
      <td>19</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>30</th>
      <td>ABCDEKFGHIJKLMNOPQR</td>
      <td>0</td>
      <td>2</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...</td>
      <td>3;0;14</td>
      <td>19</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>31</th>
      <td>ABCDEKFGHIJKLMNOPQR</td>
      <td>0</td>
      <td>2</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Acetyl@Protein N-term</td>
      <td>3;0</td>
      <td>19</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>32</th>
      <td>MABCDEKFGHIJKLMNOPQR</td>
      <td>0</td>
      <td>2</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Oxidation@M</td>
      <td>4;1</td>
      <td>20</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>33</th>
      <td>MABCDEKFGHIJKLMNOPQR</td>
      <td>0</td>
      <td>2</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Oxidation@M</td>
      <td>4;15</td>
      <td>20</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>34</th>
      <td>MABCDEKFGHIJKLMNOPQR</td>
      <td>0</td>
      <td>2</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Oxidation@M;Oxidation@M</td>
      <td>4;1;15</td>
      <td>20</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>35</th>
      <td>MABCDEKFGHIJKLMNOPQR</td>
      <td>0</td>
      <td>2</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C</td>
      <td>4</td>
      <td>20</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>36</th>
      <td>MABCDEKFGHIJKLMNOPQR</td>
      <td>0</td>
      <td>2</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...</td>
      <td>4;0;1</td>
      <td>20</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>37</th>
      <td>MABCDEKFGHIJKLMNOPQR</td>
      <td>0</td>
      <td>2</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...</td>
      <td>4;0;15</td>
      <td>20</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>38</th>
      <td>MABCDEKFGHIJKLMNOPQR</td>
      <td>0</td>
      <td>2</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Acetyl@Protein N-term;Oxidat...</td>
      <td>4;0;1;15</td>
      <td>20</td>
      <td>xx</td>
      <td></td>
    </tr>
    <tr>
      <th>39</th>
      <td>MABCDEKFGHIJKLMNOPQR</td>
      <td>0</td>
      <td>2</td>
      <td>True</td>
      <td>False</td>
      <td>Carbamidomethyl@C;Acetyl@Protein N-term</td>
      <td>4;0</td>
      <td>20</td>
      <td>xx</td>
      <td></td>
    </tr>
  </tbody>
</table>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">_lib</span> <span class="o">=</span> <span class="n">FastaLib</span><span class="p">(</span>
    <span class="p">[</span><span class="s1">&#39;b_z1&#39;</span><span class="p">,</span><span class="s1">&#39;y_z1&#39;</span><span class="p">],</span> <span class="n">I_to_L</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> 
    <span class="n">decoy</span><span class="o">=</span><span class="s1">&#39;pseudo_reverse&#39;</span>
<span class="p">)</span>
<span class="n">prot1</span> <span class="o">=</span> <span class="s1">&#39;MACDESTYKBKFGHIKLMNPQRST&#39;</span>
<span class="n">prot2</span> <span class="o">=</span> <span class="s1">&#39;FGHIKLMNPQR&#39;</span>
<span class="n">protein_dict</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;xx&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;protein_id&#39;</span><span class="p">:</span> <span class="s1">&#39;xx&#39;</span><span class="p">,</span>
        <span class="s1">&#39;sequence&#39;</span><span class="p">:</span> <span class="n">prot1</span>
    <span class="p">},</span>
    <span class="s1">&#39;yy&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;protein_id&#39;</span><span class="p">:</span> <span class="s1">&#39;yy&#39;</span><span class="p">,</span>
        <span class="s1">&#39;sequence&#39;</span><span class="p">:</span> <span class="n">prot2</span>
    <span class="p">}</span>
<span class="p">}</span>
<span class="n">_lib</span><span class="o">.</span><span class="n">import_and_process_protein_dict</span><span class="p">(</span><span class="n">protein_dict</span><span class="p">)</span>
<span class="n">_lib</span><span class="o">.</span><span class="n">calc_precursor_isotope</span><span class="p">()</span>
<span class="k">assert</span> <span class="p">(</span><span class="n">_lib</span><span class="o">.</span><span class="n">precursor_df</span><span class="o">.</span><span class="n">charge</span> <span class="o">==</span> <span class="n">_lib</span><span class="o">.</span><span class="n">min_precursor_charge</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">()</span>
<span class="k">assert</span> <span class="p">(</span><span class="n">_lib</span><span class="o">.</span><span class="n">precursor_df</span><span class="o">.</span><span class="n">charge</span> <span class="o">==</span> <span class="n">_lib</span><span class="o">.</span><span class="n">max_precursor_charge</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">()</span>
<span class="k">assert</span> <span class="p">(</span><span class="n">_lib</span><span class="o">.</span><span class="n">precursor_df</span><span class="o">.</span><span class="n">decoy</span><span class="o">==</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">()</span>
<span class="k">assert</span> <span class="p">(</span><span class="s1">&#39;MACDESTY&#39;</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;K&#39;</span><span class="p">)</span> <span class="ow">in</span> <span class="n">_lib</span><span class="o">.</span><span class="n">precursor_df</span><span class="o">.</span><span class="n">sequence</span><span class="o">.</span><span class="n">values</span>
<span class="k">assert</span> <span class="s1">&#39;isotope_apex_index&#39;</span> <span class="ow">in</span> <span class="n">_lib</span><span class="o">.</span><span class="n">precursor_df</span><span class="o">.</span><span class="n">columns</span>
<span class="k">assert</span> <span class="s1">&#39;isotope_apex_intensity&#39;</span> <span class="ow">in</span> <span class="n">_lib</span><span class="o">.</span><span class="n">precursor_df</span><span class="o">.</span><span class="n">columns</span>
<span class="k">assert</span> <span class="o">~</span><span class="n">_lib</span><span class="o">.</span><span class="n">precursor_df</span><span class="o">.</span><span class="n">sequence</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;B&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">()</span>
<span class="n">_lib</span><span class="o">.</span><span class="n">precursor_df</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sequence</th>
      <th>protein_idxes</th>
      <th>miss_cleavage</th>
      <th>is_prot_nterm</th>
      <th>is_prot_cterm</th>
      <th>mods</th>
      <th>mod_sites</th>
      <th>nAA</th>
      <th>decoy</th>
      <th>charge</th>
      <th>precursor_mz</th>
      <th>isotope_intensity_m1</th>
      <th>isotope_intensity_m2</th>
      <th>isotope_apex_intensity</th>
      <th>isotope_apex_index</th>
      <th>isotope_mz_m1</th>
      <th>isotope_mz_m2</th>
      <th>isotope_apex_mz</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>LMNPQRST</td>
      <td>0</td>
      <td>1</td>
      <td>False</td>
      <td>True</td>
      <td>Oxidation@M</td>
      <td>2</td>
      <td>8</td>
      <td>0</td>
      <td>2</td>
      <td>481.739834</td>
      <td>0.478814</td>
      <td>0.183746</td>
      <td>1.0</td>
      <td>0</td>
      <td>482.241484</td>
      <td>482.743134</td>
      <td>481.739834</td>
    </tr>
    <tr>
      <th>1</th>
      <td>LMNPQRST</td>
      <td>0</td>
      <td>1</td>
      <td>False</td>
      <td>True</td>
      <td>Oxidation@M</td>
      <td>2</td>
      <td>8</td>
      <td>0</td>
      <td>3</td>
      <td>321.495648</td>
      <td>0.478814</td>
      <td>0.183746</td>
      <td>1.0</td>
      <td>0</td>
      <td>321.830081</td>
      <td>322.164515</td>
      <td>321.495648</td>
    </tr>
    <tr>
      <th>2</th>
      <td>LMNPQRST</td>
      <td>0</td>
      <td>1</td>
      <td>False</td>
      <td>True</td>
      <td>Oxidation@M</td>
      <td>2</td>
      <td>8</td>
      <td>0</td>
      <td>4</td>
      <td>241.373555</td>
      <td>0.478814</td>
      <td>0.183746</td>
      <td>1.0</td>
      <td>0</td>
      <td>241.624380</td>
      <td>241.875205</td>
      <td>241.373555</td>
    </tr>
    <tr>
      <th>3</th>
      <td>LMNPQRST</td>
      <td>0</td>
      <td>1</td>
      <td>False</td>
      <td>True</td>
      <td></td>
      <td></td>
      <td>8</td>
      <td>0</td>
      <td>2</td>
      <td>473.742377</td>
      <td>0.478433</td>
      <td>0.181509</td>
      <td>1.0</td>
      <td>0</td>
      <td>474.244027</td>
      <td>474.745677</td>
      <td>473.742377</td>
    </tr>
    <tr>
      <th>4</th>
      <td>LMNPQRST</td>
      <td>0</td>
      <td>1</td>
      <td>False</td>
      <td>True</td>
      <td></td>
      <td></td>
      <td>8</td>
      <td>0</td>
      <td>3</td>
      <td>316.164010</td>
      <td>0.478433</td>
      <td>0.181509</td>
      <td>1.0</td>
      <td>0</td>
      <td>316.498443</td>
      <td>316.832877</td>
      <td>316.164010</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>79</th>
      <td>FGHIKLMNPQRST</td>
      <td>0</td>
      <td>2</td>
      <td>False</td>
      <td>True</td>
      <td>Oxidation@M</td>
      <td>7</td>
      <td>13</td>
      <td>0</td>
      <td>3</td>
      <td>515.604920</td>
      <td>0.828432</td>
      <td>0.420789</td>
      <td>1.0</td>
      <td>0</td>
      <td>515.939354</td>
      <td>516.273787</td>
      <td>515.604920</td>
    </tr>
    <tr>
      <th>80</th>
      <td>FGHIKLMNPQRST</td>
      <td>0</td>
      <td>2</td>
      <td>False</td>
      <td>True</td>
      <td>Oxidation@M</td>
      <td>7</td>
      <td>13</td>
      <td>0</td>
      <td>4</td>
      <td>386.955509</td>
      <td>0.828432</td>
      <td>0.420789</td>
      <td>1.0</td>
      <td>0</td>
      <td>387.206334</td>
      <td>387.457159</td>
      <td>386.955509</td>
    </tr>
    <tr>
      <th>81</th>
      <td>FGHIKLMNPQRST</td>
      <td>0</td>
      <td>2</td>
      <td>False</td>
      <td>True</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0</td>
      <td>2</td>
      <td>764.906285</td>
      <td>0.828051</td>
      <td>0.418418</td>
      <td>1.0</td>
      <td>0</td>
      <td>765.407935</td>
      <td>765.909585</td>
      <td>764.906285</td>
    </tr>
    <tr>
      <th>82</th>
      <td>FGHIKLMNPQRST</td>
      <td>0</td>
      <td>2</td>
      <td>False</td>
      <td>True</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0</td>
      <td>3</td>
      <td>510.273282</td>
      <td>0.828051</td>
      <td>0.418418</td>
      <td>1.0</td>
      <td>0</td>
      <td>510.607715</td>
      <td>510.942149</td>
      <td>510.273282</td>
    </tr>
    <tr>
      <th>83</th>
      <td>FGHIKLMNPQRST</td>
      <td>0</td>
      <td>2</td>
      <td>False</td>
      <td>True</td>
      <td></td>
      <td></td>
      <td>13</td>
      <td>0</td>
      <td>4</td>
      <td>382.956781</td>
      <td>0.828051</td>
      <td>0.418418</td>
      <td>1.0</td>
      <td>0</td>
      <td>383.207606</td>
      <td>383.458431</td>
      <td>382.956781</td>
    </tr>
  </tbody>
</table>
<p>84 rows  18 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">_lib</span><span class="o">.</span><span class="n">import_and_process_protein_dict</span><span class="p">(</span><span class="n">protein_dict</span><span class="p">)</span>
<span class="n">_lib</span><span class="o">.</span><span class="n">add_peptide_labeling</span><span class="p">({</span>
    <span class="s1">&#39;light&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;Dimethyl@Any N-term&#39;</span><span class="p">,</span><span class="s1">&#39;Dimethyl@K&#39;</span><span class="p">],</span>
    <span class="s1">&#39;heavy&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;Dimethyl:2H(6)13C(2)@Any N-term&#39;</span><span class="p">,</span><span class="s1">&#39;Dimethyl:2H(6)13C(2)@K&#39;</span><span class="p">],</span>
<span class="p">})</span>
<span class="n">_lib</span><span class="o">.</span><span class="n">calc_precursor_isotope</span><span class="p">()</span>
<span class="k">assert</span> <span class="p">(</span><span class="n">_lib</span><span class="o">.</span><span class="n">precursor_df</span><span class="o">.</span><span class="n">decoy</span><span class="o">==</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">()</span>
<span class="k">assert</span> <span class="p">(</span><span class="s1">&#39;MACDESTY&#39;</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;K&#39;</span><span class="p">)</span> <span class="ow">in</span> <span class="n">_lib</span><span class="o">.</span><span class="n">precursor_df</span><span class="o">.</span><span class="n">sequence</span><span class="o">.</span><span class="n">values</span>
<span class="k">assert</span> <span class="s1">&#39;isotope_apex_index&#39;</span> <span class="ow">in</span> <span class="n">_lib</span><span class="o">.</span><span class="n">precursor_df</span><span class="o">.</span><span class="n">columns</span>
<span class="k">assert</span> <span class="s1">&#39;isotope_apex_intensity&#39;</span> <span class="ow">in</span> <span class="n">_lib</span><span class="o">.</span><span class="n">precursor_df</span><span class="o">.</span><span class="n">columns</span>
<span class="k">assert</span> <span class="o">~</span><span class="n">_lib</span><span class="o">.</span><span class="n">precursor_df</span><span class="o">.</span><span class="n">sequence</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;B&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">()</span>
<span class="n">_lib</span><span class="o">.</span><span class="n">precursor_df</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sequence</th>
      <th>protein_idxes</th>
      <th>miss_cleavage</th>
      <th>is_prot_nterm</th>
      <th>is_prot_cterm</th>
      <th>mods</th>
      <th>mod_sites</th>
      <th>nAA</th>
      <th>decoy</th>
      <th>charge</th>
      <th>label_channel</th>
      <th>precursor_mz</th>
      <th>isotope_intensity_m1</th>
      <th>isotope_intensity_m2</th>
      <th>isotope_apex_intensity</th>
      <th>isotope_apex_index</th>
      <th>isotope_mz_m1</th>
      <th>isotope_mz_m2</th>
      <th>isotope_apex_mz</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>LMNPQRST</td>
      <td>0</td>
      <td>1</td>
      <td>False</td>
      <td>True</td>
      <td>Oxidation@M;Dimethyl@Any N-term</td>
      <td>2;0</td>
      <td>8</td>
      <td>0</td>
      <td>2</td>
      <td>light</td>
      <td>495.755484</td>
      <td>0.500906</td>
      <td>0.194451</td>
      <td>1.0</td>
      <td>0</td>
      <td>496.257134</td>
      <td>496.758784</td>
      <td>495.755484</td>
    </tr>
    <tr>
      <th>1</th>
      <td>LMNPQRST</td>
      <td>0</td>
      <td>1</td>
      <td>False</td>
      <td>True</td>
      <td>Oxidation@M;Dimethyl@Any N-term</td>
      <td>2;0</td>
      <td>8</td>
      <td>0</td>
      <td>3</td>
      <td>light</td>
      <td>330.839415</td>
      <td>0.500906</td>
      <td>0.194451</td>
      <td>1.0</td>
      <td>0</td>
      <td>331.173848</td>
      <td>331.508282</td>
      <td>330.839415</td>
    </tr>
    <tr>
      <th>2</th>
      <td>LMNPQRST</td>
      <td>0</td>
      <td>1</td>
      <td>False</td>
      <td>True</td>
      <td>Oxidation@M;Dimethyl@Any N-term</td>
      <td>2;0</td>
      <td>8</td>
      <td>0</td>
      <td>4</td>
      <td>light</td>
      <td>248.381380</td>
      <td>0.500906</td>
      <td>0.194451</td>
      <td>1.0</td>
      <td>0</td>
      <td>248.632205</td>
      <td>248.883030</td>
      <td>248.381380</td>
    </tr>
    <tr>
      <th>3</th>
      <td>LMNPQRST</td>
      <td>0</td>
      <td>1</td>
      <td>False</td>
      <td>True</td>
      <td>Dimethyl@Any N-term</td>
      <td>0</td>
      <td>8</td>
      <td>0</td>
      <td>2</td>
      <td>light</td>
      <td>487.758027</td>
      <td>0.500525</td>
      <td>0.192205</td>
      <td>1.0</td>
      <td>0</td>
      <td>488.259677</td>
      <td>488.761327</td>
      <td>487.758027</td>
    </tr>
    <tr>
      <th>4</th>
      <td>LMNPQRST</td>
      <td>0</td>
      <td>1</td>
      <td>False</td>
      <td>True</td>
      <td>Dimethyl@Any N-term</td>
      <td>0</td>
      <td>8</td>
      <td>0</td>
      <td>3</td>
      <td>light</td>
      <td>325.507777</td>
      <td>0.500525</td>
      <td>0.192205</td>
      <td>1.0</td>
      <td>0</td>
      <td>325.842210</td>
      <td>326.176643</td>
      <td>325.507777</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>163</th>
      <td>FGHIKLMNPQRST</td>
      <td>0</td>
      <td>2</td>
      <td>False</td>
      <td>True</td>
      <td>Oxidation@M;Dimethyl:2H(6)13C(2)@Any N-term;Di...</td>
      <td>7;0;5</td>
      <td>13</td>
      <td>0</td>
      <td>3</td>
      <td>heavy</td>
      <td>539.655367</td>
      <td>0.788273</td>
      <td>0.392103</td>
      <td>1.0</td>
      <td>0</td>
      <td>539.989801</td>
      <td>540.324234</td>
      <td>539.655367</td>
    </tr>
    <tr>
      <th>164</th>
      <td>FGHIKLMNPQRST</td>
      <td>0</td>
      <td>2</td>
      <td>False</td>
      <td>True</td>
      <td>Oxidation@M;Dimethyl:2H(6)13C(2)@Any N-term;Di...</td>
      <td>7;0;5</td>
      <td>13</td>
      <td>0</td>
      <td>4</td>
      <td>heavy</td>
      <td>404.993344</td>
      <td>0.788273</td>
      <td>0.392103</td>
      <td>1.0</td>
      <td>0</td>
      <td>405.244169</td>
      <td>405.494994</td>
      <td>404.993344</td>
    </tr>
    <tr>
      <th>165</th>
      <td>FGHIKLMNPQRST</td>
      <td>0</td>
      <td>2</td>
      <td>False</td>
      <td>True</td>
      <td>Dimethyl:2H(6)13C(2)@Any N-term;Dimethyl:2H(6)...</td>
      <td>0;5</td>
      <td>13</td>
      <td>0</td>
      <td>2</td>
      <td>heavy</td>
      <td>800.981955</td>
      <td>0.787646</td>
      <td>0.389779</td>
      <td>1.0</td>
      <td>0</td>
      <td>801.483605</td>
      <td>801.985255</td>
      <td>800.981955</td>
    </tr>
    <tr>
      <th>166</th>
      <td>FGHIKLMNPQRST</td>
      <td>0</td>
      <td>2</td>
      <td>False</td>
      <td>True</td>
      <td>Dimethyl:2H(6)13C(2)@Any N-term;Dimethyl:2H(6)...</td>
      <td>0;5</td>
      <td>13</td>
      <td>0</td>
      <td>3</td>
      <td>heavy</td>
      <td>534.323729</td>
      <td>0.787646</td>
      <td>0.389779</td>
      <td>1.0</td>
      <td>0</td>
      <td>534.658162</td>
      <td>534.992596</td>
      <td>534.323729</td>
    </tr>
    <tr>
      <th>167</th>
      <td>FGHIKLMNPQRST</td>
      <td>0</td>
      <td>2</td>
      <td>False</td>
      <td>True</td>
      <td>Dimethyl:2H(6)13C(2)@Any N-term;Dimethyl:2H(6)...</td>
      <td>0;5</td>
      <td>13</td>
      <td>0</td>
      <td>4</td>
      <td>heavy</td>
      <td>400.994616</td>
      <td>0.787646</td>
      <td>0.389779</td>
      <td>1.0</td>
      <td>0</td>
      <td>401.245441</td>
      <td>401.496266</td>
      <td>400.994616</td>
    </tr>
  </tbody>
</table>
<p>168 rows  19 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

</div>
 

